// @orca-so/sdk/dist/public/pools/config.js
"use strict";
/**
 * The following content is auto-generated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrcaPoolConfig = void 0;
/**
 * A list of supported Orca pools in this SDK.
 */
var OrcaPoolConfig;
(function (OrcaPoolConfig) {
    OrcaPoolConfig["SOL_USDC"] = "APDFRM3HMr8CAGXwKHiu2f5ePSpaiEJhaURwhsRrUUt9";
    OrcaPoolConfig["SOL_USDT"] = "FZthQCuYHhcfiDma7QrX7buDHwrZEd7vL8SjS6LQa3Tx";
    OrcaPoolConfig["ETH_SOL"] = "71FymgN2ZUf7VvVTLE8jYEnjP3jSK1Frp2XT1nHs8Hob";
    OrcaPoolConfig["ETH_USDC"] = "3e1W6Aqcbuk2DfHUwRiRcyzpyYRRjg6yhZZcyEARydUX";
    OrcaPoolConfig["RAY_SOL"] = "5kimD5W6yJpHRHCyPtnEyDsQRdiiJKivu5AqN3si82Jc";
    OrcaPoolConfig["ROPE_SOL"] = "ADrvfPBsRcJfGsN6Bs385zYddH52nuM5FA8UaAkX9o2V";
    OrcaPoolConfig["STEP_SOL"] = "8nTzqDXHriG2CXKbybeuEh1EqDQMtrbYMFWcP7AkiDaP";
    OrcaPoolConfig["SRM_SOL"] = "9tf8rBSEQYG7AqL896fN2nZi1iYPqpWaLEdpbeQaC1Vy";
    OrcaPoolConfig["FTT_SOL"] = "EsYaDKJCmcJtJHFuJYwQZwqohvVMCrFzcg8yo3i328No";
    OrcaPoolConfig["COPE_SOL"] = "CzieDbGRdN1QGaGDNpSqzEA18bi881ccvkkGZi51pe1k";
    OrcaPoolConfig["OXY_SOL"] = "7tYCdLN84EnTMkxM7HNamWJx7F4xgKe9KiiWvLyWjbgT";
    OrcaPoolConfig["BTC_SOL"] = "Acxs19v6eUMTEfdvkvWkRB4bwFCHm3XV9jABCy7c1mXe";
    OrcaPoolConfig["MER_SOL"] = "HiwRobjfHZ4zsPtqCC4oBS24pSmy4t8GGkXRbQj4yU6L";
    OrcaPoolConfig["FIDA_SOL"] = "EYsNdtyu4gGTaGz8N5m5iQ3G1N6rDyMbR72B3CqbWW4W";
    OrcaPoolConfig["MAPS_SOL"] = "99pfC8fWymXgbq3CvrExhx3UxQDC1fMWEWLbNT83F45e";
    OrcaPoolConfig["USDC_USDT"] = "H2uzgruPvonVpCRhwwdukcpXK8TG17swFNzYFr2rtPxy";
    OrcaPoolConfig["ORCA_SOL"] = "2uVjAuRXavpM6h1scGQaxqb6HVaNRn6T2X7HHXTabz25";
    OrcaPoolConfig["ORCA_USDC"] = "n8Mpu28RjeYD7oUX3LG1tPxzhRZh3YYLRSHcHRdS3Zx";
    OrcaPoolConfig["KIN_SOL"] = "HEvnD66WcBfTajS9adUYnGRBMDehFtLySiFHSD6kEBWs";
    OrcaPoolConfig["SAMO_SOL"] = "D6N9j8F2DhtzDtrdpT74y3u2YmYAzcggiLc3nTjqux9M";
    OrcaPoolConfig["LIQ_USDC"] = "3PD9SZFwXKkXr4akLf4ofo37ZUMycwML89R2P3qxcbZG";
    OrcaPoolConfig["SNY_USDC"] = "AZpo4BJHHRetF96v6SGinFZBMXM4yWMo4RA8C4PriDLk";
    OrcaPoolConfig["mSOL_USDC"] = "8PSfyiTVwPb6Rr2iZ8F3kNpbg65BCfJM9v8LfB916r44";
    OrcaPoolConfig["SLRS_USDC"] = "AtB4nUmdyQfuWWJ9xAHw9xyVnJFfSjSuVWkiYan8y86w";
    OrcaPoolConfig["PORT_USDC"] = "F8gPSpwVHj8FdAJAYULDuZBxFEJut87hUbARYYx3471w";
    OrcaPoolConfig["SBR_USDC"] = "CS7fA5n4c2D82dUoHrYzS3gAqgqaoVSfgsr18kitp2xo";
    OrcaPoolConfig["scnSOL_USDC"] = "Dkr8B675PGnNwEr9vTKXznjjHke5454EQdz3iaSbparB";
    OrcaPoolConfig["pSOL_USDC"] = "C2YzN6MymD5HM2kPaH7bzcbqciyjfmpqyVaR3KA5V6z1";
    OrcaPoolConfig["mSOL_SOL"] = "29cdoMgu6MS2VXpcMo1sqRdWEzdUR9tjvoh8fcK8Z87R";
    OrcaPoolConfig["ORCA_PAI"] = "C7TH2jEJJaxVwwuvkbcDGfHUiZvEkkeYjyAcdTMi5ujb";
    OrcaPoolConfig["ORCA_mSOL"] = "CVapmQn7HaU1yMDW3q6oUV4hx6XoYv54T4zfGXkuJqkA";
    OrcaPoolConfig["scnSOL_SOL"] = "APNpzQvR91v1THbsAyG3HHrUEwvexWYeNCFLQuVnxgMc";
    OrcaPoolConfig["ATLAS_USDC"] = "FZ8x1LCRSPDeHBDoAc3Gc6Y7ETCynuHEr5q5YWV7uRCJ";
    OrcaPoolConfig["POLIS_USDC"] = "GteBdo9sqE7T41G8AJsaG9WHW48uXBwsLLznmu2TBdgy";
    OrcaPoolConfig["BOP_USDC"] = "2gXDJZ7XAtQEtf4PRSQZKoq1WMuu1H44tQanbMA3YVpu";
    OrcaPoolConfig["SAMO_USDC"] = "6VK1ksrmYGMBWUUZfygGF8tHRGpNxQEWv8pfvzQHdyyc";
    OrcaPoolConfig["NINJA_SOL"] = "4X1oYoFWYtLebk51zuh889r1WFLe8Z9qWApj87hQMfML";
    OrcaPoolConfig["SLIM_USDC"] = "BVWwyiHVHZQMPHsiW7dZH7bnBVKmbxdeEjWqVRciHCyo";
    OrcaPoolConfig["wHAPI_USDC"] = "ELfBngAgvLEHVBuJQhhE7AW6eqLX7id2sfrBngVNVAUW";
    OrcaPoolConfig["COPE_USDC"] = "HsauTv9s52Zv12eaDuSp6y7BEm4e4BHEyAsbdjyyWzPK";
    OrcaPoolConfig["SUNNY_USDC"] = "GHuoeq9UnFBsBhMwH43eL3RWX5XVXbSRYJymmyMYpT7n";
    OrcaPoolConfig["GRAPE_USDC"] = "EorFh8siFyLF1QTZ7cCXQaPGqyo7eb4SAgKtRH8Jcxjd";
    OrcaPoolConfig["ABR_USDC"] = "GMzPbaCuQmeMUm1opH3oSCgKUjVgJUW14myq99RVPGX5";
    OrcaPoolConfig["KURO_USDC"] = "DRknxb4ZFxXUTG6UJ5HupNHG1SmvBSCPzsZ1o9gAhyBi";
    OrcaPoolConfig["MEDIA_USDC"] = "2toFgkQDoPrTJYGDEVoCasPXuL9uQnjvXJaDwa9LHyTx";
    OrcaPoolConfig["TULIP_USDC"] = "4SBx8GXu8HhcVHWydQv1vsDdZs3G93XSL9CtMBny6hS5";
    OrcaPoolConfig["MNGO_USDC"] = "H9yC7jDng974WwcU4kTGs7BKf7nBNswpdsP5bzbdXjib";
    OrcaPoolConfig["stSOL_wstETH"] = "Eswigpwm3xsipkTqahGi2PEJsJcULQBwZgxhQpr6yBEa";
    OrcaPoolConfig["SYP_USDC"] = "qJxKN9BhxbYvRNbjfK2uAVWboto6sonj8XC1ZEW5XTB";
    OrcaPoolConfig["stSOL_wLDO"] = "74B9aMS7SA832xKngt5VLKmWAP3pa3qkUzWncTmQSsGF";
    OrcaPoolConfig["whETH_SOL"] = "7aYnrdmdCRodDy2Czn6keUquUhjF1jPEmfwZPh488z8U";
    OrcaPoolConfig["whETH_USDC"] = "7NPtjjAP7vhp4t5NCLyY4DY5rurvyc8cgZ2a2rYabRia";
    OrcaPoolConfig["MNDE_mSOL"] = "5PHS5w6hQwFNnLz1jJFe7TVTxSQ98cDYC3akmiAoFMXs";
    OrcaPoolConfig["WAG_USDC"] = "Df6XNHMF3uRVZnz7LCEGiZVax6rXgz76owtVkBHEjSb6";
    OrcaPoolConfig["mSOL_USDT"] = "9cMWe4UYRPGAUUsTkjShJWVM7bk8DUBgxtwwH8asFJoV";
    OrcaPoolConfig["mSOL_whETH"] = "5qoTq3qC4U7vFxo3iCzbXcaD1UEmDeCD63Dsuoct71oV";
    OrcaPoolConfig["BTC_mSOL"] = "8nKJ4z9FSw6wrVZKASqBiS9DS1CiNsRnqwCCKVQjqdkB";
    OrcaPoolConfig["IVN_SOL"] = "DfgCnzaiTXfPkAH1C1Z441b5MzjjTCEh134ioxqRZxYf";
    OrcaPoolConfig["LARIX_USDC"] = "8sfThep3io4gvcGeuoAg1Rs8GDwKJjtcdAFHqQSSNAVE";
    OrcaPoolConfig["PRT_USDC"] = "6jCERp5hKj37PCXP3VTjCDJeoPuSpnMDMz5A6jWQv3yS";
    OrcaPoolConfig["JET_USDC"] = "GBijunwxa4Ni3JmYC6q6zgaVhSUJU6hVX5qTyJDRpNTc";
    OrcaPoolConfig["stSOL_USDC"] = "GtQ1NT7R5aaTiST7K6ZWdMhwDdFxsSFvVFhBo8vyHGAq";
    OrcaPoolConfig["wstETH_USDC"] = "5a6Y1ephcbKSoyLMQyD1JWbtqawCy8p2FtRL9v3zhaG5";
    OrcaPoolConfig["AURY_USDC"] = "6mJqqT5TMgveDvxzBt3hrjGkPV5VAj7tacxFCT3GebXh";
    OrcaPoolConfig["AVAX_USDC"] = "Hmfrtmo93DpSDmVNLQKcBS5D1ia5JatiRSok9ososubz";
    OrcaPoolConfig["FTT_USDC"] = "FwCombynV2fTVizxPCNA2oZKoWXLZgdJThjE4Xv9sjxc";
    OrcaPoolConfig["RAY_USDC"] = "4cXw2MYj94TFBXLL73fEpMCr8DPrW68JvrV8mzWgktbD";
    OrcaPoolConfig["SLND_USDC"] = "F59gkD7NnsdJbFKrRZsiBC8PAooN4c56T8QmahfW1iXN";
    OrcaPoolConfig["GOFX_USDC"] = "7vnps4VE5RTGAr5fmPZu7fSrk2VnM4Up838grZfqmxqE";
    OrcaPoolConfig["WOOF_USDC"] = "9EjcYfHcG8f1mccpHyaAwpoxaUPiheC6KgLQjyD9aTb6";
    OrcaPoolConfig["SDOGE_USDC"] = "CHTKUJGYRtBDqnxCFjxe5KEkZgxV98udbhuYYyzGxup5";
    OrcaPoolConfig["CATO_USDC"] = "55r9txzQtmjTykmTXmBYZCVMg5z9squB8b5cSw2AhxA4";
    OrcaPoolConfig["OOGI_USDC"] = "DSiHyHDn96bUQSZtizyCRLcQzrwohZeMpVu8rYJN1HzG";
    OrcaPoolConfig["SONAR_USDC"] = "5MvQHx8eftU39JTucFsT315JFnQASuDQg3FqxTw7xcvN";
    OrcaPoolConfig["APT_USDC"] = "HNrYngS1eoqkjWro9D3Y5Z9sWBDzPNK2tX4rfV2Up177";
    OrcaPoolConfig["DFL_USDC"] = "AWrtTWG4Zgxw8D92bb3L3sQtGLD3zDztMPWsXSph8iBP";
    OrcaPoolConfig["DFL_SOL"] = "9Y1vPaAsMz8X65DebMMnmBjbMo8i4jh4mcgiggZUUS3M";
    OrcaPoolConfig["FRKT_USDC"] = "FnDxJPNk7pPmGHUbR4XUHmHevrkXHdna5D3sQKcAtjBL";
    OrcaPoolConfig["TTT_USDC"] = "FGgP1npQTsC5Q4xBmQtNYSh51NKqNwdxBZy8JCo3igcu";
    OrcaPoolConfig["UPS_USDC"] = "E1U63VXhNiWoUkVvjrfLDdV1oJrwE6zLde3bohr6jCqz";
    OrcaPoolConfig["FANT_USDC"] = "GjG7JjTQfQpDxw4hWx4etP9oTaYCuCbPjsU8WaUT3xHB";
    OrcaPoolConfig["BLOCK_USDC"] = "D8WjqtwC9CzBrQKfSf2ccCHFQuPYwyLv5KAy8WjT5vnf";
    OrcaPoolConfig["RUN_USDC"] = "34Ppq6R8NfYBwWwPY4cBK4Afyb8hHaASQFukCzH6cV4n";
    OrcaPoolConfig["UXP_USDC"] = "HjR8JgqNKQVMvdryqJw5RJ4PCE9WGk8sgbEF7S9S3obv";
    OrcaPoolConfig["BTC_USDC"] = "J3kvcay3N16FBdawgnqoJ9v9p6XCvyCLE2Z9F5RLvGkj";
    OrcaPoolConfig["MNDE_USDC"] = "12Uj74zgUUoBe4yeackwQ4qYtFMr9fk1xL6q5Nha6t2N";
    OrcaPoolConfig["CHICKS_USDC"] = "71CBZeJ4tw38L9pSPoCz4fRsuWE64Fipyzotte7haoCS";
    OrcaPoolConfig["ONESOL_USDC"] = "6MF5CHWAj5mS7FhpxiKz37CzR2eYTu236XpBKKMXCrGg";
    OrcaPoolConfig["WMP_USDC"] = "HDgxKmiA8Pv82fNguhVeMkZqQkos2YksFPoP1KttWxX8";
    OrcaPoolConfig["UNQ_USDC"] = "2VuGzaMrDnDyZfYvDwSXk38s7M2wpud7LDY3dGA1J9sy";
    OrcaPoolConfig["BASIS_USDC"] = "GoaAiajubRgeCFEz9L6mLnSmT2QFegoJDH5tpLfivpj";
    OrcaPoolConfig["GST_USDC"] = "E6FUnQHGHJVJg7oExVr5Moeaj1QpdpZQF5odYjHXWPZb";
    OrcaPoolConfig["MEAN_USDC"] = "F5BTnwuMA6rxftTdbZ33VWKr2wrr6DuQHnd4guKmPSYQ";
    OrcaPoolConfig["AART_USDC"] = "HCtyJzFUtYecXrA52s4Y9atq4J1fhT3cYsTX17XVSFag";
    OrcaPoolConfig["SHDW_USDC"] = "DJqqvzSuPaWThfzwMjXx7H2ZmHDdwxza6NtFudtuXcpc";
    OrcaPoolConfig["SHDW_SOL"] = "2ws7g3LBPdctfKn42Di9qxzQtUJ8ZL1aEAX2rGEQMNqh";
    OrcaPoolConfig["SCY_USDC"] = "99ZHUQsgxL7K6PHrGNi1gSwawwPr7UA5fbWrYoHQ6qhX";
    OrcaPoolConfig["SLC_USDC"] = "E5kSBqTDxFLbLNQaVVtPtnhEYVLMCK2fVSEKoMKL98qR";
    OrcaPoolConfig["wUST_SOL"] = "6c13xsmyk7UaHUWZ2rm1MM3ZdrQRSBkQ9waaG25ridVs";
    OrcaPoolConfig["wUST_USDC"] = "J1KfRtP5y2warpD7LdJhfBLPKoWwSqYuovdArSv1mpQ7";
    OrcaPoolConfig["mSOL_wUST"] = "68YVjgPnTUPcBqZyghqvD2WPNsrLKsjYTmBKJzHRr4qd";
    OrcaPoolConfig["wLUNA_wUST"] = "8Mh7drLbt3jFJYwp948XyvQscGLaLkChNcaH5wwaAoWA";
    OrcaPoolConfig["stSOL_wUST"] = "HTZd53fYwYQRyAjiaPsZy9Gf41gobFdqkF4oKe3XLi95";
    OrcaPoolConfig["JSOL_USDC"] = "AzEoVuNJyo9ByoLRZ5t6vav2Zg24vULNVJM41PgCKUqR";
    OrcaPoolConfig["daoSOL_USDC"] = "CCyDxjdW3G7hPTthTMPTZ4bnhFF19XG6rx2fNiKeRQww";
    OrcaPoolConfig["ORCA_USDT"] = "Gx4PoxenyQwhGGnKagAT35iVg4im1iKhJxDWqVhgu6tk";
    OrcaPoolConfig["ORCA_whETH"] = "GsfyYHkSgC3Ta6aWR9MjB2sxoBrkGGeR2tAwXbpphf3";
    OrcaPoolConfig["GENE_USDC"] = "7cuu94swKL5PtFQohKMAzyd1mjj65rgMW3GzLY31HCnK";
    OrcaPoolConfig["CMFI_USDC"] = "85krvT9DxdYgoFLQDHTAGdvtNuLdAsc4xE5FkVLpN2aR";
    OrcaPoolConfig["CELO_USDC"] = "HVLyX8mD8YvKgZJ4oB6rXJiCYMLpHKwB6iCiCjE1XwdT";
    OrcaPoolConfig["FTM_USDC"] = "Gpzd833qSmv3kXpQmxEaqkrZTXZaRjhNAoqhf61qAhTG";
    OrcaPoolConfig["BTC_ORCA"] = "DFpLFcQZqDKykyDePgip4r6MExVmBKWqTa12ezq6qxUY";
    OrcaPoolConfig["HBB_USDC"] = "cL5WhffCYFRLM4We8VS2W684kM4pHyuvEDwp8Ddw48k";
    OrcaPoolConfig["HBB_SOL"] = "FkKzu2HeMJZf4oHwoYPxLGVy3net5Jq8HAfnA5VqETgk";
    OrcaPoolConfig["SB_USDC"] = "2Reqt4Sw9xNY8BoJ3EZLpFu5yVgNxFrbw8M3KiJpPn6o";
    OrcaPoolConfig["stSOL_USDT"] = "4ni1nho89cDKAQ9ddbNQA9ieLYpzvJVmJpuogu5Ct5ur";
    OrcaPoolConfig["SEEDED_USDC"] = "H7gyTmNCDXkD8MGMqnxqoD8ANszjcju4tjT6ERZ5dakf";
    OrcaPoolConfig["AUDIO_USDC"] = "3hksYA17VxgiKSeihjnZkBbjc2CTbEBfvDCYgQhojTo5";
    OrcaPoolConfig["MMA_USDC"] = "AaZRnJAnDyJyPD9uPJpJ8bzBGDCEi6jtBpUf92xErWPp";
    OrcaPoolConfig["ONESOL_SOL"] = "9wPhuYapychVDSxmXqCZxy2Ka8Lmav4SHM72si8bfraV";
    OrcaPoolConfig["PUFF_SOL"] = "Eho8h1BcoG5QWU7X9FzJafw5ErKUXtR2LobAJJZfWff4";
    OrcaPoolConfig["SAO_USDC"] = "4iyU77yZbg8iD344vbwruAuDAf9i1EVV3FhZJDnWStBE";
    OrcaPoolConfig["sRLYv2_SOL"] = "3dXdXg5HPyZ73GFC9LkSn3thdJUGeXWB8iSTHs5UcqiH";
    OrcaPoolConfig["ZBC_USDC"] = "2LYgm6nGXmSSjfoEriPuYeGoNiWNxUs7n3rnTbDWN5c7";
    OrcaPoolConfig["GMT_USDC"] = "CFxQF5kNAtbbDj298Xr47Sf4mkSyuzWpRH97hrdQ6kxi";
    OrcaPoolConfig["NOVA_USDC"] = "DdAFNvDxtEHCgKj3JAM64zCKfMWhs4J9wEmRrjUAFiME";
    OrcaPoolConfig["HBB_USDH"] = "7n2YW9qLkhGFArdZPLoF4hYPE2zw7xCACkVPXrUWnLuo";
    OrcaPoolConfig["TAKI_sRLYv2"] = "6atKbS2Xz5vu7cqWBNk8KYkuakRzckZ9nvtUKf2k8Sc3";
    OrcaPoolConfig["ZIG_USDC"] = "5vhh9ZnD9vnahRhFLP1EUEyYRSzvJwgw9U2xygsSJSrp";
    OrcaPoolConfig["sRLY_SOL"] = "HfRgvhgscGX5GaP3rUrZAhh7gS4aJ2UQ7rNVX976rG6P";
    OrcaPoolConfig["TAKI_sRLY"] = "3xaK5aWWLNRB73xUVX3cusLhDp65mTvP4fwW5Jwxakgs";
    OrcaPoolConfig["stSOL_SOL"] = "4jjQSgFx33DUb1a7pgPsi3FbtZXDQ94b6QywjNK3NtZw";
    OrcaPoolConfig["BTC_stSOL"] = "HYp9v7cY4wAxSsa6ijztQQ3GQ8iTttuG5vu8JNBDHoNh";
    OrcaPoolConfig["stSOL_whETH"] = "FWurFWADrgmhb6Y44LEaCMnEHS2Tu3QGqd9oBcZtr8gT";
})(OrcaPoolConfig = exports.OrcaPoolConfig || (exports.OrcaPoolConfig = {}));
