// @switchboard-xyz/switchboardv2-cli/dist/commands/lease/extend.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const anchor = tslib_1.__importStar(require("@project-serum/anchor"));
const web3_js_1 = require("@solana/web3.js");
const sbv2_utils_1 = require("@switchboard-xyz/sbv2-utils");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const BaseCommand_1 = tslib_1.__importDefault(require("../../BaseCommand"));
const utils_1 = require("../../utils");
class LeaseExtend extends BaseCommand_1.default {
    async run() {
        const { args, flags } = await this.parse(LeaseExtend);
        (0, sbv2_utils_1.verifyProgramHasPayer)(this.program);
        const payerKeypair = (0, switchboard_v2_1.programWallet)(this.program);
        const amount = this.getTokenAmount(flags.amount);
        if (amount.lte(new anchor.BN(0))) {
            throw new Error("amount to deposit must be greater than 0");
        }
        const aggregatorAccount = new switchboard_v2_1.AggregatorAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.aggregatorKey),
        });
        const aggregator = await aggregatorAccount.loadData();
        const queueAccount = new switchboard_v2_1.OracleQueueAccount({
            program: this.program,
            publicKey: aggregator.queuePubkey,
        });
        const mint = await queueAccount.loadMint();
        const [leaseAccount] = switchboard_v2_1.LeaseAccount.fromSeed(this.program, queueAccount, aggregatorAccount);
        try {
            const lease = await leaseAccount.loadData();
        }
        catch {
            throw new Error(`Failed to load lease account. Has it been created yet?`);
        }
        const lease = await leaseAccount.loadData();
        const escrow = lease.escrow;
        const initialLeaseBalance = await this.program.provider.connection.getTokenAccountBalance(escrow);
        const funderTokenAddress = await (0, sbv2_utils_1.getOrCreateSwitchboardTokenAccount)(this.program, mint);
        const initialFunderBalance = await this.program.provider.connection.getTokenAccountBalance(funderTokenAddress);
        if (!this.silent) {
            this.logger.log((0, sbv2_utils_1.chalkString)("Initial Lease Balance", initialLeaseBalance.value.uiAmountString, 24));
            this.logger.log((0, sbv2_utils_1.chalkString)("Initial Funder Balance", initialFunderBalance.value.uiAmountString, 24));
        }
        const txn = await leaseAccount.extend({
            loadAmount: amount,
            funder: funderTokenAddress,
            funderAuthority: payerKeypair,
        });
        if (!this.silent) {
            const newBalance = await this.program.provider.connection.getTokenAccountBalance(escrow);
            this.logger.log((0, sbv2_utils_1.chalkString)("Final Lease Balance", newBalance.value.uiAmountString, 24));
        }
        if (this.silent) {
            console.log(txn);
        }
        else {
            this.logger.log(`${chalk_1.default.green(`${utils_1.CHECK_ICON} Deposited ${amount} tokens into aggregator lease`)}`);
            this.logger.log(`https://explorer.solana.com/tx/${txn}?cluster=${this.cluster}`);
        }
    }
    async catch(error) {
        super.catch(error, "failed to deposit into aggregator lease account");
    }
}
exports.default = LeaseExtend;
LeaseExtend.description = "fund and re-enable an aggregator lease";
LeaseExtend.aliases = ["aggregator:lease:extend"];
LeaseExtend.flags = {
    ...BaseCommand_1.default.flags,
    amount: core_1.Flags.string({
        required: true,
        description: "token amount to load into the lease escrow. If decimals provided, amount will be normalized to raw tokenAmount",
    }),
};
LeaseExtend.args = [
    {
        name: "aggregatorKey",
        description: "public key of the aggregator to extend a lease for",
    },
];
LeaseExtend.examples = [
    "$ sbv2 aggregator:lease:extend GvDMxPzN1sCj7L26YDK2HnMRXEQmQ2aemov8YBtPS7vR --amount 1.1 --keypair ../payer-keypair.json",
];
