// @switchboard-xyz/switchboardv2-cli/dist/commands/aggregator/set/authority.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const web3_js_1 = require("@solana/web3.js");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const BaseCommand_1 = tslib_1.__importDefault(require("../../../BaseCommand"));
const utils_1 = require("../../../utils");
class AggregatorSetAuthority extends BaseCommand_1.default {
    //   static examples = ["$ sbv2 aggregator:set:authority"];
    async run() {
        const { args, flags } = await this.parse(AggregatorSetAuthority);
        const newAuthority = await (0, utils_1.loadKeypair)(args.newAuthority);
        const aggregatorAccount = new switchboard_v2_1.AggregatorAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.aggregatorKey),
        });
        const aggregator = await aggregatorAccount.loadData();
        const currentAuthority = await this.loadAuthority(flags.currentAuthority, aggregator.authority);
        const txn = await aggregatorAccount.setAuthority(newAuthority.publicKey, currentAuthority);
        if (this.silent) {
            console.log(txn);
        }
        else {
            this.logger.log(`${chalk_1.default.green(`${utils_1.CHECK_ICON}Aggregator authority set successfully`)}`);
            this.logger.log(`https://explorer.solana.com/tx/${txn}?cluster=${this.cluster}`);
        }
    }
    async catch(error) {
        super.catch(error, "failed to set aggregator authority");
    }
}
exports.default = AggregatorSetAuthority;
AggregatorSetAuthority.description = "set an aggregator's authority";
AggregatorSetAuthority.flags = {
    ...BaseCommand_1.default.flags,
    currentAuthority: core_1.Flags.string({
        char: "a",
        description: "alternate keypair that is the authority for the aggregator",
    }),
};
AggregatorSetAuthority.args = [
    {
        name: "aggregatorKey",
        description: "public key of the aggregator account",
    },
    {
        name: "newAuthority",
        description: "keypair path of new authority",
    },
];
