// @switchboard-xyz/switchboardv2-cli/dist/commands/aggregator/set/forceReportPeriod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const web3_js_1 = require("@solana/web3.js");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const BaseCommand_1 = tslib_1.__importDefault(require("../../../BaseCommand"));
const utils_1 = require("../../../utils");
class AggregatorSetForceReportPeriod extends BaseCommand_1.default {
    async run() {
        const { args, flags } = await this.parse(AggregatorSetForceReportPeriod);
        (0, utils_1.verifyProgramHasPayer)(this.program);
        const aggregatorAccount = new switchboard_v2_1.AggregatorAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.aggregatorKey),
        });
        const aggregator = await aggregatorAccount.loadData();
        const authority = await this.loadAuthority(flags.authority, aggregator.authority);
        const txn = await this.program.methods
            .aggregatorSetForceReportPeriod({
            forceReportPeriod: args.forceReportPeriod,
        })
            .accounts({
            aggregator: aggregatorAccount.publicKey,
            authority: authority.publicKey,
        })
            .signers([authority])
            .rpc();
        if (this.silent) {
            console.log(txn);
        }
        else {
            this.logger.log(`${chalk_1.default.green(`${utils_1.CHECK_ICON}Aggregator force report period set successfully`)}`);
            this.logger.log(`https://explorer.solana.com/tx/${txn}?cluster=${this.cluster}`);
        }
    }
    async catch(error) {
        super.catch(error, "failed to set aggregator's force report period");
    }
}
exports.default = AggregatorSetForceReportPeriod;
AggregatorSetForceReportPeriod.description = "set an aggregator's force report period";
AggregatorSetForceReportPeriod.aliases = ["aggregator:set:forceReport"];
AggregatorSetForceReportPeriod.flags = {
    ...BaseCommand_1.default.flags,
    authority: core_1.Flags.string({
        char: "a",
        description: "alternate keypair that is the authority for the aggregator",
    }),
};
AggregatorSetForceReportPeriod.args = [
    {
        name: "aggregatorKey",
        description: "public key of the aggregator",
    },
    {
        name: "forceReportPeriod",
        description: "Number of seconds for which, even if the variance threshold is not passed, accept new responses from oracles.",
    },
];
AggregatorSetForceReportPeriod.examples = [
    "$ sbv2 aggregator:set:forceReportPeriod GvDMxPzN1sCj7L26YDK2HnMRXEQmQ2aemov8YBtPS7vR 300 --keypair ../payer-keypair.json",
];
