// @switchboard-xyz/switchboardv2-cli/dist/commands/aggregator/set/history.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const web3_js_1 = require("@solana/web3.js");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const BaseCommand_1 = tslib_1.__importDefault(require("../../../BaseCommand"));
const utils_1 = require("../../../utils");
class AggregatorSetHistoryBuffer extends BaseCommand_1.default {
    async run() {
        const { args, flags } = await this.parse(AggregatorSetHistoryBuffer);
        (0, utils_1.verifyProgramHasPayer)(this.program);
        const aggregatorAccount = new switchboard_v2_1.AggregatorAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.aggregatorKey),
        });
        const aggregator = await aggregatorAccount.loadData();
        const authority = await this.loadAuthority(flags.authority, aggregator.authority);
        const size = Number.parseInt(args.size, 10);
        const txn = await aggregatorAccount.setHistoryBuffer({ authority, size });
        if (this.silent) {
            console.log(txn);
        }
        else {
            this.logger.log(`${chalk_1.default.green(`${utils_1.CHECK_ICON}Added a history buffer of size ${size} to aggregator successfully`)}`);
            this.logger.log(`https://explorer.solana.com/tx/${txn}?cluster=${this.cluster}`);
        }
    }
    async catch(error) {
        super.catch(error, "failed to add history buffer to aggregator");
    }
}
exports.default = AggregatorSetHistoryBuffer;
AggregatorSetHistoryBuffer.description = "set an aggregator's history buffer account to record the last N accepted results";
AggregatorSetHistoryBuffer.aliases = ["aggregator:add:history"];
AggregatorSetHistoryBuffer.flags = {
    ...BaseCommand_1.default.flags,
    authority: core_1.Flags.string({
        char: "a",
        description: "alternate keypair that is the authority for the aggregator",
    }),
};
AggregatorSetHistoryBuffer.args = [
    {
        name: "aggregatorKey",
        description: "public key of the aggregator to add to a crank",
    },
    {
        name: "size",
        description: "size of history buffer",
    },
];
AggregatorSetHistoryBuffer.examples = [
    "$ sbv2 aggregator:set:history GvDMxPzN1sCj7L26YDK2HnMRXEQmQ2aemov8YBtPS7vR 10000 --keypair ../payer-keypair.json",
];
