// @switchboard-xyz/switchboardv2-cli/dist/commands/aggregator/set/minJobs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const web3_js_1 = require("@solana/web3.js");
const sbv2_utils_1 = require("@switchboard-xyz/sbv2-utils");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const BaseCommand_1 = tslib_1.__importDefault(require("../../../BaseCommand"));
const utils_1 = require("../../../utils");
class AggregatorSetMinJobResults extends BaseCommand_1.default {
    //   static examples = ["$ sbv2 aggregator:set:authority"];
    async run() {
        const { args, flags } = await this.parse(AggregatorSetMinJobResults);
        (0, sbv2_utils_1.verifyProgramHasPayer)(this.program);
        const aggregatorAccount = new switchboard_v2_1.AggregatorAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.aggregatorKey),
        });
        const aggregator = await aggregatorAccount.loadData();
        const authority = await this.loadAuthority(flags.authority, aggregator.authority);
        const minJobResults = Number.parseInt(args.minJobResults, 10);
        if (minJobResults <= 0 || minJobResults > 16) {
            throw new Error(`Invalid min job size (1 - 16), ${minJobResults}`);
        }
        if (minJobResults > aggregator.jobPubkeysSize) {
            throw new Error(`Min jobs ${minJobResults} is greater than current number of jobs ${aggregator.jobPubkeysSize} `);
        }
        const txn = await aggregatorAccount.setMinJobs({
            authority,
            minJobResults,
        });
        if (this.silent) {
            console.log(txn);
        }
        else {
            this.logger.log(`${chalk_1.default.green(`${utils_1.CHECK_ICON}Aggregator min job set successfully`)}`);
            this.logger.log(`https://explorer.solana.com/tx/${txn}?cluster=${this.cluster}`);
        }
    }
    async catch(error) {
        super.catch(error, "failed to set aggregator min jobs");
    }
}
exports.default = AggregatorSetMinJobResults;
AggregatorSetMinJobResults.description = "set an aggregator's minimum number of jobs before an oracle responds";
AggregatorSetMinJobResults.flags = {
    ...BaseCommand_1.default.flags,
    authority: core_1.Flags.string({
        char: "a",
        description: "alternate keypair that is the authority for the aggregator",
    }),
};
AggregatorSetMinJobResults.args = [
    {
        name: "aggregatorKey",
        description: "public key of the aggregator account",
    },
    {
        name: "minJobResults",
        description: "number of jobs that must respond before an oracle responds",
    },
];
