// @switchboard-xyz/switchboardv2-cli/dist/commands/aggregator/set/minOracles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const web3_js_1 = require("@solana/web3.js");
const sbv2_utils_1 = require("@switchboard-xyz/sbv2-utils");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const BaseCommand_1 = tslib_1.__importDefault(require("../../../BaseCommand"));
const utils_1 = require("../../../utils");
class AggregatorSetMinOracleResults extends BaseCommand_1.default {
    //   static examples = ["$ sbv2 aggregator:set:authority"];
    async run() {
        const { args, flags } = await this.parse(AggregatorSetMinOracleResults);
        (0, sbv2_utils_1.verifyProgramHasPayer)(this.program);
        const aggregatorAccount = new switchboard_v2_1.AggregatorAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.aggregatorKey),
        });
        const aggregator = await aggregatorAccount.loadData();
        const minOracleResults = Number.parseInt(args.minOracleResults, 10);
        if (minOracleResults <= 0 || minOracleResults > 16) {
            throw new Error(`Invalid min oracle size (1 - 16), ${minOracleResults}`);
        }
        const authority = await this.loadAuthority(flags.authority, aggregator.authority);
        const txn = await aggregatorAccount.setMinOracles({
            authority,
            minOracleResults,
        });
        if (this.silent) {
            console.log(txn);
        }
        else {
            this.logger.log(`${chalk_1.default.green(`${utils_1.CHECK_ICON}Aggregator minimum oracles set successfully`)}`);
            this.logger.log(`https://explorer.solana.com/tx/${txn}?cluster=${this.cluster}`);
        }
    }
    async catch(error) {
        super.catch(error, "failed to set aggregator minimum oracles");
    }
}
exports.default = AggregatorSetMinOracleResults;
AggregatorSetMinOracleResults.description = "set an aggregator's minimum number of oracles that must respond before a result is accepted on-chain";
AggregatorSetMinOracleResults.flags = {
    ...BaseCommand_1.default.flags,
    authority: core_1.Flags.string({
        char: "a",
        description: "alternate keypair that is the authority for the aggregator",
    }),
};
AggregatorSetMinOracleResults.args = [
    {
        name: "aggregatorKey",
        description: "public key of the aggregator account",
    },
    {
        name: "minOracleResults",
        description: "number of oracles that must respond before a value is accepted on-chain",
    },
];
