// @switchboard-xyz/switchboardv2-cli/dist/commands/aggregator/set/updateInterval.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const web3_js_1 = require("@solana/web3.js");
const sbv2_utils_1 = require("@switchboard-xyz/sbv2-utils");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const BaseCommand_1 = tslib_1.__importDefault(require("../../../BaseCommand"));
const utils_1 = require("../../../utils");
class AggregatorSetUpdateInterval extends BaseCommand_1.default {
    async run() {
        const { args, flags } = await this.parse(AggregatorSetUpdateInterval);
        (0, sbv2_utils_1.verifyProgramHasPayer)(this.program);
        const aggregatorAccount = new switchboard_v2_1.AggregatorAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.aggregatorKey),
        });
        const aggregator = await aggregatorAccount.loadData();
        if (aggregator.minUpdateDelaySeconds === args.updateInterval) {
            throw new Error(`Aggregator already has a minUpdateDelaySeconds of ${args.updateInterval}`);
        }
        if (args.updateInterval < 5) {
            throw new Error(`Update interval should be greater than 5 seconds, ${args.updateInterval}`);
        }
        const authority = await this.loadAuthority(flags.authority, aggregator.authority);
        const txn = await aggregatorAccount.setUpdateInterval({
            newInterval: Number.parseInt(args.updateInterval, 10),
            authority,
        });
        if (this.silent) {
            console.log(txn);
        }
        else {
            this.logger.log(`${chalk_1.default.green(`${utils_1.CHECK_ICON}Aggregator minimum update delay set successfully`)}`);
            this.logger.log(`https://explorer.solana.com/tx/${txn}?cluster=${this.cluster}`);
        }
    }
    async catch(error) {
        super.catch(error, "failed to set aggregator minimum update delay");
    }
}
exports.default = AggregatorSetUpdateInterval;
AggregatorSetUpdateInterval.description = "set an aggregator's minimum update delay";
AggregatorSetUpdateInterval.flags = {
    ...BaseCommand_1.default.flags,
    authority: core_1.Flags.string({
        char: "a",
        description: "alternate keypair that is the authority for the aggregator",
    }),
};
AggregatorSetUpdateInterval.args = [
    {
        name: "aggregatorKey",
        description: "public key of the aggregator account",
    },
    {
        name: "updateInterval",
        description: "set an aggregator's minimum update delay",
    },
];
AggregatorSetUpdateInterval.examples = [
    "$ sbv2 aggregator:set:updateInterval GvDMxPzN1sCj7L26YDK2HnMRXEQmQ2aemov8YBtPS7vR 60 --keypair ../payer-keypair.json",
];
