// @switchboard-xyz/switchboardv2-cli/dist/commands/aggregator/set/varianceThreshold.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const web3_js_1 = require("@solana/web3.js");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const big_js_1 = tslib_1.__importDefault(require("big.js"));
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const BaseCommand_1 = tslib_1.__importDefault(require("../../../BaseCommand"));
const utils_1 = require("../../../utils");
class AggregatorSetVarianceThreshold extends BaseCommand_1.default {
    async run() {
        const { args, flags } = await this.parse(AggregatorSetVarianceThreshold);
        (0, utils_1.verifyProgramHasPayer)(this.program);
        const aggregatorAccount = new switchboard_v2_1.AggregatorAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.aggregatorKey),
        });
        const aggregator = await aggregatorAccount.loadData();
        const authority = await this.loadAuthority(flags.authority, aggregator.authority);
        const txn = await aggregatorAccount.setVarianceThreshold({
            authority,
            threshold: new big_js_1.default(args.varianceThreshold),
        });
        if (this.silent) {
            console.log(txn);
        }
        else {
            this.logger.log(`${chalk_1.default.green(`${utils_1.CHECK_ICON}Aggregator variance threshold set successfully`)}`);
            this.logger.log(`https://explorer.solana.com/tx/${txn}?cluster=${this.cluster}`);
        }
    }
    async catch(error) {
        super.catch(error, "failed to set aggregator's variance threshold");
    }
}
exports.default = AggregatorSetVarianceThreshold;
AggregatorSetVarianceThreshold.description = "set an aggregator's variance threshold";
AggregatorSetVarianceThreshold.aliases = ["aggregator:set:variance"];
AggregatorSetVarianceThreshold.flags = {
    ...BaseCommand_1.default.flags,
    authority: core_1.Flags.string({
        char: "a",
        description: "alternate keypair that is the authority for the aggregator",
    }),
};
AggregatorSetVarianceThreshold.args = [
    {
        name: "aggregatorKey",
        description: "public key of the aggregator",
    },
    {
        name: "varianceThreshold",
        description: "percentage change between a previous accepted result and the next round before an oracle reports a value on-chain. Used to conserve lease cost during low volatility",
    },
];
AggregatorSetVarianceThreshold.examples = [
    "$ sbv2 aggregator:set:varianceThreshold GvDMxPzN1sCj7L26YDK2HnMRXEQmQ2aemov8YBtPS7vR 0.1 --keypair ../payer-keypair.json",
];
