// @orca-so/sdk/dist/public/devnet/farms/config.js
"use strict";
/**
 * The following content is auto-generated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrcaFarmConfig = void 0;
/**
 * A list of supported Orca farms in this SDK.
 */
var OrcaFarmConfig;
(function (OrcaFarmConfig) {
    OrcaFarmConfig["SOL_USDC_AQ"] = "4GpUivZ2jvZqQ3vJRsoq5PwnYv6gdV9fJ9BzHT2JcRr7";
    OrcaFarmConfig["SOL_USDT_AQ"] = "2E4Mp6y2gFRteXiANnwzAJEhHwD3NX1wru3XvxJpGBq2";
    OrcaFarmConfig["USDC_USDT_AQ"] = "EBor1PR5XNGHjRVB6JDJuKVCQbvdr1CVZTaX1hTAdvQv";
    OrcaFarmConfig["ETH_SOL_AQ"] = "8sFnpd7mM1AWxP1LXX2FWbbkaVtAopBPmPNZ9y6172WL";
    OrcaFarmConfig["ETH_USDC_AQ"] = "9pRnvg7ihSJDLi6DGf3PLwr6xRRRrBPXsHYEgGL5hzgA";
    OrcaFarmConfig["ORCA_SOL_AQ"] = "CmDdQhusZWyi9fue27VSktYgkHefm3JXNdzc9kCpyvYi";
    OrcaFarmConfig["ORCA_USDC_AQ"] = "2ZEEntzoUN7XuMs88ukLGv5HRR1byL7wFWChryF5ZHri";
    OrcaFarmConfig["SOL_USDC_DD"] = "AtJF9fbQ1pbz76NYo2jtKjHuzzyaFToHLCizYk6UoHHL";
    OrcaFarmConfig["ETH_SOL_DD"] = "5NH9rNaoLKbPPP5zwCJTaMPkFACtfJA8wEmqWrXzgCMa";
    OrcaFarmConfig["ETH_USDC_DD"] = "EeqPcczEZH2cjYBnQyGXBQx1DGs7KG1pobwdPKcwALhD";
})(OrcaFarmConfig = exports.OrcaFarmConfig || (exports.OrcaFarmConfig = {}));
