// @orca-so/sdk/dist/constants/farms.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stsolWldoAqFarm = exports.sypUsdcAqFarm = exports.stsolWstethAqFarm = exports.mngoUsdcAqFarm = exports.tulipUsdcAqFarm = exports.mediaUsdcAqFarm = exports.kuroUsdcAqFarm = exports.abrUsdcAqFarm = exports.grapeUsdcAqFarm = exports.sunnyUsdcAqFarm = exports.copeUsdcAqFarm = exports.whapiUsdcAqFarm = exports.slimUsdcAqFarm = exports.ninjaSolAqFarm = exports.samoUsdcAqFarm = exports.bopUsdcAqFarm = exports.polisUsdcAqFarm = exports.atlasUsdcAqFarm = exports.scnsolSolAqFarm = exports.orcaMsolAqFarm = exports.orcaPaiAqFarm = exports.msolSolAqFarm = exports.psolUsdcAqFarm = exports.scnsolUsdcAqFarm = exports.sbrUsdcAqFarm = exports.portUsdcAqFarm = exports.slrsUsdcAqFarm = exports.msolUsdcAqFarm = exports.snyUsdcAqFarm = exports.liqUsdcAqFarm = exports.samoSolAqFarm = exports.kinSolAqFarm = exports.orcaUsdcAqFarm = exports.orcaSolAqFarm = exports.usdcUsdtAqFarm = exports.mapsSolAqFarm = exports.fidaSolAqFarm = exports.merSolAqFarm = exports.btcSolAqFarm = exports.oxySolAqFarm = exports.copeSolAqFarm = exports.fttSolAqFarm = exports.srmSolAqFarm = exports.stepSolAqFarm = exports.ropeSolAqFarm = exports.raySolAqFarm = exports.ethUsdcAqFarm = exports.ethSolAqFarm = exports.solUsdtAqFarm = exports.solUsdcAqFarm = void 0;
exports.wustUsdcAqFarm = exports.wustSolAqFarm = exports.slcUsdcAqFarm = exports.scyUsdcAqFarm = exports.shdwSolAqFarm = exports.shdwUsdcAqFarm = exports.aartUsdcAqFarm = exports.meanUsdcAqFarm = exports.gstUsdcAqFarm = exports.basisUsdcAqFarm = exports.unqUsdcAqFarm = exports.wmpUsdcAqFarm = exports.onesolUsdcAqFarm = exports.chicksUsdcAqFarm = exports.mndeUsdcAqFarm = exports.btcUsdcAqFarm = exports.uxpUsdcAqFarm = exports.runUsdcAqFarm = exports.blockUsdcAqFarm = exports.fantUsdcAqFarm = exports.upsUsdcAqFarm = exports.tttUsdcAqFarm = exports.frktUsdcAqFarm = exports.dflSolAqFarm = exports.dflUsdcAqFarm = exports.aptUsdcAqFarm = exports.sonarUsdcAqFarm = exports.oogiUsdcAqFarm = exports.catoUsdcAqFarm = exports.sdogeUsdcAqFarm = exports.woofUsdcAqFarm = exports.gofxUsdcAqFarm = exports.slndUsdcAqFarm = exports.rayUsdcAqFarm = exports.fttUsdcAqFarm = exports.avaxUsdcAqFarm = exports.auryUsdcAqFarm = exports.wstethUsdcAqFarm = exports.stsolUsdcAqFarm = exports.jetUsdcAqFarm = exports.prtUsdcAqFarm = exports.larixUsdcAqFarm = exports.ivnSolAqFarm = exports.btcMsolAqFarm = exports.msolWhethAqFarm = exports.msolUsdtAqFarm = exports.wagUsdcAqFarm = exports.mndeMsolAqFarm = exports.whethUsdcAqFarm = exports.whethSolAqFarm = void 0;
exports.orcaMsolDoubleDip = exports.msolUsdcDoubleDip = exports.kuroUsdcDoubleDip = exports.abrUsdcDoubleDip = exports.polisUsdcDoubleDip = exports.atlasUsdcDoubleDip = exports.ninjaSolDoubleDip = exports.slimUsdcDoubleDip = exports.whapiUsdcDoubleDip = exports.samoUsdcDoubleDip = exports.bopUsdcDoubleDip = exports.copeUsdcDoubleDip = exports.portUsdcDoubleDip = exports.slrsUsdcDoubleDip = exports.stepSolDoubleDip = exports.liqUsdcDoubleDip = exports.stsolWhethAqFarm = exports.btcStsolAqFarm = exports.stsolSolAqFarm = exports.takiSrlyAqFarm = exports.srlySolAqFarm = exports.zigUsdcAqFarm = exports.takiSrlyv2AqFarm = exports.hbbUsdhAqFarm = exports.novaUsdcAqFarm = exports.gmtUsdcAqFarm = exports.zbcUsdcAqFarm = exports.srlyv2SolAqFarm = exports.saoUsdcAqFarm = exports.puffSolAqFarm = exports.onesolSolAqFarm = exports.mmaUsdcAqFarm = exports.audioUsdcAqFarm = exports.seededUsdcAqFarm = exports.stsolUsdtAqFarm = exports.sbUsdcAqFarm = exports.hbbSolAqFarm = exports.hbbUsdcAqFarm = exports.btcOrcaAqFarm = exports.ftmUsdcAqFarm = exports.celoUsdcAqFarm = exports.cmfiUsdcAqFarm = exports.geneUsdcAqFarm = exports.orcaWhethAqFarm = exports.orcaUsdtAqFarm = exports.daosolUsdcAqFarm = exports.jsolUsdcAqFarm = exports.stsolWustAqFarm = exports.wlunaWustAqFarm = exports.msolWustAqFarm = void 0;
exports.srlySolDoubleDip = exports.zigUsdcDoubleDip = exports.hbbUsdhDoubleDip = exports.novaUsdcDoubleDip = exports.zbcUsdcDoubleDip = exports.srlyv2SolDoubleDip = exports.saoUsdcDoubleDip = exports.puffSolDoubleDip = exports.mmaUsdcDoubleDip = exports.onesolSolDoubleDip = exports.seededUsdcDoubleDip = exports.stsolUsdtDoubleDip = exports.ftmUsdcDoubleDip = exports.sbUsdcDoubleDip = exports.celoUsdcDoubleDip = exports.hbbUsdcDoubleDip = exports.cmfiUsdcDoubleDip = exports.stsolWustDoubleDip = exports.wlunaWustDoubleDip = exports.msolWustDoubleDip = exports.wustUsdcDoubleDip = exports.wustSolDoubleDip = exports.slcUsdcDoubleDip = exports.basisUsdcDoubleDip = exports.aartUsdcDoubleDip = exports.shdwSolDoubleDip = exports.shdwUsdcDoubleDip = exports.wagUsdcDoubleDip = exports.meanUsdcDoubleDip = exports.unqUsdcDoubleDip = exports.chicksUsdcDoubleDip = exports.wmpUsdcDoubleDip = exports.stsolUsdcDoubleDip = exports.fantUsdcDoubleDip = exports.upsUsdcDoubleDip = exports.sonarUsdcDoubleDip = exports.oogiUsdcDoubleDip = exports.catoUsdcDoubleDip = exports.sdogeUsdcDoubleDip = exports.woofUsdcDoubleDip = exports.gofxUsdcDoubleDip = exports.larixUsdcDoubleDip = exports.ivnSolDoubleDip = exports.btcMsolDoubleDip = exports.msolWhethDoubleDip = exports.msolUsdtDoubleDip = exports.mndeMsolDoubleDip = exports.sypUsdcDoubleDip = exports.stsolWstethDoubleDip = exports.msolSolDoubleDip = void 0;
exports.orcaFarmConfigs = exports.stsolWhethDoubleDip = exports.btcStsolDoubleDip = exports.stsolSolDoubleDip = exports.takiSrlyDoubleDip = void 0;
const web3_js_1 = require("@solana/web3.js");
/**
 * The following content is auto-generated.
 */
exports.solUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("85HrPbJtrN82aeB74WTwoFxcNgmf5aDNP2ENngbDpd5G"),
    farmTokenMint: new web3_js_1.PublicKey("FFdjrSvNALfdgxANNpt3x85WpeVMdQSH5SEP2poM8fcK"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("APDFRM3HMr8CAGXwKHiu2f5ePSpaiEJhaURwhsRrUUt9"),
    baseTokenDecimals: 6,
});
exports.solUsdtAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("4RRRJkscV2DmwJUxTQgRdYock75GfwYJn7LTxy9rGTmY"),
    farmTokenMint: new web3_js_1.PublicKey("71vZ7Jvu8fTyFzpX399dmoSovoz24rVbipLrRn2wBNzW"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("FZthQCuYHhcfiDma7QrX7buDHwrZEd7vL8SjS6LQa3Tx"),
    baseTokenDecimals: 6,
});
exports.ethSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("3ARgavt1NhqLmJWj3wAJy6XBarG6pJbEKRv1wzzRbbaN"),
    farmTokenMint: new web3_js_1.PublicKey("CGFTRh4jKLPbS9r4hZtbDfaRuC7qcA8rZpbLnVTzJBer"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("71FymgN2ZUf7VvVTLE8jYEnjP3jSK1Frp2XT1nHs8Hob"),
    baseTokenDecimals: 6,
});
exports.ethUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("FpezTR76RRjgpBb9HhR6ap8BgQfkHyNMQSqJDcoXpjAb"),
    farmTokenMint: new web3_js_1.PublicKey("HDP2AYFmvLz6sWpoSuNS62JjvW4HjMKp7doXucqpWN56"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("3e1W6Aqcbuk2DfHUwRiRcyzpyYRRjg6yhZZcyEARydUX"),
    baseTokenDecimals: 6,
});
exports.raySolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("B1aByG1fU5yUgQT2EtJrp96SC4tJhnipzdXqHx2CXRgj"),
    farmTokenMint: new web3_js_1.PublicKey("AUkn5f4N4TqPA5BiWirTDHWnG3SePfmeDpDqrFmhSgKb"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("5kimD5W6yJpHRHCyPtnEyDsQRdiiJKivu5AqN3si82Jc"),
    baseTokenDecimals: 6,
});
exports.ropeSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("NFsHa28zxuLnncHme7iLx8TXWV8ypyxPvEPaQh29zDh"),
    farmTokenMint: new web3_js_1.PublicKey("xpPyQwQ1HXHyEpvFGyTQRLY6rmj6jtAdEgLMV5uoz4m"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("ADrvfPBsRcJfGsN6Bs385zYddH52nuM5FA8UaAkX9o2V"),
    baseTokenDecimals: 6,
});
exports.stepSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("DGtiR1LWNPGV9A5byh7EjAoh9NC2cW5YDcmxCpUJosD2"),
    farmTokenMint: new web3_js_1.PublicKey("GwrBA1F8rGummDCDd8NY9Eu1cLNuJqbT8WaGxgWpFwGL"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("8nTzqDXHriG2CXKbybeuEh1EqDQMtrbYMFWcP7AkiDaP"),
    baseTokenDecimals: 6,
});
exports.srmSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("5CQVdPpaW95X1atyfqunZf7eBE1rhQBXMfgJ1wRVF1p1"),
    farmTokenMint: new web3_js_1.PublicKey("D659zwnbeTgquChbaWC3KDHrkYoqMuz1doGLHTFaqTtD"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("9tf8rBSEQYG7AqL896fN2nZi1iYPqpWaLEdpbeQaC1Vy"),
    baseTokenDecimals: 6,
});
exports.fttSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("GzZE2CEemaTGNtRSYs9sArAJFuJi5kwsPJFyz5puYhsj"),
    farmTokenMint: new web3_js_1.PublicKey("9r9BcPwCon96P5Y6JSdRAog7Uknz9p9GrnuHm4VzuB9k"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("EsYaDKJCmcJtJHFuJYwQZwqohvVMCrFzcg8yo3i328No"),
    baseTokenDecimals: 6,
});
exports.copeSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("CXYEVaqtDs4R9Bpeu4xYf8KT8S6nYis6VWkDEh9Mu1FS"),
    farmTokenMint: new web3_js_1.PublicKey("7CT19h7n2YBKiCFCaxXqMM79jNM4cmUvjXhNMjJNRYa"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("CzieDbGRdN1QGaGDNpSqzEA18bi881ccvkkGZi51pe1k"),
    baseTokenDecimals: 6,
});
exports.oxySolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("ALmRst1DksKXVfY64KXq7pJUUG5PP56kwY18ijVDGAsz"),
    farmTokenMint: new web3_js_1.PublicKey("G48RkwsNYd3A4rBfuQhCswr9YUE63fFmZGyhgH95dq3S"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("7tYCdLN84EnTMkxM7HNamWJx7F4xgKe9KiiWvLyWjbgT"),
    baseTokenDecimals: 6,
});
exports.btcSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("EeG5AAnFS56AveUrKexzivQeUdWhm6zYq6ubn6fLMCQa"),
    farmTokenMint: new web3_js_1.PublicKey("GxmjQZvgwNCh3QSRNB8CPED81hzySem62PDDuMp4B379"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("Acxs19v6eUMTEfdvkvWkRB4bwFCHm3XV9jABCy7c1mXe"),
    baseTokenDecimals: 6,
});
exports.merSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("8gJggSfM35JpfvXmEsuJoHPLB8D6br15MEWEWdQF81d6"),
    farmTokenMint: new web3_js_1.PublicKey("CrKVRnH6iGbFXxEnXMn3Emwv3Fe7VwxEqpA8zNbwsgkH"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("HiwRobjfHZ4zsPtqCC4oBS24pSmy4t8GGkXRbQj4yU6L"),
    baseTokenDecimals: 6,
});
exports.fidaSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("4X3U1tpfiwQ5zPPd4oQSPcu3UK7nMSUWyAXxCmcSA6qP"),
    farmTokenMint: new web3_js_1.PublicKey("4geGcEfgVjzJGZAaT8iTicPm1XLDPjdSpVhtA99sZ7jX"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("EYsNdtyu4gGTaGz8N5m5iQ3G1N6rDyMbR72B3CqbWW4W"),
    baseTokenDecimals: 6,
});
exports.mapsSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("Gobr8FhZDtn8yxkFQDND3KmfuwZseWEuyZ3xpD56hk1i"),
    farmTokenMint: new web3_js_1.PublicKey("7Dy84zJNHzEM9335BrtFjCuunt2VgxJ7KBT6PJarxKMq"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("99pfC8fWymXgbq3CvrExhx3UxQDC1fMWEWLbNT83F45e"),
    baseTokenDecimals: 6,
});
exports.usdcUsdtAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("5psKJrxWnPmoAbCxk3An2CGh7wHAX2cWddf5vZuYbbVw"),
    farmTokenMint: new web3_js_1.PublicKey("GjpXgKwn4VW4J2pZdS3dovM58hiXWLJtopTfqG83zY2f"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("H2uzgruPvonVpCRhwwdukcpXK8TG17swFNzYFr2rtPxy"),
    baseTokenDecimals: 6,
});
exports.orcaSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("F6pi7SyXWx56fP96mYQ4Yfh4yZ7oGNtDjwSYHT5Mz7Ld"),
    farmTokenMint: new web3_js_1.PublicKey("B5waaKnsmtqFawPspUwcuy1cRjAC7u2LrHSwxPSxK4sZ"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("2uVjAuRXavpM6h1scGQaxqb6HVaNRn6T2X7HHXTabz25"),
    baseTokenDecimals: 6,
});
exports.orcaUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("9S1BsxbDNQXQccjFamVEGgxiYQHTeudvhEYwFr4oWeaf"),
    farmTokenMint: new web3_js_1.PublicKey("Gc7W5U66iuHQcC1cQyeX9hxkPF2QUVJPTf1NWbW8fNrt"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("n8Mpu28RjeYD7oUX3LG1tPxzhRZh3YYLRSHcHRdS3Zx"),
    baseTokenDecimals: 6,
});
exports.kinSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("9XTA3t3X8KaeLgkv7XuDWATo4oDwBnTZqjnSPcE5ijYK"),
    farmTokenMint: new web3_js_1.PublicKey("7Ho3ht7krdFELBcPAsGXFfQMyG4PUvYSfpz4aNBRP3Ek"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("HEvnD66WcBfTajS9adUYnGRBMDehFtLySiFHSD6kEBWs"),
    baseTokenDecimals: 6,
});
exports.samoSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("98EDd1L47pdW1F7ne82dyFXpfhVdEcrCqUF1dYrWLFZi"),
    farmTokenMint: new web3_js_1.PublicKey("CNf8gZtLahBWxKe3YwsqywLHMTewGqvq6pJ5ecg3cTYU"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("D6N9j8F2DhtzDtrdpT74y3u2YmYAzcggiLc3nTjqux9M"),
    baseTokenDecimals: 6,
});
exports.liqUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("z2WHNcBJTZqbK5wCuBPtswEq2614T6si1cJmog7vFAL"),
    farmTokenMint: new web3_js_1.PublicKey("57vGdcMZLnbNr4TZ4hgrpGJZGR9vTPhu8L9bNKDrqxKT"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("3PD9SZFwXKkXr4akLf4ofo37ZUMycwML89R2P3qxcbZG"),
    baseTokenDecimals: 6,
});
exports.snyUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("7qAQZxYQRcmHmz3Wi9NeieujwLSd3N41M4dJipgSCNsB"),
    farmTokenMint: new web3_js_1.PublicKey("6Qw5Gzf1TkM3YRe7Dh6yMVMo2wnJxRiCUBP8abTTn9Yg"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("AZpo4BJHHRetF96v6SGinFZBMXM4yWMo4RA8C4PriDLk"),
    baseTokenDecimals: 6,
});
exports.msolUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("EvtMzreDMq1U8ytV5fEmfoWNfPhrjZ87za835GuRvZCc"),
    farmTokenMint: new web3_js_1.PublicKey("5r3vDsNTGXXb9cGQfqyNuYD2bjhRPymGJBfDmKosR9Ev"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("8PSfyiTVwPb6Rr2iZ8F3kNpbg65BCfJM9v8LfB916r44"),
    baseTokenDecimals: 6,
});
exports.slrsUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("5XXw91d4HoLjqhUhXmUcdkCDtQyXvQkLJMUcYJkKn5Dx"),
    farmTokenMint: new web3_js_1.PublicKey("66xCxkffQZKBZLiHV3PDcfR8ANJTfnDRxPCaBdv4wxB7"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("AtB4nUmdyQfuWWJ9xAHw9xyVnJFfSjSuVWkiYan8y86w"),
    baseTokenDecimals: 6,
});
exports.portUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("FpEb4eEqhnPBAA2RNvsSe8baCoev4mSNQjxKPGvx5Gjv"),
    farmTokenMint: new web3_js_1.PublicKey("4CGxvZdwiZgVMLXiTdJHTkJRUTpTSJCtmtCRbSkAxerE"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("F8gPSpwVHj8FdAJAYULDuZBxFEJut87hUbARYYx3471w"),
    baseTokenDecimals: 6,
});
exports.sbrUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("EWRNKpLc4Y8r6Aur9mo8GtBYatqfoML9A7bfEmDv2JD4"),
    farmTokenMint: new web3_js_1.PublicKey("Cum6sRPGpWYQHZapekDtMhbZ1BQ2QkYv9PAwQjypxMVo"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("CS7fA5n4c2D82dUoHrYzS3gAqgqaoVSfgsr18kitp2xo"),
    baseTokenDecimals: 6,
});
exports.scnsolUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("5MzBKRo6YqK1BKBz67sXd42jrb6gYzBuX6R5F8ywC33e"),
    farmTokenMint: new web3_js_1.PublicKey("7YFfqZGTxkj3Zeq3Et23kMznCaEYZ1WBZDt6CVrxwfqd"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("Dkr8B675PGnNwEr9vTKXznjjHke5454EQdz3iaSbparB"),
    baseTokenDecimals: 6,
});
exports.psolUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("4TK569s8SLgzmEcihyhrH8GmqFsufjZqsvHs8zrPEZYP"),
    farmTokenMint: new web3_js_1.PublicKey("8kWk6CuCAfaxhWQZvQva6qkB1DkWNHq9LRKKN6n9joUG"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("C2YzN6MymD5HM2kPaH7bzcbqciyjfmpqyVaR3KA5V6z1"),
    baseTokenDecimals: 6,
});
exports.msolSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("JADWjBW1Xs8WhW8kj3GTCRQn3LR4gwvbFTEMwv9ZNxQh"),
    farmTokenMint: new web3_js_1.PublicKey("3RTGL7gPF4V1ns1AeGFApT7cBEGVDfmJ77DqQi9AC6uG"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("29cdoMgu6MS2VXpcMo1sqRdWEzdUR9tjvoh8fcK8Z87R"),
    baseTokenDecimals: 6,
});
exports.orcaPaiAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("8KBSXu7zFjvbVXiwf1opGNBx3evVwdpaYHWR3uc4HVR6"),
    farmTokenMint: new web3_js_1.PublicKey("4aEi4A91hRbERJVDYxRWbbSrBrsxoM1Hm33KRoRzWMht"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("C7TH2jEJJaxVwwuvkbcDGfHUiZvEkkeYjyAcdTMi5ujb"),
    baseTokenDecimals: 6,
});
exports.orcaMsolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("7EVKT4iqfjiyzeVrafs23JrfhSoLd6XTanVuENNvisq7"),
    farmTokenMint: new web3_js_1.PublicKey("3Duk5b6fLztPmS4ryV48FM1Q9WXUSMwz9jehAT4UtqpE"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("CVapmQn7HaU1yMDW3q6oUV4hx6XoYv54T4zfGXkuJqkA"),
    baseTokenDecimals: 6,
});
exports.scnsolSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("5cE7V9D13k1P1qC23g5vcQEMsvrDzL5yFHhiesVUyn93"),
    farmTokenMint: new web3_js_1.PublicKey("CNqmEKGjZUUARVFHcz4w9CvX5pR8Ae2c6imHDNqsbxgj"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("APNpzQvR91v1THbsAyG3HHrUEwvexWYeNCFLQuVnxgMc"),
    baseTokenDecimals: 6,
});
exports.atlasUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("G92aJeZBCFiECwrKSQsbobfykh6cNLCf5Pd3zkiLjGLe"),
    farmTokenMint: new web3_js_1.PublicKey("HFmY1ggCsCky1zJ1sfdkNR4zb3u5n38YNRdf4vsGu17t"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("FZ8x1LCRSPDeHBDoAc3Gc6Y7ETCynuHEr5q5YWV7uRCJ"),
    baseTokenDecimals: 6,
});
exports.polisUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("5NkDw3LcFscf5WSxUiquNregP6RaP79Y2pN1jyUYepzc"),
    farmTokenMint: new web3_js_1.PublicKey("63JUKLnCAuNMPSPioEgbjjzp9Qk8qSEEM8eZqEtPqfLU"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("GteBdo9sqE7T41G8AJsaG9WHW48uXBwsLLznmu2TBdgy"),
    baseTokenDecimals: 6,
});
exports.bopUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("4bmbPAxrf5NLAD6A5fkhKhmuffdGScpu9Ei39YEaPATQ"),
    farmTokenMint: new web3_js_1.PublicKey("A7vvbqENJj8kED3ABjphe8TvwpasQYtoWGKpjpLArMxa"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("2gXDJZ7XAtQEtf4PRSQZKoq1WMuu1H44tQanbMA3YVpu"),
    baseTokenDecimals: 6,
});
exports.samoUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("F5ZvDWRVpQP5A17RFogL4dE7gZ2Uda7ZqKUy3DWJDEFx"),
    farmTokenMint: new web3_js_1.PublicKey("9voVuTq1S9bFZkF2Jo44HoVG63w2xDRT8eBzB23YbQud"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("6VK1ksrmYGMBWUUZfygGF8tHRGpNxQEWv8pfvzQHdyyc"),
    baseTokenDecimals: 6,
});
exports.ninjaSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("B8Q8e1hdCw6fhnSKmsMCrsDbvSNLZyzHjnPpwZKBkcSB"),
    farmTokenMint: new web3_js_1.PublicKey("7YyhptkxY81HPzFVfyCzA5UXxWdsNRD41ofLva3TuSpd"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("4X1oYoFWYtLebk51zuh889r1WFLe8Z9qWApj87hQMfML"),
    baseTokenDecimals: 6,
});
exports.slimUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("3muHEhQLmt7jmaicMHqe6LqXrrfFtqEuAfBog2fAXhRm"),
    farmTokenMint: new web3_js_1.PublicKey("3K7aZhtwWJ2JS6GnbbgeDVnxd1q2hwhqasmgRsAMZ4yC"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("BVWwyiHVHZQMPHsiW7dZH7bnBVKmbxdeEjWqVRciHCyo"),
    baseTokenDecimals: 6,
});
exports.whapiUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("r2MxTnFdnP7ECMxH7F4mcg1p3CowwhPmDmYoZ1syKZh"),
    farmTokenMint: new web3_js_1.PublicKey("Bfoi3RNnfdP5VeRGqvTA8MRN9ePGJoZgeKfe8WeBHUxE"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("ELfBngAgvLEHVBuJQhhE7AW6eqLX7id2sfrBngVNVAUW"),
    baseTokenDecimals: 6,
});
exports.copeUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("xigyMM8txxzpvjRw3Z1tX9euUshtqeHucW6K3f81KNQ"),
    farmTokenMint: new web3_js_1.PublicKey("9SDpBrfqNxjXcCzpKWM6yUKdfky975VJBD6xcu5cKf5s"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("HsauTv9s52Zv12eaDuSp6y7BEm4e4BHEyAsbdjyyWzPK"),
    baseTokenDecimals: 6,
});
exports.sunnyUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("Nw3caGfeqM8xQZ45iAf5zWmPMQux5wMWMjxsKEfnNk4"),
    farmTokenMint: new web3_js_1.PublicKey("9HPn1oREyNA7CEK7B1xwmBmVH6qtQaSfLBXc1JyRsdUE"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("GHuoeq9UnFBsBhMwH43eL3RWX5XVXbSRYJymmyMYpT7n"),
    baseTokenDecimals: 6,
});
exports.grapeUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("6eeK7PHkUXJ9t2qgdrHfbZfmU5S7zWZwpDw6i46Rf1jU"),
    farmTokenMint: new web3_js_1.PublicKey("97q89hnoKwqcynvwXcj83YqfqUBuCm4A8f2zHeV6bfZg"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("EorFh8siFyLF1QTZ7cCXQaPGqyo7eb4SAgKtRH8Jcxjd"),
    baseTokenDecimals: 6,
});
exports.abrUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("PyzV3qSzbj98ArVpASj2LQg6zCq6zfAixqUtYDafnRU"),
    farmTokenMint: new web3_js_1.PublicKey("5uR5STASUmoGVHzqMeut98t26TfVkQqWU9f9dsv3NfJ6"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("GMzPbaCuQmeMUm1opH3oSCgKUjVgJUW14myq99RVPGX5"),
    baseTokenDecimals: 6,
});
exports.kuroUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("6bHkoSxnK3aKW3PyG34dUz1naD6T3Pc7143s41nSuuzN"),
    farmTokenMint: new web3_js_1.PublicKey("6PGoaQdL9e463hdaFxHXsuPcjCHRK32CQ9PFKxvM7XY2"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("DRknxb4ZFxXUTG6UJ5HupNHG1SmvBSCPzsZ1o9gAhyBi"),
    baseTokenDecimals: 6,
});
exports.mediaUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("jCGam3ptTrFKhCwBGYUcqtrvt8pY4uWXaPoqC9E7uXB"),
    farmTokenMint: new web3_js_1.PublicKey("3pMYToENTB7jKrJiUPq19FCZCWE35Ph7bkRRMN6kxDXK"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("2toFgkQDoPrTJYGDEVoCasPXuL9uQnjvXJaDwa9LHyTx"),
    baseTokenDecimals: 6,
});
exports.tulipUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("AEemiZ28JTnz3xmj6LSfXruK1et317ZAGY5KCrKdXNCX"),
    farmTokenMint: new web3_js_1.PublicKey("2KYUwdRbVtaMUgHp1a6NuTomyCb33FxoZ4fkeVdwjaJA"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("4SBx8GXu8HhcVHWydQv1vsDdZs3G93XSL9CtMBny6hS5"),
    baseTokenDecimals: 6,
});
exports.mngoUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("2LM1Y428kPtLAsxcdxv2iKWaPxobt9poD2DAjeGpN6TC"),
    farmTokenMint: new web3_js_1.PublicKey("CxhcLZtbhfkwjAZ956SEkGxkAvMVQH3hfKTjKpgTV9Q5"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("H9yC7jDng974WwcU4kTGs7BKf7nBNswpdsP5bzbdXjib"),
    baseTokenDecimals: 6,
});
exports.stsolWstethAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("6eW2skHuzMxxPZjzE7x5fxtn3ZZ1Ak2SawrMH9T5KANp"),
    farmTokenMint: new web3_js_1.PublicKey("3kT3oYuS1rCfhmqfgy6EKcbZdaJimaVEjoy25QiuEaoj"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("Eswigpwm3xsipkTqahGi2PEJsJcULQBwZgxhQpr6yBEa"),
    baseTokenDecimals: 6,
});
exports.sypUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("7GDHcUdLhHxCHeQccpbJr3eNGzQvnoMdLBzkVNYvQgkh"),
    farmTokenMint: new web3_js_1.PublicKey("Ds4VGZhZzS2PMFzhzKeC3mwcQjdiCG21R76fTVbsSJyJ"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("qJxKN9BhxbYvRNbjfK2uAVWboto6sonj8XC1ZEW5XTB"),
    baseTokenDecimals: 6,
});
exports.stsolWldoAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("DMECkFUnVp1esox6Yyyfc7vJeN7spUHd5JypAWqEqCRC"),
    farmTokenMint: new web3_js_1.PublicKey("DQsbebdNDy8yQrwLTpieckhzi7Ewx9LoCPVf7G9KvY2U"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("74B9aMS7SA832xKngt5VLKmWAP3pa3qkUzWncTmQSsGF"),
    baseTokenDecimals: 6,
});
exports.whethSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("BerS3SE5G6FqZER7L7G3BhUJBrZ7BpizmuQRH9LMEYQw"),
    farmTokenMint: new web3_js_1.PublicKey("FkHQBBZGh5GS4GcXpcVksKYUUkLTNn6Yk1PCMxucR2AK"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("7aYnrdmdCRodDy2Czn6keUquUhjF1jPEmfwZPh488z8U"),
    baseTokenDecimals: 6,
});
exports.whethUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("GdQyNtN9rQWzpcm7mQMNBiXyeKRjjQoobh2waVQq5QyP"),
    farmTokenMint: new web3_js_1.PublicKey("B11Xp26xU2gzjToJEuGswvr6Jtidfh4GRUyCWzWMNdQZ"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("7NPtjjAP7vhp4t5NCLyY4DY5rurvyc8cgZ2a2rYabRia"),
    baseTokenDecimals: 6,
});
exports.mndeMsolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("5wr7m4YrJB38vuz3xyLqvq2DwLCcDkEDH5X97chKpH4T"),
    farmTokenMint: new web3_js_1.PublicKey("2wPsMuzhEsC6GhV3qtFpmJF6atEgLGbnmQ8U43Y6fPxZ"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("5PHS5w6hQwFNnLz1jJFe7TVTxSQ98cDYC3akmiAoFMXs"),
    baseTokenDecimals: 6,
});
exports.wagUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("4GWmnvMg7EGQZ6LeQtK6rbJrtTTVsQHj2ivwm4vm8mnR"),
    farmTokenMint: new web3_js_1.PublicKey("8Wu5sJpERA1J5iWcT8aMpt9cTAfKDLPbLpGjNsJoPgLc"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("Df6XNHMF3uRVZnz7LCEGiZVax6rXgz76owtVkBHEjSb6"),
    baseTokenDecimals: 6,
});
exports.msolUsdtAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("HULY26UFdfVkc2STTt1KREd57BwFV2md1tqdk253QyiK"),
    farmTokenMint: new web3_js_1.PublicKey("Afvh7TWfcT1E9eEEWJk17fPjnqk36hreTJJK5g3s4fm8"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("9cMWe4UYRPGAUUsTkjShJWVM7bk8DUBgxtwwH8asFJoV"),
    baseTokenDecimals: 6,
});
exports.msolWhethAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("D4Pmc82b9W1UDAqYNNNMGG7UYxBaZckf97AYExGbUK95"),
    farmTokenMint: new web3_js_1.PublicKey("58nifjPjF3CutGz2xMxvAMk7R9YgbVEc8Cstj4rCcs8j"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("5qoTq3qC4U7vFxo3iCzbXcaD1UEmDeCD63Dsuoct71oV"),
    baseTokenDecimals: 6,
});
exports.btcMsolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("GBrpFtiTabs14mc4Hi1RX9YiQY7res6JxrVfMTADfcQV"),
    farmTokenMint: new web3_js_1.PublicKey("DzpLz78wuwyFsQToin8iDv6YK6aBEymRqQq82swiFh7r"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("8nKJ4z9FSw6wrVZKASqBiS9DS1CiNsRnqwCCKVQjqdkB"),
    baseTokenDecimals: 6,
});
exports.ivnSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("FwzqbJZQiL3qzMx88r2o3CNKFxztuW3JC45YYk6ghMLR"),
    farmTokenMint: new web3_js_1.PublicKey("HqajzzbGMST3yCCVBJuXvNVsWkY2DXqiBz9cTRmmyBMy"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("DfgCnzaiTXfPkAH1C1Z441b5MzjjTCEh134ioxqRZxYf"),
    baseTokenDecimals: 6,
});
exports.larixUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("HeAQxAGBQdGURFFcLpdPagQ2vb66kXFTVxDfTjme8eGo"),
    farmTokenMint: new web3_js_1.PublicKey("DNAGfa7tK8csprRQmiDUwDaFfhw6ueHhVFHTCgTJ8HGs"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("8sfThep3io4gvcGeuoAg1Rs8GDwKJjtcdAFHqQSSNAVE"),
    baseTokenDecimals: 6,
});
exports.prtUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("9dCsoLdfkpYw1s7TMNE1HZPNmVSNDZrUkUHfxn2p8675"),
    farmTokenMint: new web3_js_1.PublicKey("2cYMt26745oFc7PadaQn8Vv3xFUxWBfbip2NyJeVG35F"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("6jCERp5hKj37PCXP3VTjCDJeoPuSpnMDMz5A6jWQv3yS"),
    baseTokenDecimals: 6,
});
exports.jetUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("BKdPQziyuKmXwZeVuo8Uj7usfUYpwSfnGBDbG96y266V"),
    farmTokenMint: new web3_js_1.PublicKey("4DjiLEKADWjYmiY9gzFnu5xews5oCTMRByWHWEzDa3bj"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("GBijunwxa4Ni3JmYC6q6zgaVhSUJU6hVX5qTyJDRpNTc"),
    baseTokenDecimals: 6,
});
exports.stsolUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("2P7FGV8XNXUkEAG6q5LbhfoBFkHJ7PDAjYqmAbwnVHBF"),
    farmTokenMint: new web3_js_1.PublicKey("3u2dNfGuU6C3vmSg5EvLPUpX57b3niqhWBV5Gc3WDEf5"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("GtQ1NT7R5aaTiST7K6ZWdMhwDdFxsSFvVFhBo8vyHGAq"),
    baseTokenDecimals: 6,
});
exports.wstethUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("2FAzkAgm8EpE7WpWUsEcNyj4kcVeCX2L8SR1BicGWEx9"),
    farmTokenMint: new web3_js_1.PublicKey("ojpWEdNYa5nGviUc8k8M2XLjHuaCL2EgHFdvTtdkXA1"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("5a6Y1ephcbKSoyLMQyD1JWbtqawCy8p2FtRL9v3zhaG5"),
    baseTokenDecimals: 6,
});
exports.auryUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("FvkyNXLizY4FCNkgbGpr1WA3z4A8iTRU4rPnnNJC3q2V"),
    farmTokenMint: new web3_js_1.PublicKey("7s7Veo1P8ZRy6z5MCvoAmg2kPFcQnq2Grt6yewWS8LbQ"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("6mJqqT5TMgveDvxzBt3hrjGkPV5VAj7tacxFCT3GebXh"),
    baseTokenDecimals: 6,
});
exports.avaxUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("GkmPgsKU9uYaZtKfzjWeJjBSvX8mkK8gNqoWMYBdcvUx"),
    farmTokenMint: new web3_js_1.PublicKey("5mZydNG1fBPTXzgp9SoS7Jny6EmSzBPTPrFbB1ttGnvx"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("Hmfrtmo93DpSDmVNLQKcBS5D1ia5JatiRSok9ososubz"),
    baseTokenDecimals: 6,
});
exports.fttUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("2FtSDVRmYq4x6NeTidpZVuLzgN2APzr3zHxpHx8zdgbr"),
    farmTokenMint: new web3_js_1.PublicKey("2AAzmhZ9Kh9mFcQtHJVTafu69tc5GCGpi6CTgafYta1S"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("FwCombynV2fTVizxPCNA2oZKoWXLZgdJThjE4Xv9sjxc"),
    baseTokenDecimals: 6,
});
exports.rayUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("GCcVYpBCPcPjb11LeeoVhLNJQzS5GXEEd6Dmo5eejBTx"),
    farmTokenMint: new web3_js_1.PublicKey("9MaXcCERB4DzedPNSBwyHM2P6Yo6jCFLjnatuWCtR3WF"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("4cXw2MYj94TFBXLL73fEpMCr8DPrW68JvrV8mzWgktbD"),
    baseTokenDecimals: 6,
});
exports.slndUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("5NrUSiG134QLY4JkmX82n4hdZQnmqsWMrBapQwfAtpeu"),
    farmTokenMint: new web3_js_1.PublicKey("GdySZb2nbeEjCLBg65veC5kzfMCfCWgtgqwH9YWDtDXr"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("F59gkD7NnsdJbFKrRZsiBC8PAooN4c56T8QmahfW1iXN"),
    baseTokenDecimals: 6,
});
exports.gofxUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("GmyMx4JtRrkjmxNn4WQCmVNuLfwRjkAmCfMv2AXag5se"),
    farmTokenMint: new web3_js_1.PublicKey("B95rdqSY4dqPwmt295XwBZZqZJYLmqDNXU6NvBpT4ep4"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("7vnps4VE5RTGAr5fmPZu7fSrk2VnM4Up838grZfqmxqE"),
    baseTokenDecimals: 6,
});
exports.woofUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("5K7WdUTG2yZjULeNUUiH3eoWhY9YYAnpoWFGiwVj3q2V"),
    farmTokenMint: new web3_js_1.PublicKey("4HaQXDz9gdLgKUjvNVtnLyNZoWNYKjh3XxH1TpLgiwmi"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("9EjcYfHcG8f1mccpHyaAwpoxaUPiheC6KgLQjyD9aTb6"),
    baseTokenDecimals: 6,
});
exports.sdogeUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("672wPyV8sJdqBfyay6MccLxsCj1sqQyd5WaPXkTbeB8N"),
    farmTokenMint: new web3_js_1.PublicKey("4kXYyZAMBdvgDaBKUGvRWJnHw9af7sCUPvpQ68PEdP8b"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("CHTKUJGYRtBDqnxCFjxe5KEkZgxV98udbhuYYyzGxup5"),
    baseTokenDecimals: 6,
});
exports.catoUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("4EfikjxutL59CXMBYffhSXdPEw28T963idHiwz2qpodQ"),
    farmTokenMint: new web3_js_1.PublicKey("BHtZnTBMeY4EBEW5egGnuK5bdW12v6Dod6wFav79AyYx"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("55r9txzQtmjTykmTXmBYZCVMg5z9squB8b5cSw2AhxA4"),
    baseTokenDecimals: 6,
});
exports.oogiUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("CzWsHjE7azpbksPhHxdkvH5gfpBAx4FqohwFnkHgfVpr"),
    farmTokenMint: new web3_js_1.PublicKey("FiwSk36yi1DNWcuQUeNipAc1VKxa9Wv9AR2xFvyKUxAE"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("DSiHyHDn96bUQSZtizyCRLcQzrwohZeMpVu8rYJN1HzG"),
    baseTokenDecimals: 6,
});
exports.sonarUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("2SxKGe1PEKWQC6m5Q1jUnXJuZBVXyUC37ffCPop38iiA"),
    farmTokenMint: new web3_js_1.PublicKey("GWmwwMGYBG4NqYdnsYrudzBnbgDC49MkBxdzhfLA9kVY"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("5MvQHx8eftU39JTucFsT315JFnQASuDQg3FqxTw7xcvN"),
    baseTokenDecimals: 6,
});
exports.aptUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("3YZ5GYL625vWibn7d8hMdrMBawy9HGUyeTe4AoXoME1Q"),
    farmTokenMint: new web3_js_1.PublicKey("Dx7DYSuaBufhXyQG7155ePkLmHyn6w7WeKKtQB9zscZV"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("HNrYngS1eoqkjWro9D3Y5Z9sWBDzPNK2tX4rfV2Up177"),
    baseTokenDecimals: 6,
});
exports.dflUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("GaKnyw7sYU6raJpZGjFER1eXZ9zHfqih5fAj1T5xxP6J"),
    farmTokenMint: new web3_js_1.PublicKey("C8jJeUM9s7R7gTw1ybgTW4LPkgrYtogstD6MtBjroGLB"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("AWrtTWG4Zgxw8D92bb3L3sQtGLD3zDztMPWsXSph8iBP"),
    baseTokenDecimals: 6,
});
exports.dflSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("HTEe4r3YDRqDbA2dwQNMnMBAToV7rHvpyMBe6EjZmXyz"),
    farmTokenMint: new web3_js_1.PublicKey("E3omuGbZSksAdbiyvtpdfGsUGhptA26Enwe5Y5jJKnbK"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("9Y1vPaAsMz8X65DebMMnmBjbMo8i4jh4mcgiggZUUS3M"),
    baseTokenDecimals: 6,
});
exports.frktUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("EUgwcVQhm1BDHWeMThNRx143RRay1byP6gmca9dp8HMe"),
    farmTokenMint: new web3_js_1.PublicKey("Bc4CvPm15iLRopgLmU17RJicBgjykiMwVzYxa7Wnujdh"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("FnDxJPNk7pPmGHUbR4XUHmHevrkXHdna5D3sQKcAtjBL"),
    baseTokenDecimals: 6,
});
exports.tttUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("BPsQWbFEwYRxHims8Yw14BbL8njWEQq4tNWrrJb9sgXD"),
    farmTokenMint: new web3_js_1.PublicKey("AmrPpJFQj4xtv3yVgPCFqhXEMBPhaKRP6qFyhuCdz66c"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("FGgP1npQTsC5Q4xBmQtNYSh51NKqNwdxBZy8JCo3igcu"),
    baseTokenDecimals: 6,
});
exports.upsUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("4TQzMpNP5JdZ3YDi1iz9kYqv4L7sTKQeFtpvnfc6T7aJ"),
    farmTokenMint: new web3_js_1.PublicKey("9fuv3emLQXECrTWqm2HaKT3wQhmvmgqmReZSMcu8PfpH"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("E1U63VXhNiWoUkVvjrfLDdV1oJrwE6zLde3bohr6jCqz"),
    baseTokenDecimals: 6,
});
exports.fantUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("BkiSfYmmt8KxrKFhpS4oq6TYf5C4DFnaFWDnn3eDhcfp"),
    farmTokenMint: new web3_js_1.PublicKey("C2EcthTMaC5eATXVaXg5ctvMfUYYgFyNibybHes5D3S6"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("GjG7JjTQfQpDxw4hWx4etP9oTaYCuCbPjsU8WaUT3xHB"),
    baseTokenDecimals: 6,
});
exports.blockUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("3QHfuWvAkUfr1QenKaPBDUUUwShvoMLZCmfux5bovz1J"),
    farmTokenMint: new web3_js_1.PublicKey("CkSfxpcpunHLUQktiFAeTJCYHsT4brYoJE23R3vw5zCF"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("D8WjqtwC9CzBrQKfSf2ccCHFQuPYwyLv5KAy8WjT5vnf"),
    baseTokenDecimals: 6,
});
exports.runUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("2yCL6hAsNQN7R2wrPaUffxVDyK8hhSL1hYf2knhUzLZk"),
    farmTokenMint: new web3_js_1.PublicKey("Gxh7KubJPEqR1i4oHCxxYayWcSJMzn7h2kpJeCCEAkNC"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("34Ppq6R8NfYBwWwPY4cBK4Afyb8hHaASQFukCzH6cV4n"),
    baseTokenDecimals: 6,
});
exports.uxpUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("G7W1y2Mv2kxu7ht3csCUNdduaUNBkYVcvX6QA47CJKuF"),
    farmTokenMint: new web3_js_1.PublicKey("8BxKJ3FAFbidm9rPotQEcm6dDqJkTz2PhLtaq6hnZk8L"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("HjR8JgqNKQVMvdryqJw5RJ4PCE9WGk8sgbEF7S9S3obv"),
    baseTokenDecimals: 6,
});
exports.btcUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("3japFRAWZTq8FLjeTsi4hmNHpeo7RhP9ift4joCjRkga"),
    farmTokenMint: new web3_js_1.PublicKey("3aDHNJpWPEzXHZNeuak3LgcB8MbAwnDKkB2YftaPyhev"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("J3kvcay3N16FBdawgnqoJ9v9p6XCvyCLE2Z9F5RLvGkj"),
    baseTokenDecimals: 6,
});
exports.mndeUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("FBSHQPtDZLvv8tmj1S75GW4TKGhxXVqU1zcSK57ZWHGF"),
    farmTokenMint: new web3_js_1.PublicKey("BtXnDsnidDZnkVFJJnqtAMMCtkoG5uioYF72yC8fhVTn"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("12Uj74zgUUoBe4yeackwQ4qYtFMr9fk1xL6q5Nha6t2N"),
    baseTokenDecimals: 6,
});
exports.chicksUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("BQivr16DZwUyzx9Qa5LR7V7eg1ASMiPjBtyj1bsoLXw2"),
    farmTokenMint: new web3_js_1.PublicKey("GTcxXu3yw7yotq2UtAMvpEbx84ozJyrBwy77Cmem7m4X"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("71CBZeJ4tw38L9pSPoCz4fRsuWE64Fipyzotte7haoCS"),
    baseTokenDecimals: 6,
});
exports.onesolUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("5FSwvhovxgEJPg4yufxCXiLXD3RY2SJKQ2fFy8mMupJj"),
    farmTokenMint: new web3_js_1.PublicKey("4tcadvVpXQPswr5dRt6jJW1sVFW2CxrKgzP594ZxaUTv"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("6MF5CHWAj5mS7FhpxiKz37CzR2eYTu236XpBKKMXCrGg"),
    baseTokenDecimals: 6,
});
exports.wmpUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("8brkHbXG9fNoYZu68D7cQy9Q1gC7hfTZQD7jfKFv4hvC"),
    farmTokenMint: new web3_js_1.PublicKey("Ce3VXSQGFEHqrQSEzcVb2Ro88Mcn56cYBUAXFx5tL9bo"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("HDgxKmiA8Pv82fNguhVeMkZqQkos2YksFPoP1KttWxX8"),
    baseTokenDecimals: 6,
});
exports.unqUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("9hQXfKCdD4v9v8vzimd6KExoSnS36q5kUPTW3PdX6KRu"),
    farmTokenMint: new web3_js_1.PublicKey("CsGTrgJ6oLx9UQow9aLjuAacjXTY53zT8B2FnAbfVZAS"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("2VuGzaMrDnDyZfYvDwSXk38s7M2wpud7LDY3dGA1J9sy"),
    baseTokenDecimals: 6,
});
exports.basisUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("A8CNiARq7zYMMGKYbqJVfByVyBzdMexhc5EEGzCN13dS"),
    farmTokenMint: new web3_js_1.PublicKey("4yx2aHMa7N4m1uUaBRy9QPtpstw3HFPtvcCPJQaGFHKL"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("GoaAiajubRgeCFEz9L6mLnSmT2QFegoJDH5tpLfivpj"),
    baseTokenDecimals: 6,
});
exports.gstUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("44aNS8nnj4r3WWvnR1Ud929iAqn5jJ7ugVkJJvH4XuW9"),
    farmTokenMint: new web3_js_1.PublicKey("72vxFxfeSN2DRKmSQAkJCoFBNYb2WNevyaDh4v2t8TqP"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("E6FUnQHGHJVJg7oExVr5Moeaj1QpdpZQF5odYjHXWPZb"),
    baseTokenDecimals: 6,
});
exports.meanUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("4ZqD3uYerAhUrVCPSKV15JT7ToR4EW1TwK6HHckAZNio"),
    farmTokenMint: new web3_js_1.PublicKey("3h2VBX8533NB8eEH8rPXMdayodFDbgTHpbav6JqLZAQq"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("F5BTnwuMA6rxftTdbZ33VWKr2wrr6DuQHnd4guKmPSYQ"),
    baseTokenDecimals: 6,
});
exports.aartUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("6pvWTK3UZJ2gKpVZBXMEUEbnSbqf8GxEFbAQjWuuW4C7"),
    farmTokenMint: new web3_js_1.PublicKey("Bg7pZq7KdsQsnCQBeNpa4XVnrWfjUJNu3ViHoNB7YUZU"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("HCtyJzFUtYecXrA52s4Y9atq4J1fhT3cYsTX17XVSFag"),
    baseTokenDecimals: 6,
});
exports.shdwUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("ABmFqgfvQjjU8uBkZL2KdH5AvYEMsuddtdvpm4s62Pzq"),
    farmTokenMint: new web3_js_1.PublicKey("7WWHfufv8vuBC1x7GXA3pu7kgNhEQkXoq3CtbaQihAJ9"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("DJqqvzSuPaWThfzwMjXx7H2ZmHDdwxza6NtFudtuXcpc"),
    baseTokenDecimals: 6,
});
exports.shdwSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("EKe5CgBnBJA2cgeEUe67aQS57bAQsPvfVdArEWuEuEEW"),
    farmTokenMint: new web3_js_1.PublicKey("BDStVBt4NS5bfda25ubK51kVRioV4yjKKCPbe96jeEms"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("2ws7g3LBPdctfKn42Di9qxzQtUJ8ZL1aEAX2rGEQMNqh"),
    baseTokenDecimals: 6,
});
exports.scyUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("GDf3GpSU1V1UCvpcU19rV8p8CUC4cken9b5RpKzBBgiN"),
    farmTokenMint: new web3_js_1.PublicKey("DkxeyV1TEt9umvqjPNf9tdjJjgyvW5Mdi4AQnExA5bZa"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("99ZHUQsgxL7K6PHrGNi1gSwawwPr7UA5fbWrYoHQ6qhX"),
    baseTokenDecimals: 6,
});
exports.slcUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("AgscvT8HJ4uqznp4SeYMgKf9qUjTrQL7AzU6mjcqpPCB"),
    farmTokenMint: new web3_js_1.PublicKey("X8GnAvxq942xXjNzqYPFQQ9JstJQNEA81uPeNoQpgrN"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("E5kSBqTDxFLbLNQaVVtPtnhEYVLMCK2fVSEKoMKL98qR"),
    baseTokenDecimals: 6,
});
exports.wustSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("7tHjwHKvoSPjoAZepTZsw7uV9XobkgRXhMbhbZmqA8mS"),
    farmTokenMint: new web3_js_1.PublicKey("F49Cm3srGucQCBanA2xL7nSKHfH1QZd9vLdFtkg4LKnq"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("6c13xsmyk7UaHUWZ2rm1MM3ZdrQRSBkQ9waaG25ridVs"),
    baseTokenDecimals: 6,
});
exports.wustUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("Aq289TqufDB4A4TEGVUkVquVggyrPjCwxkEv7pDEVpWZ"),
    farmTokenMint: new web3_js_1.PublicKey("2mhVUMsG7eb3XhHjAbKpRZgWNyGZNCiWU7dRxauzZaGL"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("J1KfRtP5y2warpD7LdJhfBLPKoWwSqYuovdArSv1mpQ7"),
    baseTokenDecimals: 6,
});
exports.msolWustAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("9iXAweywwdGR76L58FNg9siQzxyzbeJsomjEohPztpHF"),
    farmTokenMint: new web3_js_1.PublicKey("6wqME6zPQzGDLugpnwZnVBKBbATC5nTaEa781Vj98yvM"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("68YVjgPnTUPcBqZyghqvD2WPNsrLKsjYTmBKJzHRr4qd"),
    baseTokenDecimals: 6,
});
exports.wlunaWustAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("52VxPD3pag6QrtEKXTe4jYBvSTvfggjbXZv3vqo455uU"),
    farmTokenMint: new web3_js_1.PublicKey("3ZGKT28NXAqb2YtUAMvKixQvHNHT31Q5mK8AC2iBMs29"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("8Mh7drLbt3jFJYwp948XyvQscGLaLkChNcaH5wwaAoWA"),
    baseTokenDecimals: 6,
});
exports.stsolWustAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("GxhewC22S6wsXT156yC9SARvDnijoc3YEYyLVcQFDUCx"),
    farmTokenMint: new web3_js_1.PublicKey("3Q44iV4URXdbS4Tk1PGs5VdWQoCxHB7zdcdMnemo8jfH"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("HTZd53fYwYQRyAjiaPsZy9Gf41gobFdqkF4oKe3XLi95"),
    baseTokenDecimals: 6,
});
exports.jsolUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("62JumLDtnsQ5YsEZrJaQsvDMP5pRmok3JEWHghFkL8D2"),
    farmTokenMint: new web3_js_1.PublicKey("B3JDURP58ooQT8zvhLgDRKVjstqgrHhn61qzzudHWrS1"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("AzEoVuNJyo9ByoLRZ5t6vav2Zg24vULNVJM41PgCKUqR"),
    baseTokenDecimals: 6,
});
exports.daosolUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("EaxYr2iZhrgwUKQVeFFeHQLW1afzS8pgZGn6hLhgFMgy"),
    farmTokenMint: new web3_js_1.PublicKey("A9kPybFFdj3MGKWWSwgb3KvzQyeNXESayh4ngmkmee2R"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("CCyDxjdW3G7hPTthTMPTZ4bnhFF19XG6rx2fNiKeRQww"),
    baseTokenDecimals: 6,
});
exports.orcaUsdtAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("4SNMBiDGjDSZnyQFqSodJTTmRqEkisBWciXZ4hGbDujc"),
    farmTokenMint: new web3_js_1.PublicKey("C2YSdSesufbMVWtKXSZGhc12f6UL2j3WDe9VT6rRbmA8"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("Gx4PoxenyQwhGGnKagAT35iVg4im1iKhJxDWqVhgu6tk"),
    baseTokenDecimals: 6,
});
exports.orcaWhethAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("AgWH92f8cPAuURyqehF6A9wFgupEsnDX3VP5eHyytabA"),
    farmTokenMint: new web3_js_1.PublicKey("A7EAHvFEWzW8qSTFNLrEHN894GAZWZesNiebttEwtUGJ"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("GsfyYHkSgC3Ta6aWR9MjB2sxoBrkGGeR2tAwXbpphf3"),
    baseTokenDecimals: 6,
});
exports.geneUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("EoRaCH4cwmmxc78s5E1fTDt2zPkivXNLf1MvoGhJywKd"),
    farmTokenMint: new web3_js_1.PublicKey("GrBZ4HLhL28JTxMfAH2Vm2hVv3RE4TYXCG9p9BpyUaRq"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("7cuu94swKL5PtFQohKMAzyd1mjj65rgMW3GzLY31HCnK"),
    baseTokenDecimals: 6,
});
exports.cmfiUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("2CGcou8YJoannsazSYQ16jSUZzPP8zZ9jdqgSRSGsVXa"),
    farmTokenMint: new web3_js_1.PublicKey("E8RVjS24pBuF3oCCeJVAgC4RQ7mVa5P3FGXqGbiczzvD"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("85krvT9DxdYgoFLQDHTAGdvtNuLdAsc4xE5FkVLpN2aR"),
    baseTokenDecimals: 6,
});
exports.celoUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("EowPc9YUmqJDjAujhD1eaLzDm99JnCu88fA1W8skKgQv"),
    farmTokenMint: new web3_js_1.PublicKey("9RRBDWZGWGVnHgazCpb9R1XApmHLBDya8Tq1yrzoMsKY"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("HVLyX8mD8YvKgZJ4oB6rXJiCYMLpHKwB6iCiCjE1XwdT"),
    baseTokenDecimals: 6,
});
exports.ftmUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("EsE6fq5rP5Dg3ivS42JodpaQNpmkB6LTFX2nDSRZZ2og"),
    farmTokenMint: new web3_js_1.PublicKey("E7Af9Fa2U1YqEkQpAYjXwDB5TJUou9VcN3ot33Gj6UY9"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("Gpzd833qSmv3kXpQmxEaqkrZTXZaRjhNAoqhf61qAhTG"),
    baseTokenDecimals: 6,
});
exports.btcOrcaAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("8TcCqDgPgim1CV6zgAXFq3PxSBH7MDCeVSGharRzJ4SA"),
    farmTokenMint: new web3_js_1.PublicKey("EQsfnq9d1R2MbXGKDPSgJXnDk1oXocgMFoSd15pEAPGD"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("DFpLFcQZqDKykyDePgip4r6MExVmBKWqTa12ezq6qxUY"),
    baseTokenDecimals: 6,
});
exports.hbbUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("3saySWJwipyCXqqhYT8XaSMFAS7NNksKKdP3zFJuUMtT"),
    farmTokenMint: new web3_js_1.PublicKey("BxPd4x7gm6WQF6jsDCRQuQVZ2bopQ4KegMrqyQgdkPDk"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("cL5WhffCYFRLM4We8VS2W684kM4pHyuvEDwp8Ddw48k"),
    baseTokenDecimals: 6,
});
exports.hbbSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("9fEfGhHjvqDPcFx6cBNuWHHjQaa6fSEP5pRHjdFkPenE"),
    farmTokenMint: new web3_js_1.PublicKey("62q8m79WkXR1MAer4H2zLpA38s47Fywkqqv4MMwEHjad"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("FkKzu2HeMJZf4oHwoYPxLGVy3net5Jq8HAfnA5VqETgk"),
    baseTokenDecimals: 6,
});
exports.sbUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("Ax8vkRHj3Gqo5Tv6PfWZzmbr8zHfPdeYbJpG7QYkP3r1"),
    farmTokenMint: new web3_js_1.PublicKey("3Wppx86xN7Dg7GLUTD9C7AqCt68qZkWzNYUdCS7t1pCk"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("2Reqt4Sw9xNY8BoJ3EZLpFu5yVgNxFrbw8M3KiJpPn6o"),
    baseTokenDecimals: 6,
});
exports.stsolUsdtAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("2xvT1Z2SU1iKvmWpUNP9y4vRXyB8GpGq7sD9CforJoYd"),
    farmTokenMint: new web3_js_1.PublicKey("CtvKaRLzCzRZCcYPwai7NCYBK4ArBj2oD6BfvMgJoiCN"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("4ni1nho89cDKAQ9ddbNQA9ieLYpzvJVmJpuogu5Ct5ur"),
    baseTokenDecimals: 6,
});
exports.seededUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("n7VuBP5qpMt2fZxNevR2gtxsgogDv22GJsj5f2WNJCN"),
    farmTokenMint: new web3_js_1.PublicKey("H15WptGntFQifZmJHUzYBV9Mv7P27ofavEsF6yqpLTdX"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("H7gyTmNCDXkD8MGMqnxqoD8ANszjcju4tjT6ERZ5dakf"),
    baseTokenDecimals: 6,
});
exports.audioUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("DvzYNQoPYFWQCjBs6rxvMmarcKMiEaQHC7sWVAm5K2XM"),
    farmTokenMint: new web3_js_1.PublicKey("BbRVh76k2jYsSWnd7qpDQ8ptzVzgRq67viK6tAFEkB8z"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("3hksYA17VxgiKSeihjnZkBbjc2CTbEBfvDCYgQhojTo5"),
    baseTokenDecimals: 6,
});
exports.mmaUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("C54TkajKsLqHZP1taiayDAaAHrr6KBmCVRT1BuZHxa9q"),
    farmTokenMint: new web3_js_1.PublicKey("H9qkbU2XYSQTk6JJJ2TMuZBthfkeCHigmuxB5jiQaamn"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("AaZRnJAnDyJyPD9uPJpJ8bzBGDCEi6jtBpUf92xErWPp"),
    baseTokenDecimals: 6,
});
exports.onesolSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("64zLHH6zA8yuLxxPAgbVx5Cp9F8afHZWSyUndHsFBP95"),
    farmTokenMint: new web3_js_1.PublicKey("Hva9oLa2GjoKdB45WoHujsX7MTpehByPshMrQpNDmkFq"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("9wPhuYapychVDSxmXqCZxy2Ka8Lmav4SHM72si8bfraV"),
    baseTokenDecimals: 6,
});
exports.puffSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("Gp4r2PFLSf9Y1Y1K46gpEYsM7mkFNVY4Ga7s6YXZmxGD"),
    farmTokenMint: new web3_js_1.PublicKey("Ge5kuYg5PekrPUeKzngw97Cnfngj8j6NCX5q1jTBkWSW"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("Eho8h1BcoG5QWU7X9FzJafw5ErKUXtR2LobAJJZfWff4"),
    baseTokenDecimals: 6,
});
exports.saoUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("4rKjrmHAmeT6bu3JNhP2NYjFgVHtt71U9yumBDKHTe14"),
    farmTokenMint: new web3_js_1.PublicKey("A9BeGSRJJYXPrMs81rVZxvkk16fopzgG5YkngntgTu7p"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("4iyU77yZbg8iD344vbwruAuDAf9i1EVV3FhZJDnWStBE"),
    baseTokenDecimals: 6,
});
exports.srlyv2SolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("9SV2NL59i1PfD72AUHDXa2KT1xWnFD3VyMgfzCRDARW1"),
    farmTokenMint: new web3_js_1.PublicKey("6gkZ7QUmxmwPLS2NK3Dr6YHtTPZs6GQrkA595WSx5iLe"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("3dXdXg5HPyZ73GFC9LkSn3thdJUGeXWB8iSTHs5UcqiH"),
    baseTokenDecimals: 6,
});
exports.zbcUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("gBBNuRf5VsGoAEwKzTAaVsPSqTiYLh5VmaxvzvbCEBb"),
    farmTokenMint: new web3_js_1.PublicKey("7Hoi4adCSBzERdvSiUXtVDz79tiDxMD5HpDv7m9rs3Sb"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("2LYgm6nGXmSSjfoEriPuYeGoNiWNxUs7n3rnTbDWN5c7"),
    baseTokenDecimals: 6,
});
exports.gmtUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("DbJT9UTTt8U8xAk7BtRWC3nQHxJnNmZdZydUAhJj14TZ"),
    farmTokenMint: new web3_js_1.PublicKey("8EnEoVX1aXkzbTzhrqDQ2aVGybbPpeWZDCYEGjjw1dyG"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("CFxQF5kNAtbbDj298Xr47Sf4mkSyuzWpRH97hrdQ6kxi"),
    baseTokenDecimals: 6,
});
exports.novaUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("B8Kv2mjJGvbDnEUX5EHhg1Erg3s7qBYpjDzrmvMMMrRx"),
    farmTokenMint: new web3_js_1.PublicKey("H2jZrZKx1fLY9LPi2ad5G2DVBWzMmUwPQLjRFERUroEH"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("DdAFNvDxtEHCgKj3JAM64zCKfMWhs4J9wEmRrjUAFiME"),
    baseTokenDecimals: 6,
});
exports.hbbUsdhAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("cPFH2eNZnyf85RpSkPM2ZG3S4RdedX51ywqSk7kjmp2"),
    farmTokenMint: new web3_js_1.PublicKey("DUGuXcpHX9y6WdTKgDm8XsECbf7pApA7ZvAsjUqHUoYn"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("7n2YW9qLkhGFArdZPLoF4hYPE2zw7xCACkVPXrUWnLuo"),
    baseTokenDecimals: 6,
});
exports.takiSrlyv2AqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("ADTFK1eFV3oKCzGex2wHTCctBNYxqSvHv7fN1tnbHBGq"),
    farmTokenMint: new web3_js_1.PublicKey("EkPqSfr4ZiRiLgjZw45QmF36TzZ3EEV51DyRVd2iuhA2"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("6atKbS2Xz5vu7cqWBNk8KYkuakRzckZ9nvtUKf2k8Sc3"),
    baseTokenDecimals: 6,
});
exports.zigUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("CAG4eyg41phvb4KZ1hkfPikbUreZcVTVVH9Rc8yf7ADc"),
    farmTokenMint: new web3_js_1.PublicKey("9nKbvVW9TjdnKt2NbfMaf9PHcg8zWWFgGTBfzUXdFcke"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("5vhh9ZnD9vnahRhFLP1EUEyYRSzvJwgw9U2xygsSJSrp"),
    baseTokenDecimals: 6,
});
exports.srlySolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("3851tQsk6PdqWm3pZgPdHaqZcBdNioKxnGM3ah8Yq5rK"),
    farmTokenMint: new web3_js_1.PublicKey("HfkZkoPqZzd7S1F3BhkJN2Es1C5vabVYqfCecXG2MGHP"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("HfRgvhgscGX5GaP3rUrZAhh7gS4aJ2UQ7rNVX976rG6P"),
    baseTokenDecimals: 6,
});
exports.takiSrlyAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("FSy7cpNu9gC2Grr8J3pnorm5jTR9D2xHngUS8vZDUeug"),
    farmTokenMint: new web3_js_1.PublicKey("13HwDUH7yREGAwHSWnx4Kgnn6huL8yfCCBqaBr6NGVH6"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("3xaK5aWWLNRB73xUVX3cusLhDp65mTvP4fwW5Jwxakgs"),
    baseTokenDecimals: 6,
});
exports.stsolSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("9bKvCTiVrnSzFPrd4g7Dd9SeEZvKNn3SQ7dMpuE9VKuo"),
    farmTokenMint: new web3_js_1.PublicKey("7ZvhrqdnP2NrQi2YCjQNd8jKJpLDMUuZg4wP7XJPjnVX"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("4jjQSgFx33DUb1a7pgPsi3FbtZXDQ94b6QywjNK3NtZw"),
    baseTokenDecimals: 6,
});
exports.btcStsolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("F2mdxT26uvf16X9vV5ACUqa6FicPBq5fLZkAXXxYxD8b"),
    farmTokenMint: new web3_js_1.PublicKey("GjBW6751wNJ99zYkr9QwSi98bvj1NgyAcYHfQDpemJfq"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("HYp9v7cY4wAxSsa6ijztQQ3GQ8iTttuG5vu8JNBDHoNh"),
    baseTokenDecimals: 6,
});
exports.stsolWhethAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("5xpcR2aPqE3WWQa6ZMv6xSmBakdyKBsAi3pmLJj2ARH1"),
    farmTokenMint: new web3_js_1.PublicKey("9V9CavjccKJH66m4MJHzVBYzY2ZM3U3D5xUCP7HSyAUz"),
    rewardTokenMint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("FWurFWADrgmhb6Y44LEaCMnEHS2Tu3QGqd9oBcZtr8gT"),
    baseTokenDecimals: 6,
});
exports.liqUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("AraZDjfmkqzDJ3CdbjjYbtpujUYivgTEKKM2TPf6hJ27"),
    farmTokenMint: new web3_js_1.PublicKey("5rGtJDiJhD5Mx2fvdEYuLrCiWaMD9z3wpmJSxwGHmo4u"),
    rewardTokenMint: new web3_js_1.PublicKey("4wjPQJ6PrkC4dHhYghwJzGBVP78DkBzA2U3kHoFNBuhj"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("57vGdcMZLnbNr4TZ4hgrpGJZGR9vTPhu8L9bNKDrqxKT"),
    baseTokenDecimals: 6,
});
exports.stepSolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("FmHGpt2scyJ8NuDuknnzQ7jectLyRRQmTyXxu5tRoD2j"),
    farmTokenMint: new web3_js_1.PublicKey("Gs1fM7EFS1rXkxhqs4mwu9uvSkupNzZgRbHGxG2NGRh7"),
    rewardTokenMint: new web3_js_1.PublicKey("StepAscQoEioFxxWGnh2sLBDFp9d8rvKz2Yp39iDpyT"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("GwrBA1F8rGummDCDd8NY9Eu1cLNuJqbT8WaGxgWpFwGL"),
    baseTokenDecimals: 6,
});
exports.slrsUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("8JkTAVHXChDPzsNkcJYRJ69mXN5nCcyb1qVBmbVhCH1n"),
    farmTokenMint: new web3_js_1.PublicKey("F3rWkGAtdjWcU1rr16Wq4YPTgFdsyb1oS1xdy5tr9K1r"),
    rewardTokenMint: new web3_js_1.PublicKey("SLRSSpSLUTP7okbCUBYStWCo1vUgyt775faPqz8HUMr"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("66xCxkffQZKBZLiHV3PDcfR8ANJTfnDRxPCaBdv4wxB7"),
    baseTokenDecimals: 6,
});
exports.portUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("C8k63XU8xzzvBpSraYWjHCNzvprjYKDadAML3MWSFbxW"),
    farmTokenMint: new web3_js_1.PublicKey("Zm2dmUuuBicmvHxGAnAzaohZR2Y86gXEV2WMfo8AoCa"),
    rewardTokenMint: new web3_js_1.PublicKey("PoRTjZMPXb9T7dyU7tpLEZRQj7e6ssfAE62j2oQuc6y"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("4CGxvZdwiZgVMLXiTdJHTkJRUTpTSJCtmtCRbSkAxerE"),
    baseTokenDecimals: 6,
});
exports.copeUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("HW21NT7v6ViM2Cs3S2tVgUnawACF8JRRwd2AA41a8HUh"),
    farmTokenMint: new web3_js_1.PublicKey("AtcMEt9caZxpunQV99pxED2rhpQmaDykBreEqBsYU11v"),
    rewardTokenMint: new web3_js_1.PublicKey("8HGyAAB1yoM1ttS7pXjHMa3dukTFGQggnFFH3hJZgzQh"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("9SDpBrfqNxjXcCzpKWM6yUKdfky975VJBD6xcu5cKf5s"),
    baseTokenDecimals: 6,
});
exports.bopUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("GYipJSPD7zcDoEABRCZfDSqh7zqYMmjyP8dNKEUnW1iC"),
    farmTokenMint: new web3_js_1.PublicKey("CjGUbKiH1QmFFjMqhAbJn4DrbjgBWUhQHV4LuzrgpFqi"),
    rewardTokenMint: new web3_js_1.PublicKey("BLwTnYKqf7u4qjgZrrsKeNs2EzWkMLqVCu6j8iHyrNA3"),
    rewardTokenDecimals: 8,
    baseTokenMint: new web3_js_1.PublicKey("A7vvbqENJj8kED3ABjphe8TvwpasQYtoWGKpjpLArMxa"),
    baseTokenDecimals: 6,
});
exports.samoUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("8VJT2SYGXgvQ8jYvh1Cq6mg83gkAVuTY3cHiULUy6Cit"),
    farmTokenMint: new web3_js_1.PublicKey("EdfAy8jwnvU1z61UaFUjwoRPFgD3UkkPvnhEBZjzwhv8"),
    rewardTokenMint: new web3_js_1.PublicKey("7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("9voVuTq1S9bFZkF2Jo44HoVG63w2xDRT8eBzB23YbQud"),
    baseTokenDecimals: 6,
});
exports.whapiUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("GFFQdLWvbWzrZBQrPQVyzGQfd6SWuZaiRwmkKA2bJPeY"),
    farmTokenMint: new web3_js_1.PublicKey("41VBoy8SGJzQnWGcxiBL4yM6H68FiPp74aMvsZGNGCbt"),
    rewardTokenMint: new web3_js_1.PublicKey("6VNKqgz9hk7zRShTFdg5AnkfKwZUcojzwAkzxSH3bnUm"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("Bfoi3RNnfdP5VeRGqvTA8MRN9ePGJoZgeKfe8WeBHUxE"),
    baseTokenDecimals: 6,
});
exports.slimUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("3RsBqqz9MiXa1uhFFd6tpGVQ66ptkFPCBAUyqLAXeCTE"),
    farmTokenMint: new web3_js_1.PublicKey("HhDk3ySWkVbMZjgBsFSnLtAeudDCrfZ6DNSRgxh2oRUp"),
    rewardTokenMint: new web3_js_1.PublicKey("xxxxa1sKNGwFtw2kFn8XauW9xq8hBZ5kVtcSesTT9fW"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("3K7aZhtwWJ2JS6GnbbgeDVnxd1q2hwhqasmgRsAMZ4yC"),
    baseTokenDecimals: 6,
});
exports.ninjaSolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("HPeAGzDEhYUNm4z4aV2PTjVjbESnNg6jBy6LtFjFJzQj"),
    farmTokenMint: new web3_js_1.PublicKey("Db7mPGrZbswvFmJ7MgZsM6CFhnXHMnrUDqr2hrzmi7Re"),
    rewardTokenMint: new web3_js_1.PublicKey("FgX1WD9WzMU3yLwXaFSarPfkgzjLb2DZCqmkx9ExpuvJ"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("7YyhptkxY81HPzFVfyCzA5UXxWdsNRD41ofLva3TuSpd"),
    baseTokenDecimals: 6,
});
exports.atlasUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("DTP1xr4EzFf1YDu4CeWTtWVsCBzFPk4HDEsL3AzoR3kB"),
    farmTokenMint: new web3_js_1.PublicKey("894ptAFT7d3inPsWTniCGL2NZpJDiXGvFZFfuHXA1w8F"),
    rewardTokenMint: new web3_js_1.PublicKey("ATLASXmbPQxBUYbxPsV97usA3fPQYEqzQBUHgiFCUsXx"),
    rewardTokenDecimals: 8,
    baseTokenMint: new web3_js_1.PublicKey("HFmY1ggCsCky1zJ1sfdkNR4zb3u5n38YNRdf4vsGu17t"),
    baseTokenDecimals: 6,
});
exports.polisUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("7h1zAHj2xzEw3eKfprYqG36aN5XwcZXBsYwM2haWQVzR"),
    farmTokenMint: new web3_js_1.PublicKey("FE1QJzi5RA5aKnTfSV3DAMN3z4uHUzSR5Z4drs9S5vB"),
    rewardTokenMint: new web3_js_1.PublicKey("poLisWXnNRwC6oBu1vHiuKQzFjGL4XDSu4g9qjz9qVk"),
    rewardTokenDecimals: 8,
    baseTokenMint: new web3_js_1.PublicKey("63JUKLnCAuNMPSPioEgbjjzp9Qk8qSEEM8eZqEtPqfLU"),
    baseTokenDecimals: 6,
});
exports.abrUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("98htZRc2QNd8BS9GGHoxkySZ9BiL9MAgHLEQfxzXYKk6"),
    farmTokenMint: new web3_js_1.PublicKey("7bp7psdaC3DVc86Hmdz5tAMEjgPjmCzgFEVALfqBwMmz"),
    rewardTokenMint: new web3_js_1.PublicKey("a11bdAAuV8iB2fu7X6AxAvDTo1QZ8FXB3kk5eecdasp"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("5uR5STASUmoGVHzqMeut98t26TfVkQqWU9f9dsv3NfJ6"),
    baseTokenDecimals: 6,
});
exports.kuroUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("BK3VXXDA4KxVurHbP3yPytGWp1HifbBtB4ugpzfDeq4v"),
    farmTokenMint: new web3_js_1.PublicKey("88RCQs9VFvqPjsRe3PKNzBeMtzCS9oS1a1CJuAnGnLZJ"),
    rewardTokenMint: new web3_js_1.PublicKey("2Kc38rfQ49DFaKHQaWbijkE7fcymUMLY5guUiUsDmFfn"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("6PGoaQdL9e463hdaFxHXsuPcjCHRK32CQ9PFKxvM7XY2"),
    baseTokenDecimals: 6,
});
exports.msolUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("5fhDMuGKRDPWVWXf7BBEwifRFrp6XwXctDQoG7UHGVt6"),
    farmTokenMint: new web3_js_1.PublicKey("9y3QYM5mcaB8tU7oXRzAQnzHVa75P8riDuPievLp64cY"),
    rewardTokenMint: new web3_js_1.PublicKey("MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("5r3vDsNTGXXb9cGQfqyNuYD2bjhRPymGJBfDmKosR9Ev"),
    baseTokenDecimals: 6,
});
exports.orcaMsolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("41ZYSekqDNtJ1BdGkTZVR1CJfBiFrud6HcT3HVUdSyWN"),
    farmTokenMint: new web3_js_1.PublicKey("876yhw4J4GHyynNJUtARYEnWGaejhrWC7Hy3DAm1pZxi"),
    rewardTokenMint: new web3_js_1.PublicKey("MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("3Duk5b6fLztPmS4ryV48FM1Q9WXUSMwz9jehAT4UtqpE"),
    baseTokenDecimals: 6,
});
exports.msolSolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("2SciNw7cEsKJc1PMRDzWCcEzvuScmEaUgmrJXCi9UFxY"),
    farmTokenMint: new web3_js_1.PublicKey("576ABEdvLG1iFU3bLC8AMJ3mo5LhfgPPhMtTeVAGG6u7"),
    rewardTokenMint: new web3_js_1.PublicKey("MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("3RTGL7gPF4V1ns1AeGFApT7cBEGVDfmJ77DqQi9AC6uG"),
    baseTokenDecimals: 6,
});
exports.stsolWstethDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("EmWtmApj1PtJMgMfzbYMmnmtuwfcBy7es3Tg2AR8xfW6"),
    farmTokenMint: new web3_js_1.PublicKey("5WXyG6zL1HmESPCSHHKBtqLuRPZCNgd9mTB25op87FkU"),
    rewardTokenMint: new web3_js_1.PublicKey("HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p"),
    rewardTokenDecimals: 8,
    baseTokenMint: new web3_js_1.PublicKey("3kT3oYuS1rCfhmqfgy6EKcbZdaJimaVEjoy25QiuEaoj"),
    baseTokenDecimals: 6,
});
exports.sypUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("gpy1dZRbPbLZ2KNr4wd2r9zoxERbWV8gWTqTL47KNnh"),
    farmTokenMint: new web3_js_1.PublicKey("BpHfwFwJwkZKWY5xVMC3oifMvWRy42R4VE1vPeBzg2G1"),
    rewardTokenMint: new web3_js_1.PublicKey("FnKE9n6aGjQoNWRBZXy4RW6LZVao7qwBonUbiD7edUmZ"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("Ds4VGZhZzS2PMFzhzKeC3mwcQjdiCG21R76fTVbsSJyJ"),
    baseTokenDecimals: 6,
});
exports.mndeMsolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("C9AfeUkti1ykMTebw9N1WoiEhNqwNiXinfXBwuLUv1BT"),
    farmTokenMint: new web3_js_1.PublicKey("2FMpVEhvxiFxhfideFUMNxCoUZK3TfhezzajoHGTQKP2"),
    rewardTokenMint: new web3_js_1.PublicKey("MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("2wPsMuzhEsC6GhV3qtFpmJF6atEgLGbnmQ8U43Y6fPxZ"),
    baseTokenDecimals: 6,
});
exports.msolUsdtDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("FNV9pGMWTYSMq5dRhmjpRwHpDJKV6JD4HTKsqZndKvuY"),
    farmTokenMint: new web3_js_1.PublicKey("7iKG16aukdXXw43MowbfrGqXhAoYe51iVR9u2Nf2dCEY"),
    rewardTokenMint: new web3_js_1.PublicKey("MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("Afvh7TWfcT1E9eEEWJk17fPjnqk36hreTJJK5g3s4fm8"),
    baseTokenDecimals: 6,
});
exports.msolWhethDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("D6oqo3F2KkJcePDoNZfbb8F7SPnRhP7WCC9FNktzVCDT"),
    farmTokenMint: new web3_js_1.PublicKey("3kFeVJUxhQS7PE7vV8pt9bhTCQrUDqeGf6AU4sjkLzVt"),
    rewardTokenMint: new web3_js_1.PublicKey("MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("58nifjPjF3CutGz2xMxvAMk7R9YgbVEc8Cstj4rCcs8j"),
    baseTokenDecimals: 6,
});
exports.btcMsolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("Cn7QNyosNQ8DyKEeMDPmtg66R7vKMXigcQ561kTkFD8E"),
    farmTokenMint: new web3_js_1.PublicKey("6uA1ADUJbvwYJZpzUn9z9LuyKoRVngBKcQTKdXsSivA8"),
    rewardTokenMint: new web3_js_1.PublicKey("MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("DzpLz78wuwyFsQToin8iDv6YK6aBEymRqQq82swiFh7r"),
    baseTokenDecimals: 6,
});
exports.ivnSolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("9DNgTpphCRXhkf8ySiSnf1L2CHwACZybgPNwsnLUdzfA"),
    farmTokenMint: new web3_js_1.PublicKey("5X71f6zUnVYWSxWM8wf942pzWLv1ZtCDhGCYqZipYutD"),
    rewardTokenMint: new web3_js_1.PublicKey("iVNcrNE9BRZBC9Aqf753iZiZfbszeAVUoikgT9yvr2a"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("HqajzzbGMST3yCCVBJuXvNVsWkY2DXqiBz9cTRmmyBMy"),
    baseTokenDecimals: 6,
});
exports.larixUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("GnFV3S7H6eM9V12EJwptv111VmV7W3AnEEfzdHJL25n7"),
    farmTokenMint: new web3_js_1.PublicKey("Huy453KXTaWaA3AiJeqLrAWWSMwoGjFU8nsEi7GiPY7n"),
    rewardTokenMint: new web3_js_1.PublicKey("Lrxqnh6ZHKbGy3dcrCED43nsoLkM1LTzU2jRfWe8qUC"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("DNAGfa7tK8csprRQmiDUwDaFfhw6ueHhVFHTCgTJ8HGs"),
    baseTokenDecimals: 6,
});
exports.gofxUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("EVDvbsD5f1qz7FpVyknrzB3bLK86FHfqVbcH9WqpWZtK"),
    farmTokenMint: new web3_js_1.PublicKey("BzZ3mLyXEaGt1Na1zxfZYjRJBFDbDckurc2LDq46irUx"),
    rewardTokenMint: new web3_js_1.PublicKey("GFX1ZjR2P15tmrSwow6FjyDYcEkoFb4p4gJCpLBjaxHD"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("B95rdqSY4dqPwmt295XwBZZqZJYLmqDNXU6NvBpT4ep4"),
    baseTokenDecimals: 6,
});
exports.woofUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("9M9JKUWatibJKghbDDjHXLGGY18dLPLHgs6a4WiP9EAe"),
    farmTokenMint: new web3_js_1.PublicKey("G6kaJ4NrCS9VoNgPiMuR5ENFrGJBfMe98PXFgVYkBKts"),
    rewardTokenMint: new web3_js_1.PublicKey("9nEqaUcb16sQ3Tn1psbkWqyhPdLmfHWjKGymREjsAgTE"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("4HaQXDz9gdLgKUjvNVtnLyNZoWNYKjh3XxH1TpLgiwmi"),
    baseTokenDecimals: 6,
});
exports.sdogeUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("J3AmdMedeE1mhwbZzZFysER8CZdMqxL3HoyLxCmub9mc"),
    farmTokenMint: new web3_js_1.PublicKey("3upLraKWWJ3Tv8JgwygBszT1nQwksYBnx2XWEgaGMTt6"),
    rewardTokenMint: new web3_js_1.PublicKey("8ymi88q5DtmdNTn2sPRNFkvMkszMHuLJ1e3RVdWjPa3s"),
    rewardTokenDecimals: 0,
    baseTokenMint: new web3_js_1.PublicKey("4kXYyZAMBdvgDaBKUGvRWJnHw9af7sCUPvpQ68PEdP8b"),
    baseTokenDecimals: 6,
});
exports.catoUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("7YMRTqXna5vtCkncV6hxhUd7qCrv621iRq96b7LYySR2"),
    farmTokenMint: new web3_js_1.PublicKey("2YbEbQD8DP8QyRQnp7CXwdD6BqeJsry4MkWZgxovZPMm"),
    rewardTokenMint: new web3_js_1.PublicKey("5p2zjqCd1WJzAVgcEnjhb9zWDU7b9XVhFhx4usiyN7jB"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("BHtZnTBMeY4EBEW5egGnuK5bdW12v6Dod6wFav79AyYx"),
    baseTokenDecimals: 6,
});
exports.oogiUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("F1jqYgps3MsGvWJ9WKq7NYx1eT4xjPDruFtQ9gMD1pLk"),
    farmTokenMint: new web3_js_1.PublicKey("3CiL5pm15BiLH8PBrhx7uFJr2uYoeBNZk55kuphp48Py"),
    rewardTokenMint: new web3_js_1.PublicKey("H7Qc9APCWWGDVxGD5fJHmLTmdEgT9GFatAKFNg6sHh8A"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("FiwSk36yi1DNWcuQUeNipAc1VKxa9Wv9AR2xFvyKUxAE"),
    baseTokenDecimals: 6,
});
exports.sonarUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("E3hyqjh9R7UWMwkMSxYHws4qdnpMSzXwkuJfVcEfmJwb"),
    farmTokenMint: new web3_js_1.PublicKey("9tpC9vawMtim6q3MaQHfPPwFnnErh58qXhdqczS2p7sd"),
    rewardTokenMint: new web3_js_1.PublicKey("sonarX4VtVkQemriJeLm6CKeW3GDMyiBnnAEMw1MRAE"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("GWmwwMGYBG4NqYdnsYrudzBnbgDC49MkBxdzhfLA9kVY"),
    baseTokenDecimals: 6,
});
exports.upsUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("8V2RqGTNmz2xYBdonfvkLstGWV7ibHopoZMQeaV9rykb"),
    farmTokenMint: new web3_js_1.PublicKey("9Ux3QdsLpLXCNFsACXjp4F2D5ihkWZnVhdDiTDTvw2Ny"),
    rewardTokenMint: new web3_js_1.PublicKey("EwJN2GqUGXXzYmoAciwuABtorHczTA5LqbukKXV1viH7"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("9fuv3emLQXECrTWqm2HaKT3wQhmvmgqmReZSMcu8PfpH"),
    baseTokenDecimals: 6,
});
exports.fantUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("2J5exwdrNhqK1y1DF5jPQV2JieZ5xBsR3jDrZRh8PfPZ"),
    farmTokenMint: new web3_js_1.PublicKey("FftGKB7DThU5N96xYQ8BWkRBpFTKDvSzN7hCCNxuypsE"),
    rewardTokenMint: new web3_js_1.PublicKey("FANTafPFBAt93BNJVpdu25pGPmca3RfwdsDsRrT3LX1r"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("C2EcthTMaC5eATXVaXg5ctvMfUYYgFyNibybHes5D3S6"),
    baseTokenDecimals: 6,
});
exports.stsolUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("HhBtZgPaFb5rodxXsZbtZUqxk2vEg6tdzM4AcaG2PjcL"),
    farmTokenMint: new web3_js_1.PublicKey("CejKA1pePxny3iprRDEyiojfTKNxxX2bjmKToDGZqwvh"),
    rewardTokenMint: new web3_js_1.PublicKey("HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p"),
    rewardTokenDecimals: 8,
    baseTokenMint: new web3_js_1.PublicKey("3u2dNfGuU6C3vmSg5EvLPUpX57b3niqhWBV5Gc3WDEf5"),
    baseTokenDecimals: 6,
});
exports.wmpUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("2aUMqCFWK3PKeScspP4SqRFS7fag18TDHv2VsaBfSS72"),
    farmTokenMint: new web3_js_1.PublicKey("4SWDBKeb44RQMYmyorUiuF94Fqf1toketzJyvLSjVtHN"),
    rewardTokenMint: new web3_js_1.PublicKey("BygDd5LURoqztD3xETc99WCxLUbTi6WYSht9XiBgZ4HW"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("Ce3VXSQGFEHqrQSEzcVb2Ro88Mcn56cYBUAXFx5tL9bo"),
    baseTokenDecimals: 6,
});
exports.chicksUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("61uQt5e3Vd3cps6mS5xnL1UVLe3rwNS4aDaa3F2V7h3z"),
    farmTokenMint: new web3_js_1.PublicKey("3EPqpch8B8nk37gcxKCnQ5n6VSu4NyetcyZen636NYfb"),
    rewardTokenMint: new web3_js_1.PublicKey("cxxShYRVcepDudXhe7U62QHvw8uBJoKFifmzggGKVC2"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("GTcxXu3yw7yotq2UtAMvpEbx84ozJyrBwy77Cmem7m4X"),
    baseTokenDecimals: 6,
});
exports.unqUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("G5cvdbZwbwAV4eW8ENQy5tmbzaJu8TbdgF5MLXgLPXj3"),
    farmTokenMint: new web3_js_1.PublicKey("75jdeQfjFnNzg4dSbkA3HxTN63hXo99MrpafDJsz5Gcy"),
    rewardTokenMint: new web3_js_1.PublicKey("UNQtEecZ5Zb4gSSVHCAWUQEoNnSVEbWiKCi1v9kdUJJ"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("CsGTrgJ6oLx9UQow9aLjuAacjXTY53zT8B2FnAbfVZAS"),
    baseTokenDecimals: 6,
});
exports.meanUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("9TQ4HQXymNLVKEMve5peFXst5VPDXX1oZ8VTVuQeG7W7"),
    farmTokenMint: new web3_js_1.PublicKey("FaFGyRihhc8XUqWXNrVAd6gjRsyJ98mogonH4EN8WUPb"),
    rewardTokenMint: new web3_js_1.PublicKey("MEANeD3XDdUmNMsRGjASkSWdC8prLYsoRJ61pPeHctD"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("3h2VBX8533NB8eEH8rPXMdayodFDbgTHpbav6JqLZAQq"),
    baseTokenDecimals: 6,
});
exports.wagUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("5dA9HWoFnBagq25bpMPA79bKxQkQ6YXvBtYQn6Le6vyp"),
    farmTokenMint: new web3_js_1.PublicKey("BYno9HNAiLoi7K8Wa5c1f5EE7hukzZPUQZDesGxBPzhh"),
    rewardTokenMint: new web3_js_1.PublicKey("5tN42n9vMi6ubp67Uy4NnmM5DMZYN8aS8GeB3bEDHr6E"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("8Wu5sJpERA1J5iWcT8aMpt9cTAfKDLPbLpGjNsJoPgLc"),
    baseTokenDecimals: 6,
});
exports.shdwUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("GJ9EixfM3noFT1b7Y5uAzV1qJSFev5uPaRrmfMoibck6"),
    farmTokenMint: new web3_js_1.PublicKey("HPv7XJ16t4pZVNTBPHoYY19xiv4pHjjSbarE7Km3jJ1R"),
    rewardTokenMint: new web3_js_1.PublicKey("SHDWyBxihqiCj6YekG2GUr7wqKLeLAMK1gHZck9pL6y"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("7WWHfufv8vuBC1x7GXA3pu7kgNhEQkXoq3CtbaQihAJ9"),
    baseTokenDecimals: 6,
});
exports.shdwSolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("2EHJ8ToKpJfXyAfechjH9QSbVMKTSViYPdJRepQz7V8S"),
    farmTokenMint: new web3_js_1.PublicKey("8HgXuNigmLvfsgDun1vQso6pBuj7sVvDqpcergjtu3dz"),
    rewardTokenMint: new web3_js_1.PublicKey("SHDWyBxihqiCj6YekG2GUr7wqKLeLAMK1gHZck9pL6y"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("BDStVBt4NS5bfda25ubK51kVRioV4yjKKCPbe96jeEms"),
    baseTokenDecimals: 6,
});
exports.aartUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("BKFSHwhqDHRFM3PzNh7BG1vgxtEbx1A9Co1JN6z9SLYz"),
    farmTokenMint: new web3_js_1.PublicKey("FvmyENsZf68ezNbHxRqv1hAb7phfhdnXMfQm74Vcpnkn"),
    rewardTokenMint: new web3_js_1.PublicKey("F3nefJBcejYbtdREjui1T9DPh5dBgpkKq7u2GAAMXs5B"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("Bg7pZq7KdsQsnCQBeNpa4XVnrWfjUJNu3ViHoNB7YUZU"),
    baseTokenDecimals: 6,
});
exports.basisUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("DasaXe2Wqcks6csFv1bWwdW41mV8rMD5c27Uw9rFYVu4"),
    farmTokenMint: new web3_js_1.PublicKey("8XtNSYBhLHa4cYzNsXd6yDAweMECumrxFJ7F2qxk2xN"),
    rewardTokenMint: new web3_js_1.PublicKey("Basis9oJw9j8cw53oMV7iqsgo6ihi9ALw4QR31rcjUJa"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("4yx2aHMa7N4m1uUaBRy9QPtpstw3HFPtvcCPJQaGFHKL"),
    baseTokenDecimals: 6,
});
exports.slcUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("DdYbYsTgpp3pBdGhN9sSdgRJs4ijKZQuDqzqKz6qmScs"),
    farmTokenMint: new web3_js_1.PublicKey("6VWtaecVXHuorMwVrZ4GAnYC3T5MG1YdQLYNPGyWzQHh"),
    rewardTokenMint: new web3_js_1.PublicKey("METAmTMXwdb8gYzyCPfXXFmZZw4rUsXX58PNsDg7zjL"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("X8GnAvxq942xXjNzqYPFQQ9JstJQNEA81uPeNoQpgrN"),
    baseTokenDecimals: 6,
});
exports.wustSolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("6p47BJ5d9JmywGpbZa8MapJmhcLdu5w2uNg9NQTJKvi"),
    farmTokenMint: new web3_js_1.PublicKey("BkcNGF6iV6NTfoH1ocoywXrfDpPWiWUu7eARBT42e57v"),
    rewardTokenMint: new web3_js_1.PublicKey("F6v4wfAdJB8D8p77bMXZgYt8TDKsYxLYxH5AFhUkYx9W"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("F49Cm3srGucQCBanA2xL7nSKHfH1QZd9vLdFtkg4LKnq"),
    baseTokenDecimals: 6,
});
exports.wustUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("87ArJQZCaUwLkZfaRHbjU8geq6F8HsFcCtoppCV7Buby"),
    farmTokenMint: new web3_js_1.PublicKey("DGxLh6BykS8C4zLrBRRD4cHZRdy73eVqZ5283JkAdU7d"),
    rewardTokenMint: new web3_js_1.PublicKey("F6v4wfAdJB8D8p77bMXZgYt8TDKsYxLYxH5AFhUkYx9W"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("2mhVUMsG7eb3XhHjAbKpRZgWNyGZNCiWU7dRxauzZaGL"),
    baseTokenDecimals: 6,
});
exports.msolWustDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("5z183wdSnkc1k9EmnXncE66VQYYLn8MkxG9gJTo5gabj"),
    farmTokenMint: new web3_js_1.PublicKey("2q4rZYZgicviVBsb8CTWGR3bGWys3ePFDYtrayyQpwyi"),
    rewardTokenMint: new web3_js_1.PublicKey("F6v4wfAdJB8D8p77bMXZgYt8TDKsYxLYxH5AFhUkYx9W"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("6wqME6zPQzGDLugpnwZnVBKBbATC5nTaEa781Vj98yvM"),
    baseTokenDecimals: 6,
});
exports.wlunaWustDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("4G49eoJrt6DGWnVR84tgNzw987cjwD4cPYd3SUuo6mrd"),
    farmTokenMint: new web3_js_1.PublicKey("HQnMrrLP5RxYN6Peuu9cfDNQK81yd3TtBt3ZCjoD8UAf"),
    rewardTokenMint: new web3_js_1.PublicKey("F6v4wfAdJB8D8p77bMXZgYt8TDKsYxLYxH5AFhUkYx9W"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("3ZGKT28NXAqb2YtUAMvKixQvHNHT31Q5mK8AC2iBMs29"),
    baseTokenDecimals: 6,
});
exports.stsolWustDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("CXbiLWJoYcVmV7GcF9xSwwMPSgHh5rHRLJB84F54R4qU"),
    farmTokenMint: new web3_js_1.PublicKey("DxiftFoeRxHk15N4rDYzpwtGhfK3LqSn4gWDCaEkMksE"),
    rewardTokenMint: new web3_js_1.PublicKey("HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p"),
    rewardTokenDecimals: 8,
    baseTokenMint: new web3_js_1.PublicKey("3Q44iV4URXdbS4Tk1PGs5VdWQoCxHB7zdcdMnemo8jfH"),
    baseTokenDecimals: 6,
});
exports.cmfiUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("E5m2jmo8DhLiY93nh95K4XCxvZgt66JzaJnwyt5CPmVR"),
    farmTokenMint: new web3_js_1.PublicKey("Dr7ks8oV6iRTT2R2uqE4x6Egmg3djt14aCEciNhq3SF9"),
    rewardTokenMint: new web3_js_1.PublicKey("5Wsd311hY8NXQhkt9cWHwTnqafk7BGEbLu8Py3DSnPAr"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("E8RVjS24pBuF3oCCeJVAgC4RQ7mVa5P3FGXqGbiczzvD"),
    baseTokenDecimals: 6,
});
exports.hbbUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("CvYum3awvbxPVa5FpZRBs97Prg5GDmLha4X6mggX7R1p"),
    farmTokenMint: new web3_js_1.PublicKey("FYdet59VXibMU8FhnbZamZTVMsjnCyNevUqzWUrJfGum"),
    rewardTokenMint: new web3_js_1.PublicKey("HBB111SCo9jkCejsZfz8Ec8nH7T6THF8KEKSnvwT6XK6"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("BxPd4x7gm6WQF6jsDCRQuQVZ2bopQ4KegMrqyQgdkPDk"),
    baseTokenDecimals: 6,
});
exports.celoUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("DrDpUFTvHjznYEugMqcXTqspdmQ6GZhuBzjodU7A6qWU"),
    farmTokenMint: new web3_js_1.PublicKey("4EUgJ5LSF3UDVbVAYzqE133NXAxfhHZSvqx214EJkNkQ"),
    rewardTokenMint: new web3_js_1.PublicKey("a11bdAAuV8iB2fu7X6AxAvDTo1QZ8FXB3kk5eecdasp"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("9RRBDWZGWGVnHgazCpb9R1XApmHLBDya8Tq1yrzoMsKY"),
    baseTokenDecimals: 6,
});
exports.sbUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("74hUtc4MaZUt4YDDjVSqHyUbPT8kbxW16mbAharpZhPV"),
    farmTokenMint: new web3_js_1.PublicKey("EdkL9TXiT2NV1LND5nj4kR1L9SYjozWpJvbcStvrUUJF"),
    rewardTokenMint: new web3_js_1.PublicKey("SuperbZyz7TsSdSoFAZ6RYHfAWe9NmjXBLVQpS8hqdx"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("3Wppx86xN7Dg7GLUTD9C7AqCt68qZkWzNYUdCS7t1pCk"),
    baseTokenDecimals: 6,
});
exports.ftmUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("7cYU6ux4GwB1X2FKaRruQ9dbDZKUrxZAaWGNJfeoHeGJ"),
    farmTokenMint: new web3_js_1.PublicKey("4r3uYr4Ph4ctw1chiUaAtwon4stFGHqbSJ6Lym139rM9"),
    rewardTokenMint: new web3_js_1.PublicKey("a11bdAAuV8iB2fu7X6AxAvDTo1QZ8FXB3kk5eecdasp"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("E7Af9Fa2U1YqEkQpAYjXwDB5TJUou9VcN3ot33Gj6UY9"),
    baseTokenDecimals: 6,
});
exports.stsolUsdtDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("AihUw1sSxuzL2i1ELWMXmiRT3UgyRds68iNXwFKbyQnP"),
    farmTokenMint: new web3_js_1.PublicKey("2kftNiMcrjp9CRqtoLZNwApGEoirfunHPvunenGLz17y"),
    rewardTokenMint: new web3_js_1.PublicKey("HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p"),
    rewardTokenDecimals: 8,
    baseTokenMint: new web3_js_1.PublicKey("CtvKaRLzCzRZCcYPwai7NCYBK4ArBj2oD6BfvMgJoiCN"),
    baseTokenDecimals: 6,
});
exports.seededUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("EsSWWXMipKC4aVHcKjkYdvNjKphxk86DGCYhdF2huoso"),
    farmTokenMint: new web3_js_1.PublicKey("AjCuWDEZYC9FuwdYpdAy4vLoY6PxMNF8xitQeaNZd1nW"),
    rewardTokenMint: new web3_js_1.PublicKey("seedEDBqu63tJ7PFqvcbwvThrYUkQeqT6NLf81kLibs"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("H15WptGntFQifZmJHUzYBV9Mv7P27ofavEsF6yqpLTdX"),
    baseTokenDecimals: 6,
});
exports.onesolSolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("2GpXwBPbooPwqxLuM1bW9FaSK5hxaJqNkr7WdPsjoPca"),
    farmTokenMint: new web3_js_1.PublicKey("DuTybhuJgwECRHDLgtrBuKnfhxve2zXoYXD6cnDikzoT"),
    rewardTokenMint: new web3_js_1.PublicKey("4ThReWAbAVZjNVgs5Ui9Pk3cZ5TYaD9u6Y89fp6EFzoF"),
    rewardTokenDecimals: 8,
    baseTokenMint: new web3_js_1.PublicKey("Hva9oLa2GjoKdB45WoHujsX7MTpehByPshMrQpNDmkFq"),
    baseTokenDecimals: 6,
});
exports.mmaUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("7w9ztS9wcnsk2gqCEvKmQFUBbwhPZPfb5CXhC2onUWXs"),
    farmTokenMint: new web3_js_1.PublicKey("9hSvWrDqgAr7E8sWpatomKcgAGy3hvgAWPXtGPySNbsy"),
    rewardTokenMint: new web3_js_1.PublicKey("MMAx26JtJgSWv6yH48nEHCGZcVvRbf9Lt9ALa7jSipe"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("H9qkbU2XYSQTk6JJJ2TMuZBthfkeCHigmuxB5jiQaamn"),
    baseTokenDecimals: 6,
});
exports.puffSolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("HNTRmR36ZXBNakYpfPPq7VJaXzLrNY6hN1Xz9p5s8wWA"),
    farmTokenMint: new web3_js_1.PublicKey("Grn1ipjetWK7sz5Bq9S1nwD7F7dUsFzM44diFxbMruvK"),
    rewardTokenMint: new web3_js_1.PublicKey("G9tt98aYSznRk7jWsfuz9FnTdokxS6Brohdo9hSmjTRB"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("Ge5kuYg5PekrPUeKzngw97Cnfngj8j6NCX5q1jTBkWSW"),
    baseTokenDecimals: 6,
});
exports.saoUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("9eVq1gfxbgRpX8W4XctocHEm2U8eT5KzTcnMduFThoz1"),
    farmTokenMint: new web3_js_1.PublicKey("47PY3ET5oziYHEt1mddYFE7opJnqPbp98yZ1xf6cpgER"),
    rewardTokenMint: new web3_js_1.PublicKey("2HeykdKjzHKGm2LKHw8pDYwjKPiFEoXAz74dirhUgQvq"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("A9BeGSRJJYXPrMs81rVZxvkk16fopzgG5YkngntgTu7p"),
    baseTokenDecimals: 6,
});
exports.srlyv2SolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("7GAp1xR64GjjbyHbvE28Rt9xLEBLTjk4W3AemRMvTcrS"),
    farmTokenMint: new web3_js_1.PublicKey("4Xtia3w6AKBNPCYmFZCAQV8CNwUmHYrVQiNMEGdMkRMg"),
    rewardTokenMint: new web3_js_1.PublicKey("RLYv2ubRMDLcGG2UyvPmnPmkfuQTsMbg4Jtygc7dmnq"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("6gkZ7QUmxmwPLS2NK3Dr6YHtTPZs6GQrkA595WSx5iLe"),
    baseTokenDecimals: 6,
});
exports.zbcUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("AMVd1e1mvWxvJ9LPU3GWjRQ7bfhkwVB69RiL5oRkyAtC"),
    farmTokenMint: new web3_js_1.PublicKey("2ygRp6eeFaBmfEZngpkVwuVs8hmRbe3FMnF4qiJiFzhT"),
    rewardTokenMint: new web3_js_1.PublicKey("zebeczgi5fSEtbpfQKVZKCJ3WgYXxjkMUkNNx7fLKAF"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("7Hoi4adCSBzERdvSiUXtVDz79tiDxMD5HpDv7m9rs3Sb"),
    baseTokenDecimals: 6,
});
exports.novaUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("DQW177Qu36iFAjeXmywHPoeBVvVHa6ngwm2RWzL6ffVj"),
    farmTokenMint: new web3_js_1.PublicKey("22MzBGjYuoZuAyoXEnNSAxBatpRwUFhDzusY1NboiMLj"),
    rewardTokenMint: new web3_js_1.PublicKey("BDrL8huis6S5tpmozaAaT5zhE5A7ZBAB2jMMvpKEeF8A"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("H2jZrZKx1fLY9LPi2ad5G2DVBWzMmUwPQLjRFERUroEH"),
    baseTokenDecimals: 6,
});
exports.hbbUsdhDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("9qCNAmcuo8odN67UJopN7BEKXBfT2kc7xf98612Ria1z"),
    farmTokenMint: new web3_js_1.PublicKey("8Lb3tMPCUwYPWDRWsujSHs2grXBssXm8GdXyn3zh1GmU"),
    rewardTokenMint: new web3_js_1.PublicKey("HBB111SCo9jkCejsZfz8Ec8nH7T6THF8KEKSnvwT6XK6"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("DUGuXcpHX9y6WdTKgDm8XsECbf7pApA7ZvAsjUqHUoYn"),
    baseTokenDecimals: 6,
});
exports.zigUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("6rGP2znXXGU9vWN6Te5D3DaKvSYQKXB8chaLdVcsSJsF"),
    farmTokenMint: new web3_js_1.PublicKey("87om2hz9nKKdS3sinYtGXCBuPACPd7xrnKBS3UwyZAoJ"),
    rewardTokenMint: new web3_js_1.PublicKey("ANqY8h3sRSwkC29FvGJenAh7VGRABVVx7Ls6Mq4BuGT"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("9nKbvVW9TjdnKt2NbfMaf9PHcg8zWWFgGTBfzUXdFcke"),
    baseTokenDecimals: 6,
});
exports.srlySolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("5dGXUyewQ652r6FTC1yduvxxSVspvmcQRY5wFCVNZeg3"),
    farmTokenMint: new web3_js_1.PublicKey("919hW5ubB18RFEU6KmP9AKfMxk6drESjc54sxewfz1Mp"),
    rewardTokenMint: new web3_js_1.PublicKey("sRLY3migNrkC1HLgqotpvi66qGkdNedqPZ9TJpAQhyh"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("HfkZkoPqZzd7S1F3BhkJN2Es1C5vabVYqfCecXG2MGHP"),
    baseTokenDecimals: 6,
});
exports.takiSrlyDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("GMa3zuo9DGHu9hfnQ6noUpsmtX1wJH9AF84q37TfmtfW"),
    farmTokenMint: new web3_js_1.PublicKey("5gR2iQfkNe5XgApAUqMsbBjkXkmf4yineBFkhpm4yKgD"),
    rewardTokenMint: new web3_js_1.PublicKey("Taki7fi3Zicv7Du1xNAWLaf6mRK7ikdn77HeGzgwvo4"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("13HwDUH7yREGAwHSWnx4Kgnn6huL8yfCCBqaBr6NGVH6"),
    baseTokenDecimals: 6,
});
exports.stsolSolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("D7tPD1ZP46fH2jJoL8YH75aPkGQVXFJVYiX7G2NVKt3T"),
    farmTokenMint: new web3_js_1.PublicKey("HaBymyP8zL1pgpnkdAn14pmTQbLQpiraYwd1uEzkReG9"),
    rewardTokenMint: new web3_js_1.PublicKey("HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p"),
    rewardTokenDecimals: 8,
    baseTokenMint: new web3_js_1.PublicKey("7ZvhrqdnP2NrQi2YCjQNd8jKJpLDMUuZg4wP7XJPjnVX"),
    baseTokenDecimals: 6,
});
exports.btcStsolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("kDV1QjYknEeqMRaTGssdPmnnKLxTavUBFf3VkNn89aP"),
    farmTokenMint: new web3_js_1.PublicKey("B91eD8Z2TfR21NxnwHv8H2GihT25qNRKtMxhtZfz4Kgk"),
    rewardTokenMint: new web3_js_1.PublicKey("HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p"),
    rewardTokenDecimals: 8,
    baseTokenMint: new web3_js_1.PublicKey("GjBW6751wNJ99zYkr9QwSi98bvj1NgyAcYHfQDpemJfq"),
    baseTokenDecimals: 6,
});
exports.stsolWhethDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("HcWSgX8HKHTDrbdHwMe2KgQSixs3KM2PNXEowT6bkyB7"),
    farmTokenMint: new web3_js_1.PublicKey("DkMeCWXirFsQZPYfgDs4voqvQbYGGQiwYoPQkUn4fU9W"),
    rewardTokenMint: new web3_js_1.PublicKey("HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p"),
    rewardTokenDecimals: 8,
    baseTokenMint: new web3_js_1.PublicKey("9V9CavjccKJH66m4MJHzVBYzY2ZM3U3D5xUCP7HSyAUz"),
    baseTokenDecimals: 6,
});
/**
 * Mapping for OrcaFarm parameters
 * Key: baseTokenMint : OrcaFarmParams
 */
exports.orcaFarmConfigs = {
    APDFRM3HMr8CAGXwKHiu2f5ePSpaiEJhaURwhsRrUUt9: exports.solUsdcAqFarm,
    FZthQCuYHhcfiDma7QrX7buDHwrZEd7vL8SjS6LQa3Tx: exports.solUsdtAqFarm,
    "71FymgN2ZUf7VvVTLE8jYEnjP3jSK1Frp2XT1nHs8Hob": exports.ethSolAqFarm,
    "3e1W6Aqcbuk2DfHUwRiRcyzpyYRRjg6yhZZcyEARydUX": exports.ethUsdcAqFarm,
    "5kimD5W6yJpHRHCyPtnEyDsQRdiiJKivu5AqN3si82Jc": exports.raySolAqFarm,
    ADrvfPBsRcJfGsN6Bs385zYddH52nuM5FA8UaAkX9o2V: exports.ropeSolAqFarm,
    "8nTzqDXHriG2CXKbybeuEh1EqDQMtrbYMFWcP7AkiDaP": exports.stepSolAqFarm,
    "9tf8rBSEQYG7AqL896fN2nZi1iYPqpWaLEdpbeQaC1Vy": exports.srmSolAqFarm,
    EsYaDKJCmcJtJHFuJYwQZwqohvVMCrFzcg8yo3i328No: exports.fttSolAqFarm,
    CzieDbGRdN1QGaGDNpSqzEA18bi881ccvkkGZi51pe1k: exports.copeSolAqFarm,
    "7tYCdLN84EnTMkxM7HNamWJx7F4xgKe9KiiWvLyWjbgT": exports.oxySolAqFarm,
    Acxs19v6eUMTEfdvkvWkRB4bwFCHm3XV9jABCy7c1mXe: exports.btcSolAqFarm,
    HiwRobjfHZ4zsPtqCC4oBS24pSmy4t8GGkXRbQj4yU6L: exports.merSolAqFarm,
    EYsNdtyu4gGTaGz8N5m5iQ3G1N6rDyMbR72B3CqbWW4W: exports.fidaSolAqFarm,
    "99pfC8fWymXgbq3CvrExhx3UxQDC1fMWEWLbNT83F45e": exports.mapsSolAqFarm,
    H2uzgruPvonVpCRhwwdukcpXK8TG17swFNzYFr2rtPxy: exports.usdcUsdtAqFarm,
    "2uVjAuRXavpM6h1scGQaxqb6HVaNRn6T2X7HHXTabz25": exports.orcaSolAqFarm,
    n8Mpu28RjeYD7oUX3LG1tPxzhRZh3YYLRSHcHRdS3Zx: exports.orcaUsdcAqFarm,
    HEvnD66WcBfTajS9adUYnGRBMDehFtLySiFHSD6kEBWs: exports.kinSolAqFarm,
    D6N9j8F2DhtzDtrdpT74y3u2YmYAzcggiLc3nTjqux9M: exports.samoSolAqFarm,
    "3PD9SZFwXKkXr4akLf4ofo37ZUMycwML89R2P3qxcbZG": exports.liqUsdcAqFarm,
    AZpo4BJHHRetF96v6SGinFZBMXM4yWMo4RA8C4PriDLk: exports.snyUsdcAqFarm,
    "8PSfyiTVwPb6Rr2iZ8F3kNpbg65BCfJM9v8LfB916r44": exports.msolUsdcAqFarm,
    AtB4nUmdyQfuWWJ9xAHw9xyVnJFfSjSuVWkiYan8y86w: exports.slrsUsdcAqFarm,
    F8gPSpwVHj8FdAJAYULDuZBxFEJut87hUbARYYx3471w: exports.portUsdcAqFarm,
    CS7fA5n4c2D82dUoHrYzS3gAqgqaoVSfgsr18kitp2xo: exports.sbrUsdcAqFarm,
    Dkr8B675PGnNwEr9vTKXznjjHke5454EQdz3iaSbparB: exports.scnsolUsdcAqFarm,
    C2YzN6MymD5HM2kPaH7bzcbqciyjfmpqyVaR3KA5V6z1: exports.psolUsdcAqFarm,
    "29cdoMgu6MS2VXpcMo1sqRdWEzdUR9tjvoh8fcK8Z87R": exports.msolSolAqFarm,
    C7TH2jEJJaxVwwuvkbcDGfHUiZvEkkeYjyAcdTMi5ujb: exports.orcaPaiAqFarm,
    CVapmQn7HaU1yMDW3q6oUV4hx6XoYv54T4zfGXkuJqkA: exports.orcaMsolAqFarm,
    APNpzQvR91v1THbsAyG3HHrUEwvexWYeNCFLQuVnxgMc: exports.scnsolSolAqFarm,
    FZ8x1LCRSPDeHBDoAc3Gc6Y7ETCynuHEr5q5YWV7uRCJ: exports.atlasUsdcAqFarm,
    GteBdo9sqE7T41G8AJsaG9WHW48uXBwsLLznmu2TBdgy: exports.polisUsdcAqFarm,
    "2gXDJZ7XAtQEtf4PRSQZKoq1WMuu1H44tQanbMA3YVpu": exports.bopUsdcAqFarm,
    "6VK1ksrmYGMBWUUZfygGF8tHRGpNxQEWv8pfvzQHdyyc": exports.samoUsdcAqFarm,
    "4X1oYoFWYtLebk51zuh889r1WFLe8Z9qWApj87hQMfML": exports.ninjaSolAqFarm,
    BVWwyiHVHZQMPHsiW7dZH7bnBVKmbxdeEjWqVRciHCyo: exports.slimUsdcAqFarm,
    ELfBngAgvLEHVBuJQhhE7AW6eqLX7id2sfrBngVNVAUW: exports.whapiUsdcAqFarm,
    HsauTv9s52Zv12eaDuSp6y7BEm4e4BHEyAsbdjyyWzPK: exports.copeUsdcAqFarm,
    GHuoeq9UnFBsBhMwH43eL3RWX5XVXbSRYJymmyMYpT7n: exports.sunnyUsdcAqFarm,
    EorFh8siFyLF1QTZ7cCXQaPGqyo7eb4SAgKtRH8Jcxjd: exports.grapeUsdcAqFarm,
    GMzPbaCuQmeMUm1opH3oSCgKUjVgJUW14myq99RVPGX5: exports.abrUsdcAqFarm,
    DRknxb4ZFxXUTG6UJ5HupNHG1SmvBSCPzsZ1o9gAhyBi: exports.kuroUsdcAqFarm,
    "2toFgkQDoPrTJYGDEVoCasPXuL9uQnjvXJaDwa9LHyTx": exports.mediaUsdcAqFarm,
    "4SBx8GXu8HhcVHWydQv1vsDdZs3G93XSL9CtMBny6hS5": exports.tulipUsdcAqFarm,
    H9yC7jDng974WwcU4kTGs7BKf7nBNswpdsP5bzbdXjib: exports.mngoUsdcAqFarm,
    Eswigpwm3xsipkTqahGi2PEJsJcULQBwZgxhQpr6yBEa: exports.stsolWstethAqFarm,
    qJxKN9BhxbYvRNbjfK2uAVWboto6sonj8XC1ZEW5XTB: exports.sypUsdcAqFarm,
    "74B9aMS7SA832xKngt5VLKmWAP3pa3qkUzWncTmQSsGF": exports.stsolWldoAqFarm,
    "7aYnrdmdCRodDy2Czn6keUquUhjF1jPEmfwZPh488z8U": exports.whethSolAqFarm,
    "7NPtjjAP7vhp4t5NCLyY4DY5rurvyc8cgZ2a2rYabRia": exports.whethUsdcAqFarm,
    "5PHS5w6hQwFNnLz1jJFe7TVTxSQ98cDYC3akmiAoFMXs": exports.mndeMsolAqFarm,
    Df6XNHMF3uRVZnz7LCEGiZVax6rXgz76owtVkBHEjSb6: exports.wagUsdcAqFarm,
    "9cMWe4UYRPGAUUsTkjShJWVM7bk8DUBgxtwwH8asFJoV": exports.msolUsdtAqFarm,
    "5qoTq3qC4U7vFxo3iCzbXcaD1UEmDeCD63Dsuoct71oV": exports.msolWhethAqFarm,
    "8nKJ4z9FSw6wrVZKASqBiS9DS1CiNsRnqwCCKVQjqdkB": exports.btcMsolAqFarm,
    DfgCnzaiTXfPkAH1C1Z441b5MzjjTCEh134ioxqRZxYf: exports.ivnSolAqFarm,
    "8sfThep3io4gvcGeuoAg1Rs8GDwKJjtcdAFHqQSSNAVE": exports.larixUsdcAqFarm,
    "6jCERp5hKj37PCXP3VTjCDJeoPuSpnMDMz5A6jWQv3yS": exports.prtUsdcAqFarm,
    GBijunwxa4Ni3JmYC6q6zgaVhSUJU6hVX5qTyJDRpNTc: exports.jetUsdcAqFarm,
    GtQ1NT7R5aaTiST7K6ZWdMhwDdFxsSFvVFhBo8vyHGAq: exports.stsolUsdcAqFarm,
    "5a6Y1ephcbKSoyLMQyD1JWbtqawCy8p2FtRL9v3zhaG5": exports.wstethUsdcAqFarm,
    "6mJqqT5TMgveDvxzBt3hrjGkPV5VAj7tacxFCT3GebXh": exports.auryUsdcAqFarm,
    Hmfrtmo93DpSDmVNLQKcBS5D1ia5JatiRSok9ososubz: exports.avaxUsdcAqFarm,
    FwCombynV2fTVizxPCNA2oZKoWXLZgdJThjE4Xv9sjxc: exports.fttUsdcAqFarm,
    "4cXw2MYj94TFBXLL73fEpMCr8DPrW68JvrV8mzWgktbD": exports.rayUsdcAqFarm,
    F59gkD7NnsdJbFKrRZsiBC8PAooN4c56T8QmahfW1iXN: exports.slndUsdcAqFarm,
    "7vnps4VE5RTGAr5fmPZu7fSrk2VnM4Up838grZfqmxqE": exports.gofxUsdcAqFarm,
    "9EjcYfHcG8f1mccpHyaAwpoxaUPiheC6KgLQjyD9aTb6": exports.woofUsdcAqFarm,
    CHTKUJGYRtBDqnxCFjxe5KEkZgxV98udbhuYYyzGxup5: exports.sdogeUsdcAqFarm,
    "55r9txzQtmjTykmTXmBYZCVMg5z9squB8b5cSw2AhxA4": exports.catoUsdcAqFarm,
    DSiHyHDn96bUQSZtizyCRLcQzrwohZeMpVu8rYJN1HzG: exports.oogiUsdcAqFarm,
    "5MvQHx8eftU39JTucFsT315JFnQASuDQg3FqxTw7xcvN": exports.sonarUsdcAqFarm,
    HNrYngS1eoqkjWro9D3Y5Z9sWBDzPNK2tX4rfV2Up177: exports.aptUsdcAqFarm,
    AWrtTWG4Zgxw8D92bb3L3sQtGLD3zDztMPWsXSph8iBP: exports.dflUsdcAqFarm,
    "9Y1vPaAsMz8X65DebMMnmBjbMo8i4jh4mcgiggZUUS3M": exports.dflSolAqFarm,
    FnDxJPNk7pPmGHUbR4XUHmHevrkXHdna5D3sQKcAtjBL: exports.frktUsdcAqFarm,
    FGgP1npQTsC5Q4xBmQtNYSh51NKqNwdxBZy8JCo3igcu: exports.tttUsdcAqFarm,
    E1U63VXhNiWoUkVvjrfLDdV1oJrwE6zLde3bohr6jCqz: exports.upsUsdcAqFarm,
    GjG7JjTQfQpDxw4hWx4etP9oTaYCuCbPjsU8WaUT3xHB: exports.fantUsdcAqFarm,
    D8WjqtwC9CzBrQKfSf2ccCHFQuPYwyLv5KAy8WjT5vnf: exports.blockUsdcAqFarm,
    "34Ppq6R8NfYBwWwPY4cBK4Afyb8hHaASQFukCzH6cV4n": exports.runUsdcAqFarm,
    HjR8JgqNKQVMvdryqJw5RJ4PCE9WGk8sgbEF7S9S3obv: exports.uxpUsdcAqFarm,
    J3kvcay3N16FBdawgnqoJ9v9p6XCvyCLE2Z9F5RLvGkj: exports.btcUsdcAqFarm,
    "12Uj74zgUUoBe4yeackwQ4qYtFMr9fk1xL6q5Nha6t2N": exports.mndeUsdcAqFarm,
    "71CBZeJ4tw38L9pSPoCz4fRsuWE64Fipyzotte7haoCS": exports.chicksUsdcAqFarm,
    "6MF5CHWAj5mS7FhpxiKz37CzR2eYTu236XpBKKMXCrGg": exports.onesolUsdcAqFarm,
    HDgxKmiA8Pv82fNguhVeMkZqQkos2YksFPoP1KttWxX8: exports.wmpUsdcAqFarm,
    "2VuGzaMrDnDyZfYvDwSXk38s7M2wpud7LDY3dGA1J9sy": exports.unqUsdcAqFarm,
    GoaAiajubRgeCFEz9L6mLnSmT2QFegoJDH5tpLfivpj: exports.basisUsdcAqFarm,
    E6FUnQHGHJVJg7oExVr5Moeaj1QpdpZQF5odYjHXWPZb: exports.gstUsdcAqFarm,
    F5BTnwuMA6rxftTdbZ33VWKr2wrr6DuQHnd4guKmPSYQ: exports.meanUsdcAqFarm,
    HCtyJzFUtYecXrA52s4Y9atq4J1fhT3cYsTX17XVSFag: exports.aartUsdcAqFarm,
    DJqqvzSuPaWThfzwMjXx7H2ZmHDdwxza6NtFudtuXcpc: exports.shdwUsdcAqFarm,
    "2ws7g3LBPdctfKn42Di9qxzQtUJ8ZL1aEAX2rGEQMNqh": exports.shdwSolAqFarm,
    "99ZHUQsgxL7K6PHrGNi1gSwawwPr7UA5fbWrYoHQ6qhX": exports.scyUsdcAqFarm,
    E5kSBqTDxFLbLNQaVVtPtnhEYVLMCK2fVSEKoMKL98qR: exports.slcUsdcAqFarm,
    "6c13xsmyk7UaHUWZ2rm1MM3ZdrQRSBkQ9waaG25ridVs": exports.wustSolAqFarm,
    J1KfRtP5y2warpD7LdJhfBLPKoWwSqYuovdArSv1mpQ7: exports.wustUsdcAqFarm,
    "68YVjgPnTUPcBqZyghqvD2WPNsrLKsjYTmBKJzHRr4qd": exports.msolWustAqFarm,
    "8Mh7drLbt3jFJYwp948XyvQscGLaLkChNcaH5wwaAoWA": exports.wlunaWustAqFarm,
    HTZd53fYwYQRyAjiaPsZy9Gf41gobFdqkF4oKe3XLi95: exports.stsolWustAqFarm,
    AzEoVuNJyo9ByoLRZ5t6vav2Zg24vULNVJM41PgCKUqR: exports.jsolUsdcAqFarm,
    CCyDxjdW3G7hPTthTMPTZ4bnhFF19XG6rx2fNiKeRQww: exports.daosolUsdcAqFarm,
    Gx4PoxenyQwhGGnKagAT35iVg4im1iKhJxDWqVhgu6tk: exports.orcaUsdtAqFarm,
    GsfyYHkSgC3Ta6aWR9MjB2sxoBrkGGeR2tAwXbpphf3: exports.orcaWhethAqFarm,
    "7cuu94swKL5PtFQohKMAzyd1mjj65rgMW3GzLY31HCnK": exports.geneUsdcAqFarm,
    "85krvT9DxdYgoFLQDHTAGdvtNuLdAsc4xE5FkVLpN2aR": exports.cmfiUsdcAqFarm,
    HVLyX8mD8YvKgZJ4oB6rXJiCYMLpHKwB6iCiCjE1XwdT: exports.celoUsdcAqFarm,
    Gpzd833qSmv3kXpQmxEaqkrZTXZaRjhNAoqhf61qAhTG: exports.ftmUsdcAqFarm,
    DFpLFcQZqDKykyDePgip4r6MExVmBKWqTa12ezq6qxUY: exports.btcOrcaAqFarm,
    cL5WhffCYFRLM4We8VS2W684kM4pHyuvEDwp8Ddw48k: exports.hbbUsdcAqFarm,
    FkKzu2HeMJZf4oHwoYPxLGVy3net5Jq8HAfnA5VqETgk: exports.hbbSolAqFarm,
    "2Reqt4Sw9xNY8BoJ3EZLpFu5yVgNxFrbw8M3KiJpPn6o": exports.sbUsdcAqFarm,
    "4ni1nho89cDKAQ9ddbNQA9ieLYpzvJVmJpuogu5Ct5ur": exports.stsolUsdtAqFarm,
    H7gyTmNCDXkD8MGMqnxqoD8ANszjcju4tjT6ERZ5dakf: exports.seededUsdcAqFarm,
    "3hksYA17VxgiKSeihjnZkBbjc2CTbEBfvDCYgQhojTo5": exports.audioUsdcAqFarm,
    AaZRnJAnDyJyPD9uPJpJ8bzBGDCEi6jtBpUf92xErWPp: exports.mmaUsdcAqFarm,
    "9wPhuYapychVDSxmXqCZxy2Ka8Lmav4SHM72si8bfraV": exports.onesolSolAqFarm,
    Eho8h1BcoG5QWU7X9FzJafw5ErKUXtR2LobAJJZfWff4: exports.puffSolAqFarm,
    "4iyU77yZbg8iD344vbwruAuDAf9i1EVV3FhZJDnWStBE": exports.saoUsdcAqFarm,
    "3dXdXg5HPyZ73GFC9LkSn3thdJUGeXWB8iSTHs5UcqiH": exports.srlyv2SolAqFarm,
    "2LYgm6nGXmSSjfoEriPuYeGoNiWNxUs7n3rnTbDWN5c7": exports.zbcUsdcAqFarm,
    CFxQF5kNAtbbDj298Xr47Sf4mkSyuzWpRH97hrdQ6kxi: exports.gmtUsdcAqFarm,
    DdAFNvDxtEHCgKj3JAM64zCKfMWhs4J9wEmRrjUAFiME: exports.novaUsdcAqFarm,
    "7n2YW9qLkhGFArdZPLoF4hYPE2zw7xCACkVPXrUWnLuo": exports.hbbUsdhAqFarm,
    "6atKbS2Xz5vu7cqWBNk8KYkuakRzckZ9nvtUKf2k8Sc3": exports.takiSrlyv2AqFarm,
    "5vhh9ZnD9vnahRhFLP1EUEyYRSzvJwgw9U2xygsSJSrp": exports.zigUsdcAqFarm,
    HfRgvhgscGX5GaP3rUrZAhh7gS4aJ2UQ7rNVX976rG6P: exports.srlySolAqFarm,
    "3xaK5aWWLNRB73xUVX3cusLhDp65mTvP4fwW5Jwxakgs": exports.takiSrlyAqFarm,
    "4jjQSgFx33DUb1a7pgPsi3FbtZXDQ94b6QywjNK3NtZw": exports.stsolSolAqFarm,
    HYp9v7cY4wAxSsa6ijztQQ3GQ8iTttuG5vu8JNBDHoNh: exports.btcStsolAqFarm,
    FWurFWADrgmhb6Y44LEaCMnEHS2Tu3QGqd9oBcZtr8gT: exports.stsolWhethAqFarm,
    "57vGdcMZLnbNr4TZ4hgrpGJZGR9vTPhu8L9bNKDrqxKT": exports.liqUsdcDoubleDip,
    GwrBA1F8rGummDCDd8NY9Eu1cLNuJqbT8WaGxgWpFwGL: exports.stepSolDoubleDip,
    "66xCxkffQZKBZLiHV3PDcfR8ANJTfnDRxPCaBdv4wxB7": exports.slrsUsdcDoubleDip,
    "4CGxvZdwiZgVMLXiTdJHTkJRUTpTSJCtmtCRbSkAxerE": exports.portUsdcDoubleDip,
    "9SDpBrfqNxjXcCzpKWM6yUKdfky975VJBD6xcu5cKf5s": exports.copeUsdcDoubleDip,
    A7vvbqENJj8kED3ABjphe8TvwpasQYtoWGKpjpLArMxa: exports.bopUsdcDoubleDip,
    "9voVuTq1S9bFZkF2Jo44HoVG63w2xDRT8eBzB23YbQud": exports.samoUsdcDoubleDip,
    Bfoi3RNnfdP5VeRGqvTA8MRN9ePGJoZgeKfe8WeBHUxE: exports.whapiUsdcDoubleDip,
    "3K7aZhtwWJ2JS6GnbbgeDVnxd1q2hwhqasmgRsAMZ4yC": exports.slimUsdcDoubleDip,
    "7YyhptkxY81HPzFVfyCzA5UXxWdsNRD41ofLva3TuSpd": exports.ninjaSolDoubleDip,
    HFmY1ggCsCky1zJ1sfdkNR4zb3u5n38YNRdf4vsGu17t: exports.atlasUsdcDoubleDip,
    "63JUKLnCAuNMPSPioEgbjjzp9Qk8qSEEM8eZqEtPqfLU": exports.polisUsdcDoubleDip,
    "5uR5STASUmoGVHzqMeut98t26TfVkQqWU9f9dsv3NfJ6": exports.abrUsdcDoubleDip,
    "6PGoaQdL9e463hdaFxHXsuPcjCHRK32CQ9PFKxvM7XY2": exports.kuroUsdcDoubleDip,
    "5r3vDsNTGXXb9cGQfqyNuYD2bjhRPymGJBfDmKosR9Ev": exports.msolUsdcDoubleDip,
    "3Duk5b6fLztPmS4ryV48FM1Q9WXUSMwz9jehAT4UtqpE": exports.orcaMsolDoubleDip,
    "3RTGL7gPF4V1ns1AeGFApT7cBEGVDfmJ77DqQi9AC6uG": exports.msolSolDoubleDip,
    "3kT3oYuS1rCfhmqfgy6EKcbZdaJimaVEjoy25QiuEaoj": exports.stsolWstethDoubleDip,
    Ds4VGZhZzS2PMFzhzKeC3mwcQjdiCG21R76fTVbsSJyJ: exports.sypUsdcDoubleDip,
    "2wPsMuzhEsC6GhV3qtFpmJF6atEgLGbnmQ8U43Y6fPxZ": exports.mndeMsolDoubleDip,
    Afvh7TWfcT1E9eEEWJk17fPjnqk36hreTJJK5g3s4fm8: exports.msolUsdtDoubleDip,
    "58nifjPjF3CutGz2xMxvAMk7R9YgbVEc8Cstj4rCcs8j": exports.msolWhethDoubleDip,
    DzpLz78wuwyFsQToin8iDv6YK6aBEymRqQq82swiFh7r: exports.btcMsolDoubleDip,
    HqajzzbGMST3yCCVBJuXvNVsWkY2DXqiBz9cTRmmyBMy: exports.ivnSolDoubleDip,
    DNAGfa7tK8csprRQmiDUwDaFfhw6ueHhVFHTCgTJ8HGs: exports.larixUsdcDoubleDip,
    B95rdqSY4dqPwmt295XwBZZqZJYLmqDNXU6NvBpT4ep4: exports.gofxUsdcDoubleDip,
    "4HaQXDz9gdLgKUjvNVtnLyNZoWNYKjh3XxH1TpLgiwmi": exports.woofUsdcDoubleDip,
    "4kXYyZAMBdvgDaBKUGvRWJnHw9af7sCUPvpQ68PEdP8b": exports.sdogeUsdcDoubleDip,
    BHtZnTBMeY4EBEW5egGnuK5bdW12v6Dod6wFav79AyYx: exports.catoUsdcDoubleDip,
    FiwSk36yi1DNWcuQUeNipAc1VKxa9Wv9AR2xFvyKUxAE: exports.oogiUsdcDoubleDip,
    GWmwwMGYBG4NqYdnsYrudzBnbgDC49MkBxdzhfLA9kVY: exports.sonarUsdcDoubleDip,
    "9fuv3emLQXECrTWqm2HaKT3wQhmvmgqmReZSMcu8PfpH": exports.upsUsdcDoubleDip,
    C2EcthTMaC5eATXVaXg5ctvMfUYYgFyNibybHes5D3S6: exports.fantUsdcDoubleDip,
    "3u2dNfGuU6C3vmSg5EvLPUpX57b3niqhWBV5Gc3WDEf5": exports.stsolUsdcDoubleDip,
    Ce3VXSQGFEHqrQSEzcVb2Ro88Mcn56cYBUAXFx5tL9bo: exports.wmpUsdcDoubleDip,
    GTcxXu3yw7yotq2UtAMvpEbx84ozJyrBwy77Cmem7m4X: exports.chicksUsdcDoubleDip,
    CsGTrgJ6oLx9UQow9aLjuAacjXTY53zT8B2FnAbfVZAS: exports.unqUsdcDoubleDip,
    "3h2VBX8533NB8eEH8rPXMdayodFDbgTHpbav6JqLZAQq": exports.meanUsdcDoubleDip,
    "8Wu5sJpERA1J5iWcT8aMpt9cTAfKDLPbLpGjNsJoPgLc": exports.wagUsdcDoubleDip,
    "7WWHfufv8vuBC1x7GXA3pu7kgNhEQkXoq3CtbaQihAJ9": exports.shdwUsdcDoubleDip,
    BDStVBt4NS5bfda25ubK51kVRioV4yjKKCPbe96jeEms: exports.shdwSolDoubleDip,
    Bg7pZq7KdsQsnCQBeNpa4XVnrWfjUJNu3ViHoNB7YUZU: exports.aartUsdcDoubleDip,
    "4yx2aHMa7N4m1uUaBRy9QPtpstw3HFPtvcCPJQaGFHKL": exports.basisUsdcDoubleDip,
    X8GnAvxq942xXjNzqYPFQQ9JstJQNEA81uPeNoQpgrN: exports.slcUsdcDoubleDip,
    F49Cm3srGucQCBanA2xL7nSKHfH1QZd9vLdFtkg4LKnq: exports.wustSolDoubleDip,
    "2mhVUMsG7eb3XhHjAbKpRZgWNyGZNCiWU7dRxauzZaGL": exports.wustUsdcDoubleDip,
    "6wqME6zPQzGDLugpnwZnVBKBbATC5nTaEa781Vj98yvM": exports.msolWustDoubleDip,
    "3ZGKT28NXAqb2YtUAMvKixQvHNHT31Q5mK8AC2iBMs29": exports.wlunaWustDoubleDip,
    "3Q44iV4URXdbS4Tk1PGs5VdWQoCxHB7zdcdMnemo8jfH": exports.stsolWustDoubleDip,
    E8RVjS24pBuF3oCCeJVAgC4RQ7mVa5P3FGXqGbiczzvD: exports.cmfiUsdcDoubleDip,
    BxPd4x7gm6WQF6jsDCRQuQVZ2bopQ4KegMrqyQgdkPDk: exports.hbbUsdcDoubleDip,
    "9RRBDWZGWGVnHgazCpb9R1XApmHLBDya8Tq1yrzoMsKY": exports.celoUsdcDoubleDip,
    "3Wppx86xN7Dg7GLUTD9C7AqCt68qZkWzNYUdCS7t1pCk": exports.sbUsdcDoubleDip,
    E7Af9Fa2U1YqEkQpAYjXwDB5TJUou9VcN3ot33Gj6UY9: exports.ftmUsdcDoubleDip,
    CtvKaRLzCzRZCcYPwai7NCYBK4ArBj2oD6BfvMgJoiCN: exports.stsolUsdtDoubleDip,
    H15WptGntFQifZmJHUzYBV9Mv7P27ofavEsF6yqpLTdX: exports.seededUsdcDoubleDip,
    Hva9oLa2GjoKdB45WoHujsX7MTpehByPshMrQpNDmkFq: exports.onesolSolDoubleDip,
    H9qkbU2XYSQTk6JJJ2TMuZBthfkeCHigmuxB5jiQaamn: exports.mmaUsdcDoubleDip,
    Ge5kuYg5PekrPUeKzngw97Cnfngj8j6NCX5q1jTBkWSW: exports.puffSolDoubleDip,
    A9BeGSRJJYXPrMs81rVZxvkk16fopzgG5YkngntgTu7p: exports.saoUsdcDoubleDip,
    "6gkZ7QUmxmwPLS2NK3Dr6YHtTPZs6GQrkA595WSx5iLe": exports.srlyv2SolDoubleDip,
    "7Hoi4adCSBzERdvSiUXtVDz79tiDxMD5HpDv7m9rs3Sb": exports.zbcUsdcDoubleDip,
    H2jZrZKx1fLY9LPi2ad5G2DVBWzMmUwPQLjRFERUroEH: exports.novaUsdcDoubleDip,
    DUGuXcpHX9y6WdTKgDm8XsECbf7pApA7ZvAsjUqHUoYn: exports.hbbUsdhDoubleDip,
    "9nKbvVW9TjdnKt2NbfMaf9PHcg8zWWFgGTBfzUXdFcke": exports.zigUsdcDoubleDip,
    HfkZkoPqZzd7S1F3BhkJN2Es1C5vabVYqfCecXG2MGHP: exports.srlySolDoubleDip,
    "13HwDUH7yREGAwHSWnx4Kgnn6huL8yfCCBqaBr6NGVH6": exports.takiSrlyDoubleDip,
    "7ZvhrqdnP2NrQi2YCjQNd8jKJpLDMUuZg4wP7XJPjnVX": exports.stsolSolDoubleDip,
    GjBW6751wNJ99zYkr9QwSi98bvj1NgyAcYHfQDpemJfq: exports.btcStsolDoubleDip,
    "9V9CavjccKJH66m4MJHzVBYzY2ZM3U3D5xUCP7HSyAUz": exports.stsolWhethDoubleDip,
};
