// @orca-so/sdk/dist/constants/tokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.novaToken = exports.ninjaToken = exports.mngoToken = exports.mndeToken = exports.mmaToken = exports.merToken = exports.mediaToken = exports.meanToken = exports.mapsToken = exports.liqToken = exports.larixToken = exports.kuroToken = exports.kinToken = exports.jsolToken = exports.jetToken = exports.ivnToken = exports.hdgToken = exports.hbbToken = exports.gstToken = exports.grapeToken = exports.gofxToken = exports.gmtToken = exports.geneToken = exports.fttToken = exports.ftmToken = exports.frktToken = exports.fidaToken = exports.fantToken = exports.fabToken = exports.ethToken = exports.dflToken = exports.copeToken = exports.cmfiToken = exports.chicksToken = exports.celoToken = exports.caveToken = exports.catoToken = exports.btcToken = exports.bopToken = exports.blockToken = exports.basisToken = exports.avaxToken = exports.auryToken = exports.audioToken = exports.atlasToken = exports.aptToken = exports.anaToken = exports.abrToken = exports.aartToken = exports.onesolToken = void 0;
exports.srlyToken = exports.srlyv2Token = exports.psolToken = exports.msolToken = exports.daosolToken = exports.zigToken = exports.zbcToken = exports.woofToken = exports.wmpToken = exports.wagToken = exports.uxpToken = exports.ushToken = exports.usdtToken = exports.usdhToken = exports.usdcToken = exports.upsToken = exports.unqToken = exports.tulipToken = exports.tttToken = exports.takiToken = exports.sypToken = exports.sunnyToken = exports.stepToken = exports.srmToken = exports.sonarToken = exports.solToken = exports.snyToken = exports.slrsToken = exports.slndToken = exports.slimToken = exports.slcToken = exports.shdwToken = exports.seededToken = exports.sdogeToken = exports.scyToken = exports.sbrToken = exports.sbToken = exports.saoToken = exports.samoToken = exports.runToken = exports.ropeToken = exports.rayToken = exports.puffToken = exports.prtToken = exports.portToken = exports.polisToken = exports.paiToken = exports.oxyToken = exports.orcaToken = exports.oogiToken = void 0;
exports.wstethToken = exports.whethToken = exports.wustToken = exports.wlunaToken = exports.wldoToken = exports.whapiToken = exports.stsolToken = exports.scnsolToken = void 0;
const web3_js_1 = require("@solana/web3.js");
/**
 * The following content is auto-generated.
 */
exports.onesolToken = Object.freeze({
    tag: "1SOL",
    name: "1Sol",
    mint: new web3_js_1.PublicKey("4ThReWAbAVZjNVgs5Ui9Pk3cZ5TYaD9u6Y89fp6EFzoF"),
    scale: 8,
});
exports.aartToken = Object.freeze({
    tag: "AART",
    name: "ALL.ART",
    mint: new web3_js_1.PublicKey("F3nefJBcejYbtdREjui1T9DPh5dBgpkKq7u2GAAMXs5B"),
    scale: 6,
});
exports.abrToken = Object.freeze({
    tag: "ABR",
    name: "Allbridge",
    mint: new web3_js_1.PublicKey("a11bdAAuV8iB2fu7X6AxAvDTo1QZ8FXB3kk5eecdasp"),
    scale: 9,
});
exports.anaToken = Object.freeze({
    tag: "ANA",
    name: "Nirvana ANA",
    mint: new web3_js_1.PublicKey("ANAxByE6G2WjFp7A4NqtWYXb3mgruyzZYg3spfxe6Lbo"),
    scale: 6,
});
exports.aptToken = Object.freeze({
    tag: "APT",
    name: "Apricot",
    mint: new web3_js_1.PublicKey("APTtJyaRX5yGTsJU522N4VYWg3vCvSb65eam5GrPT5Rt"),
    scale: 6,
});
exports.atlasToken = Object.freeze({
    tag: "ATLAS",
    name: "Star Atlas",
    mint: new web3_js_1.PublicKey("ATLASXmbPQxBUYbxPsV97usA3fPQYEqzQBUHgiFCUsXx"),
    scale: 8,
});
exports.audioToken = Object.freeze({
    tag: "AUDIO",
    name: "Audius (Wormhole)",
    mint: new web3_js_1.PublicKey("9LzCMqDgTKYz9Drzqnpgee3SGa89up3a247ypMj2xrqM"),
    scale: 8,
});
exports.auryToken = Object.freeze({
    tag: "AURY",
    name: "Aurory",
    mint: new web3_js_1.PublicKey("AURYydfxJib1ZkTir1Jn1J9ECYUtjb6rKQVmtYaixWPP"),
    scale: 9,
});
exports.avaxToken = Object.freeze({
    tag: "AVAX",
    name: "Avalanche",
    mint: new web3_js_1.PublicKey("AUrMpCDYYcPuHhyNX8gEEqbmDPFUpBpHrNW3vPeCFn5Z"),
    scale: 9,
});
exports.basisToken = Object.freeze({
    tag: "BASIS",
    name: "basis.markets",
    mint: new web3_js_1.PublicKey("Basis9oJw9j8cw53oMV7iqsgo6ihi9ALw4QR31rcjUJa"),
    scale: 6,
});
exports.blockToken = Object.freeze({
    tag: "BLOCK",
    name: "Blockasset",
    mint: new web3_js_1.PublicKey("NFTUkR4u7wKxy9QLaX2TGvd9oZSWoMo4jqSJqdMb7Nk"),
    scale: 6,
});
exports.bopToken = Object.freeze({
    tag: "BOP",
    name: "Boring Protocol",
    mint: new web3_js_1.PublicKey("BLwTnYKqf7u4qjgZrrsKeNs2EzWkMLqVCu6j8iHyrNA3"),
    scale: 8,
});
exports.btcToken = Object.freeze({
    tag: "BTC",
    name: "Bitcoin",
    mint: new web3_js_1.PublicKey("9n4nbM75f5Ui33ZbPYXn59EwSgE8CGsHtAeTH5YFeJ9E"),
    scale: 6,
});
exports.catoToken = Object.freeze({
    tag: "CATO",
    name: "CATO",
    mint: new web3_js_1.PublicKey("5p2zjqCd1WJzAVgcEnjhb9zWDU7b9XVhFhx4usiyN7jB"),
    scale: 9,
});
exports.caveToken = Object.freeze({
    tag: "CAVE",
    name: "Crypto Cavemen",
    mint: new web3_js_1.PublicKey("4SZjjNABoqhbd4hnapbvoEPEqT8mnNkfbEoAwALf1V8t"),
    scale: 6,
});
exports.celoToken = Object.freeze({
    tag: "CELO",
    name: "Celo",
    mint: new web3_js_1.PublicKey("GNCjk3FmPPgZTkbQRSxr6nCvLtYMbXKMnRxg8BgJs62e"),
    scale: 9,
});
exports.chicksToken = Object.freeze({
    tag: "CHICKS",
    name: "SolChicks Token",
    mint: new web3_js_1.PublicKey("cxxShYRVcepDudXhe7U62QHvw8uBJoKFifmzggGKVC2"),
    scale: 9,
});
exports.cmfiToken = Object.freeze({
    tag: "CMFI",
    name: "Compendium.Fi",
    mint: new web3_js_1.PublicKey("5Wsd311hY8NXQhkt9cWHwTnqafk7BGEbLu8Py3DSnPAr"),
    scale: 6,
});
exports.copeToken = Object.freeze({
    tag: "COPE",
    name: "Cope",
    mint: new web3_js_1.PublicKey("8HGyAAB1yoM1ttS7pXjHMa3dukTFGQggnFFH3hJZgzQh"),
    scale: 6,
});
exports.dflToken = Object.freeze({
    tag: "DFL",
    name: "DeFi Land",
    mint: new web3_js_1.PublicKey("DFL1zNkaGPWm1BqAVqRjCZvHmwTFrEaJtbzJWgseoNJh"),
    scale: 9,
});
exports.ethToken = Object.freeze({
    tag: "ETH",
    name: "Ethereum",
    mint: new web3_js_1.PublicKey("2FPyTwcZLUg1MDrwsyoP4D6s1tM7hAkHYRjkNb5w6Pxk"),
    scale: 6,
});
exports.fabToken = Object.freeze({
    tag: "FAB",
    name: "Fabric",
    mint: new web3_js_1.PublicKey("EdAhkbj5nF9sRM7XN7ewuW8C9XEUMs8P7cnoQ57SYE96"),
    scale: 9,
});
exports.fantToken = Object.freeze({
    tag: "FANT",
    name: "Phantasia",
    mint: new web3_js_1.PublicKey("FANTafPFBAt93BNJVpdu25pGPmca3RfwdsDsRrT3LX1r"),
    scale: 6,
});
exports.fidaToken = Object.freeze({
    tag: "FIDA",
    name: "Bonfida",
    mint: new web3_js_1.PublicKey("EchesyfXePKdLtoiZSL8pBe8Myagyy8ZRqsACNCFGnvp"),
    scale: 6,
});
exports.frktToken = Object.freeze({
    tag: "FRKT",
    name: "FRAKT Token",
    mint: new web3_js_1.PublicKey("ErGB9xa24Szxbk1M28u2Tx8rKPqzL6BroNkkzk5rG4zj"),
    scale: 8,
});
exports.ftmToken = Object.freeze({
    tag: "FTM",
    name: "Fantom",
    mint: new web3_js_1.PublicKey("EsPKhGTMf3bGoy4Qm7pCv3UCcWqAmbC1UGHBTDxRjjD4"),
    scale: 9,
});
exports.fttToken = Object.freeze({
    tag: "FTT",
    name: "FTX Token",
    mint: new web3_js_1.PublicKey("AGFEad2et2ZJif9jaGpdMixQqvW5i81aBdvKe7PHNfz3"),
    scale: 6,
});
exports.geneToken = Object.freeze({
    tag: "GENE",
    name: "Genopets",
    mint: new web3_js_1.PublicKey("GENEtH5amGSi8kHAtQoezp1XEXwZJ8vcuePYnXdKrMYz"),
    scale: 9,
});
exports.gmtToken = Object.freeze({
    tag: "GMT",
    name: "STEPN",
    mint: new web3_js_1.PublicKey("7i5KKsX2weiTkry7jA4ZwSuXGhs5eJBEjY8vVxR4pfRx"),
    scale: 9,
});
exports.gofxToken = Object.freeze({
    tag: "GOFX",
    name: "GooseFX",
    mint: new web3_js_1.PublicKey("GFX1ZjR2P15tmrSwow6FjyDYcEkoFb4p4gJCpLBjaxHD"),
    scale: 9,
});
exports.grapeToken = Object.freeze({
    tag: "GRAPE",
    name: "Grape",
    mint: new web3_js_1.PublicKey("8upjSpvjcdpuzhfR1zriwg5NXkwDruejqNE9WNbPRtyA"),
    scale: 6,
});
exports.gstToken = Object.freeze({
    tag: "GST",
    name: "GST",
    mint: new web3_js_1.PublicKey("AFbX8oGjGpmVFywbVouvhQSRmiW2aR1mohfahi4Y2AdB"),
    scale: 9,
});
exports.hbbToken = Object.freeze({
    tag: "HBB",
    name: "Hubble",
    mint: new web3_js_1.PublicKey("HBB111SCo9jkCejsZfz8Ec8nH7T6THF8KEKSnvwT6XK6"),
    scale: 6,
});
exports.hdgToken = Object.freeze({
    tag: "HDG",
    name: "Hedge Token",
    mint: new web3_js_1.PublicKey("5PmpMzWjraf3kSsGEKtqdUsCoLhptg4yriZ17LKKdBBy"),
    scale: 9,
});
exports.ivnToken = Object.freeze({
    tag: "IVN",
    name: "Investin",
    mint: new web3_js_1.PublicKey("iVNcrNE9BRZBC9Aqf753iZiZfbszeAVUoikgT9yvr2a"),
    scale: 6,
});
exports.jetToken = Object.freeze({
    tag: "JET",
    name: "JET",
    mint: new web3_js_1.PublicKey("JET6zMJWkCN9tpRT2v2jfAmm5VnQFDpUBCyaKojmGtz"),
    scale: 9,
});
exports.jsolToken = Object.freeze({
    tag: "JSOL",
    name: "JPool",
    mint: new web3_js_1.PublicKey("7Q2afV64in6N6SeZsAAB81TJzwDoD6zpqmHkzi9Dcavn"),
    scale: 9,
});
exports.kinToken = Object.freeze({
    tag: "KIN",
    name: "Kin",
    mint: new web3_js_1.PublicKey("kinXdEcpDQeHPEuQnqmUgtYykqKGVFq6CeVX5iAHJq6"),
    scale: 5,
});
exports.kuroToken = Object.freeze({
    tag: "KURO",
    name: "Kurobi",
    mint: new web3_js_1.PublicKey("2Kc38rfQ49DFaKHQaWbijkE7fcymUMLY5guUiUsDmFfn"),
    scale: 6,
});
exports.larixToken = Object.freeze({
    tag: "LARIX",
    name: "Larix",
    mint: new web3_js_1.PublicKey("Lrxqnh6ZHKbGy3dcrCED43nsoLkM1LTzU2jRfWe8qUC"),
    scale: 6,
});
exports.liqToken = Object.freeze({
    tag: "LIQ",
    name: "LIQ Protocol",
    mint: new web3_js_1.PublicKey("4wjPQJ6PrkC4dHhYghwJzGBVP78DkBzA2U3kHoFNBuhj"),
    scale: 6,
});
exports.mapsToken = Object.freeze({
    tag: "MAPS",
    name: "MAPS",
    mint: new web3_js_1.PublicKey("MAPS41MDahZ9QdKXhVa4dWB9RuyfV4XqhyAZ8XcYepb"),
    scale: 6,
});
exports.meanToken = Object.freeze({
    tag: "MEAN",
    name: "Meanfi",
    mint: new web3_js_1.PublicKey("MEANeD3XDdUmNMsRGjASkSWdC8prLYsoRJ61pPeHctD"),
    scale: 6,
});
exports.mediaToken = Object.freeze({
    tag: "MEDIA",
    name: "Media Network",
    mint: new web3_js_1.PublicKey("ETAtLmCmsoiEEKfNrHKJ2kYy3MoABhU6NQvpSfij5tDs"),
    scale: 6,
});
exports.merToken = Object.freeze({
    tag: "MER",
    name: "Mercurial",
    mint: new web3_js_1.PublicKey("MERt85fc5boKw3BW1eYdxonEuJNvXbiMbs6hvheau5K"),
    scale: 6,
});
exports.mmaToken = Object.freeze({
    tag: "MMA",
    name: "MMA Gaming",
    mint: new web3_js_1.PublicKey("MMAx26JtJgSWv6yH48nEHCGZcVvRbf9Lt9ALa7jSipe"),
    scale: 9,
});
exports.mndeToken = Object.freeze({
    tag: "MNDE",
    name: "Marinade Governance",
    mint: new web3_js_1.PublicKey("MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey"),
    scale: 9,
});
exports.mngoToken = Object.freeze({
    tag: "MNGO",
    name: "Mango Markets",
    mint: new web3_js_1.PublicKey("MangoCzJ36AjZyKwVj3VnYU4GTonjfVEnJmvvWaxLac"),
    scale: 6,
});
exports.ninjaToken = Object.freeze({
    tag: "NINJA",
    name: "NINJA",
    mint: new web3_js_1.PublicKey("FgX1WD9WzMU3yLwXaFSarPfkgzjLb2DZCqmkx9ExpuvJ"),
    scale: 6,
});
exports.novaToken = Object.freeze({
    tag: "NOVA",
    name: "Nova Finance",
    mint: new web3_js_1.PublicKey("BDrL8huis6S5tpmozaAaT5zhE5A7ZBAB2jMMvpKEeF8A"),
    scale: 9,
});
exports.oogiToken = Object.freeze({
    tag: "OOGI",
    name: "OOGI",
    mint: new web3_js_1.PublicKey("H7Qc9APCWWGDVxGD5fJHmLTmdEgT9GFatAKFNg6sHh8A"),
    scale: 9,
});
exports.orcaToken = Object.freeze({
    tag: "ORCA",
    name: "Orca",
    mint: new web3_js_1.PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    scale: 6,
});
exports.oxyToken = Object.freeze({
    tag: "OXY",
    name: "Oxygen",
    mint: new web3_js_1.PublicKey("z3dn17yLaGMKffVogeFHQ9zWVcXgqgf3PQnDsNs2g6M"),
    scale: 6,
});
exports.paiToken = Object.freeze({
    tag: "PAI",
    name: "Parrot Stable",
    mint: new web3_js_1.PublicKey("Ea5SjE2Y6yvCeW5dYTn7PYMuW5ikXkvbGdcmSnXeaLjS"),
    scale: 6,
});
exports.polisToken = Object.freeze({
    tag: "POLIS",
    name: "Star Atlas DAO",
    mint: new web3_js_1.PublicKey("poLisWXnNRwC6oBu1vHiuKQzFjGL4XDSu4g9qjz9qVk"),
    scale: 8,
});
exports.portToken = Object.freeze({
    tag: "PORT",
    name: "Port Finance",
    mint: new web3_js_1.PublicKey("PoRTjZMPXb9T7dyU7tpLEZRQj7e6ssfAE62j2oQuc6y"),
    scale: 6,
});
exports.prtToken = Object.freeze({
    tag: "PRT",
    name: "Parrot Protocol",
    mint: new web3_js_1.PublicKey("PRT88RkA4Kg5z7pKnezeNH4mafTvtQdfFgpQTGRjz44"),
    scale: 6,
});
exports.puffToken = Object.freeze({
    tag: "PUFF",
    name: "PUFF",
    mint: new web3_js_1.PublicKey("G9tt98aYSznRk7jWsfuz9FnTdokxS6Brohdo9hSmjTRB"),
    scale: 9,
});
exports.rayToken = Object.freeze({
    tag: "RAY",
    name: "Raydium",
    mint: new web3_js_1.PublicKey("4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R"),
    scale: 6,
});
exports.ropeToken = Object.freeze({
    tag: "ROPE",
    name: "Rope",
    mint: new web3_js_1.PublicKey("8PMHT4swUMtBzgHnh5U564N5sjPSiUz2cjEQzFnnP1Fo"),
    scale: 9,
});
exports.runToken = Object.freeze({
    tag: "RUN",
    name: "Run",
    mint: new web3_js_1.PublicKey("6F9XriABHfWhit6zmMUYAQBSy6XK5VF1cHXuW5LDpRtC"),
    scale: 9,
});
exports.samoToken = Object.freeze({
    tag: "SAMO",
    name: "Samoyedcoin",
    mint: new web3_js_1.PublicKey("7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU"),
    scale: 9,
});
exports.saoToken = Object.freeze({
    tag: "SAO",
    name: "Sator",
    mint: new web3_js_1.PublicKey("2HeykdKjzHKGm2LKHw8pDYwjKPiFEoXAz74dirhUgQvq"),
    scale: 9,
});
exports.sbToken = Object.freeze({
    tag: "SB",
    name: "SuperBonds",
    mint: new web3_js_1.PublicKey("SuperbZyz7TsSdSoFAZ6RYHfAWe9NmjXBLVQpS8hqdx"),
    scale: 6,
});
exports.sbrToken = Object.freeze({
    tag: "SBR",
    name: "Saber",
    mint: new web3_js_1.PublicKey("Saber2gLauYim4Mvftnrasomsv6NvAuncvMEZwcLpD1"),
    scale: 6,
});
exports.scyToken = Object.freeze({
    tag: "SCY",
    name: "Synchrony",
    mint: new web3_js_1.PublicKey("SCYfrGCw8aDiqdgcpdGjV6jp4UVVQLuphxTDLNWu36f"),
    scale: 9,
});
exports.sdogeToken = Object.freeze({
    tag: "SDOGE",
    name: "SolDoge",
    mint: new web3_js_1.PublicKey("8ymi88q5DtmdNTn2sPRNFkvMkszMHuLJ1e3RVdWjPa3s"),
    scale: 0,
});
exports.seededToken = Object.freeze({
    tag: "SEEDED",
    name: "Seeded Network",
    mint: new web3_js_1.PublicKey("seedEDBqu63tJ7PFqvcbwvThrYUkQeqT6NLf81kLibs"),
    scale: 9,
});
exports.shdwToken = Object.freeze({
    tag: "SHDW",
    name: "Shadow",
    mint: new web3_js_1.PublicKey("SHDWyBxihqiCj6YekG2GUr7wqKLeLAMK1gHZck9pL6y"),
    scale: 9,
});
exports.slcToken = Object.freeze({
    tag: "SLC",
    name: "Solice",
    mint: new web3_js_1.PublicKey("METAmTMXwdb8gYzyCPfXXFmZZw4rUsXX58PNsDg7zjL"),
    scale: 6,
});
exports.slimToken = Object.freeze({
    tag: "SLIM",
    name: "Solanium",
    mint: new web3_js_1.PublicKey("xxxxa1sKNGwFtw2kFn8XauW9xq8hBZ5kVtcSesTT9fW"),
    scale: 6,
});
exports.slndToken = Object.freeze({
    tag: "SLND",
    name: "Solend",
    mint: new web3_js_1.PublicKey("SLNDpmoWTVADgEdndyvWzroNL7zSi1dF9PC3xHGtPwp"),
    scale: 6,
});
exports.slrsToken = Object.freeze({
    tag: "SLRS",
    name: "Solrise Finance",
    mint: new web3_js_1.PublicKey("SLRSSpSLUTP7okbCUBYStWCo1vUgyt775faPqz8HUMr"),
    scale: 6,
});
exports.snyToken = Object.freeze({
    tag: "SNY",
    name: "SNY",
    mint: new web3_js_1.PublicKey("4dmKkXNHdgYsXqBHCuMikNQWwVomZURhYvkkX5c4pQ7y"),
    scale: 6,
});
exports.solToken = Object.freeze({
    tag: "SOL",
    name: "Solana",
    mint: new web3_js_1.PublicKey("So11111111111111111111111111111111111111112"),
    scale: 9,
});
exports.sonarToken = Object.freeze({
    tag: "SONAR",
    name: "SonarWatch",
    mint: new web3_js_1.PublicKey("sonarX4VtVkQemriJeLm6CKeW3GDMyiBnnAEMw1MRAE"),
    scale: 9,
});
exports.srmToken = Object.freeze({
    tag: "SRM",
    name: "Serum",
    mint: new web3_js_1.PublicKey("SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt"),
    scale: 6,
});
exports.stepToken = Object.freeze({
    tag: "STEP",
    name: "Step",
    mint: new web3_js_1.PublicKey("StepAscQoEioFxxWGnh2sLBDFp9d8rvKz2Yp39iDpyT"),
    scale: 9,
});
exports.sunnyToken = Object.freeze({
    tag: "SUNNY",
    name: "Sunny Aggregator",
    mint: new web3_js_1.PublicKey("SUNNYWgPQmFxe9wTZzNK7iPnJ3vYDrkgnxJRJm1s3ag"),
    scale: 6,
});
exports.sypToken = Object.freeze({
    tag: "SYP",
    name: "Sypool",
    mint: new web3_js_1.PublicKey("FnKE9n6aGjQoNWRBZXy4RW6LZVao7qwBonUbiD7edUmZ"),
    scale: 9,
});
exports.takiToken = Object.freeze({
    tag: "TAKI",
    name: "Taki",
    mint: new web3_js_1.PublicKey("Taki7fi3Zicv7Du1xNAWLaf6mRK7ikdn77HeGzgwvo4"),
    scale: 9,
});
exports.tttToken = Object.freeze({
    tag: "TTT",
    name: "TabTrader",
    mint: new web3_js_1.PublicKey("FNFKRV3V8DtA3gVJN6UshMiLGYA8izxFwkNWmJbFjmRj"),
    scale: 6,
});
exports.tulipToken = Object.freeze({
    tag: "TULIP",
    name: "Tulip Protocol",
    mint: new web3_js_1.PublicKey("TuLipcqtGVXP9XR62wM8WWCm6a9vhLs7T1uoWBk6FDs"),
    scale: 6,
});
exports.unqToken = Object.freeze({
    tag: "UNQ",
    name: "UNQ",
    mint: new web3_js_1.PublicKey("UNQtEecZ5Zb4gSSVHCAWUQEoNnSVEbWiKCi1v9kdUJJ"),
    scale: 6,
});
exports.upsToken = Object.freeze({
    tag: "UPS",
    name: "UPFI Network",
    mint: new web3_js_1.PublicKey("EwJN2GqUGXXzYmoAciwuABtorHczTA5LqbukKXV1viH7"),
    scale: 6,
});
exports.usdcToken = Object.freeze({
    tag: "USDC",
    name: "USD Coin",
    mint: new web3_js_1.PublicKey("EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"),
    scale: 6,
});
exports.usdhToken = Object.freeze({
    tag: "USDH",
    name: "USDH",
    mint: new web3_js_1.PublicKey("USDH1SM1ojwWUga67PGrgFWUHibbjqMvuMaDkRJTgkX"),
    scale: 6,
});
exports.usdtToken = Object.freeze({
    tag: "USDT",
    name: "Tether USD",
    mint: new web3_js_1.PublicKey("Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB"),
    scale: 6,
});
exports.ushToken = Object.freeze({
    tag: "USH",
    name: "Hedge USD",
    mint: new web3_js_1.PublicKey("9iLH8T7zoWhY7sBmj1WK9ENbWdS1nL8n9wAxaeRitTa6"),
    scale: 9,
});
exports.uxpToken = Object.freeze({
    tag: "UXP",
    name: "UXD Protocol Token",
    mint: new web3_js_1.PublicKey("UXPhBoR3qG4UCiGNJfV7MqhHyFqKN68g45GoYvAeL2M"),
    scale: 9,
});
exports.wagToken = Object.freeze({
    tag: "WAG",
    name: "Waggle",
    mint: new web3_js_1.PublicKey("5tN42n9vMi6ubp67Uy4NnmM5DMZYN8aS8GeB3bEDHr6E"),
    scale: 9,
});
exports.wmpToken = Object.freeze({
    tag: "WMP",
    name: "Whalemap",
    mint: new web3_js_1.PublicKey("BygDd5LURoqztD3xETc99WCxLUbTi6WYSht9XiBgZ4HW"),
    scale: 9,
});
exports.woofToken = Object.freeze({
    tag: "WOOF",
    name: "WOOF Token",
    mint: new web3_js_1.PublicKey("9nEqaUcb16sQ3Tn1psbkWqyhPdLmfHWjKGymREjsAgTE"),
    scale: 6,
});
exports.zbcToken = Object.freeze({
    tag: "ZBC",
    name: "Zebec Protocol",
    mint: new web3_js_1.PublicKey("zebeczgi5fSEtbpfQKVZKCJ3WgYXxjkMUkNNx7fLKAF"),
    scale: 9,
});
exports.zigToken = Object.freeze({
    tag: "ZIG",
    name: "Zignaly",
    mint: new web3_js_1.PublicKey("ANqY8h3sRSwkC29FvGJenAh7VGRABVVx7Ls6Mq4BuGT"),
    scale: 9,
});
exports.daosolToken = Object.freeze({
    tag: "daoSOL",
    name: "daoSOL Token",
    mint: new web3_js_1.PublicKey("GEJpt3Wjmr628FqXxTgxMce1pLntcPV4uFi8ksxMyPQh"),
    scale: 9,
});
exports.msolToken = Object.freeze({
    tag: "mSOL",
    name: "Marinade.finance",
    mint: new web3_js_1.PublicKey("mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So"),
    scale: 9,
});
exports.psolToken = Object.freeze({
    tag: "pSOL",
    name: "pSOL",
    mint: new web3_js_1.PublicKey("9EaLkQrbjmbbuZG9Wdpo8qfNUEjHATJFSycEmw6f1rGX"),
    scale: 9,
});
exports.srlyv2Token = Object.freeze({
    tag: "sRLYv2",
    name: "Rally (Solana)",
    mint: new web3_js_1.PublicKey("RLYv2ubRMDLcGG2UyvPmnPmkfuQTsMbg4Jtygc7dmnq"),
    scale: 9,
});
exports.srlyToken = Object.freeze({
    tag: "sRLY",
    name: "Rally Solana",
    mint: new web3_js_1.PublicKey("sRLY3migNrkC1HLgqotpvi66qGkdNedqPZ9TJpAQhyh"),
    scale: 9,
});
exports.scnsolToken = Object.freeze({
    tag: "scnSOL",
    name: "Socean Staked Sol",
    mint: new web3_js_1.PublicKey("5oVNBeEEQvYi1cX3ir8Dx5n1P7pdxydbGF2X4TxVusJm"),
    scale: 9,
});
exports.stsolToken = Object.freeze({
    tag: "stSOL",
    name: "Lido Staked SOL",
    mint: new web3_js_1.PublicKey("7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj"),
    scale: 9,
});
exports.whapiToken = Object.freeze({
    tag: "wHAPI",
    name: "HAPI",
    mint: new web3_js_1.PublicKey("6VNKqgz9hk7zRShTFdg5AnkfKwZUcojzwAkzxSH3bnUm"),
    scale: 9,
});
exports.wldoToken = Object.freeze({
    tag: "wLDO",
    name: "Lido DAO",
    mint: new web3_js_1.PublicKey("HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p"),
    scale: 8,
});
exports.wlunaToken = Object.freeze({
    tag: "wLUNA",
    name: "Terra",
    mint: new web3_js_1.PublicKey("F6v4wfAdJB8D8p77bMXZgYt8TDKsYxLYxH5AFhUkYx9W"),
    scale: 6,
});
exports.wustToken = Object.freeze({
    tag: "wUST",
    name: "TerraUSD (Wormhole)",
    mint: new web3_js_1.PublicKey("9vMJfxuKxXBoEa7rM12mYLMwTacLMLDJqHozw96WQL8i"),
    scale: 6,
});
exports.whethToken = Object.freeze({
    tag: "whETH",
    name: "Ethereum",
    mint: new web3_js_1.PublicKey("7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs"),
    scale: 8,
});
exports.wstethToken = Object.freeze({
    tag: "wstETH",
    name: "Lido Staked Ether",
    mint: new web3_js_1.PublicKey("ZScHuTtqZukUrtZS43teTKGs2VqkKL8k4QCouR2n6Uo"),
    scale: 8,
});
