// @switchboard-xyz/switchboardv2-cli/dist/commands/crank/turn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const web3_js_1 = require("@solana/web3.js");
const sbv2_utils_1 = require("@switchboard-xyz/sbv2-utils");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const BaseCommand_1 = tslib_1.__importDefault(require("../../BaseCommand"));
const utils_1 = require("../../utils");
class CrankTurn extends BaseCommand_1.default {
    async run() {
        const { args } = await this.parse(CrankTurn);
        (0, utils_1.verifyProgramHasPayer)(this.program);
        const payer = (0, switchboard_v2_1.programWallet)(this.program);
        // load crank
        const crankAccount = new switchboard_v2_1.CrankAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.crankKey),
        });
        const crank = await crankAccount.loadData();
        // load queue
        const queueAccount = new switchboard_v2_1.OracleQueueAccount({
            program: this.program,
            publicKey: crank.queuePubkey,
        });
        const queue = await queueAccount.loadData();
        // load program state
        const [programStateAccount] = switchboard_v2_1.ProgramStateAccount.fromSeed(this.program);
        const progamState = await programStateAccount.loadData();
        // get payer payout wallet
        const mint = await programStateAccount.getTokenMint();
        const payoutTokenAddress = await (0, sbv2_utils_1.getOrCreateSwitchboardTokenAccount)(this.program, mint, payer);
        const txn = await crankAccount.pop({
            payoutWallet: payoutTokenAddress,
            queuePubkey: queueAccount.publicKey,
            queueAuthority: queue.authority,
            crank: 0,
            queue: 0,
            tokenMint: progamState.tokenMintPublicKey,
        });
        if (this.silent) {
            console.log(txn);
        }
        else {
            this.logger.log(`${chalk_1.default.green(`${utils_1.CHECK_ICON}Crank turned successfully`)}`);
            this.logger.log(`https://explorer.solana.com/tx/${txn}?cluster=${this.cluster}`);
        }
    }
    async catch(error) {
        super.catch(error, "failed to turn the crank");
    }
}
exports.default = CrankTurn;
CrankTurn.description = "turn the crank and get rewarded if aggregator updates available";
CrankTurn.flags = {
    ...BaseCommand_1.default.flags,
};
CrankTurn.args = [
    {
        name: "crankKey",
        description: "public key of the crank to turn",
    },
];
CrankTurn.examples = [
    "$ sbv2 crank:turn 85L2cFUvXaeGQ4HrzP8RJEVCL7WvRrXM2msvEmQ82AVr --keypair ../payer-keypair.json",
];
