// @switchboard-xyz/switchboardv2-cli/dist/commands/aggregator/remove/job.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@oclif/core");
const web3_js_1 = require("@solana/web3.js");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const BaseCommand_1 = tslib_1.__importDefault(require("../../../BaseCommand"));
const utils_1 = require("../../../utils");
class AggregatorRemoveJob extends BaseCommand_1.default {
    async run() {
        const { args, flags } = await this.parse(AggregatorRemoveJob);
        (0, utils_1.verifyProgramHasPayer)(this.program);
        const aggregatorAccount = new switchboard_v2_1.AggregatorAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.aggregatorKey),
        });
        const aggregator = await aggregatorAccount.loadData();
        const authority = await this.loadAuthority(flags.authority, aggregator.authority);
        const jobAccount = new switchboard_v2_1.JobAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.jobKey),
        });
        const txn = await aggregatorAccount.removeJob(jobAccount, authority);
        if (this.silent) {
            console.log(txn);
        }
        else {
            this.logger.log(`${chalk_1.default.green(`${utils_1.CHECK_ICON}Job succesfully removed from aggregator account\r\n`)}`);
            this.logger.log(`https://explorer.solana.com/tx/${txn}?cluster=${this.cluster}`);
        }
    }
    async catch(error) {
        super.catch(error, "failed to remove job to aggregator account");
    }
}
exports.default = AggregatorRemoveJob;
AggregatorRemoveJob.description = "remove a switchboard job account from an aggregator";
AggregatorRemoveJob.flags = {
    ...BaseCommand_1.default.flags,
    authority: core_1.Flags.string({
        char: "a",
        description: "alternate keypair that is the authority for the aggregator",
    }),
};
AggregatorRemoveJob.args = [
    {
        name: "aggregatorKey",
        description: "public key of the aggregator account",
    },
    {
        name: "jobKey",
        description: "public key of an existing job account to remove from an aggregator",
    },
];
AggregatorRemoveJob.examples = ["$ sbv2 aggregator:remove:job"];
