// @switchboard-xyz/switchboardv2-cli/dist/commands/print/buffer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/* eslint-disable unicorn/import-style */
const core_1 = require("@oclif/core");
const sbv2_utils_1 = require("@switchboard-xyz/sbv2-utils");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const BaseCommand_1 = tslib_1.__importDefault(require("../../BaseCommand"));
class BufferPrint extends BaseCommand_1.default {
    async run() {
        const { args, flags } = await this.parse(BufferPrint);
        const bufferAccount = new switchboard_v2_1.BufferRelayerAccount({
            program: this.program,
            publicKey: args.bufferRelayerKey,
        });
        const bufferAccountData = await bufferAccount.loadData();
        this.logger.log(await (0, sbv2_utils_1.prettyPrintBufferRelayer)(bufferAccount, bufferAccountData, flags.job));
    }
    async catch(error) {
        super.catch(error, "failed to print buffer relayer account");
    }
}
exports.default = BufferPrint;
BufferPrint.description = "Print the deserialized Switchboard buffer relayer account";
BufferPrint.aliases = ["buffer:print"];
BufferPrint.flags = {
    ...BaseCommand_1.default.flags,
    job: core_1.Flags.boolean({
        description: "output job definitions",
        default: false,
    }),
};
BufferPrint.args = [
    {
        name: "bufferRelayerKey",
        description: "public key of the buffer relayer account to deserialize",
    },
];
BufferPrint.examples = [
    "$ sbv2 buffer:print 23GvzENjwgqqaLejsAtAWgTkSzWjSMo2LUYTAETT8URp",
];
