// @switchboard-xyz/switchboardv2-cli/dist/commands/print/crank.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const web3_js_1 = require("@solana/web3.js");
const sbv2_utils_1 = require("@switchboard-xyz/sbv2-utils");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const BaseCommand_1 = tslib_1.__importDefault(require("../../BaseCommand"));
class CrankPrint extends BaseCommand_1.default {
    async run() {
        const { args, flags } = await this.parse(CrankPrint);
        const crankAccount = new switchboard_v2_1.CrankAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.crankKey),
        });
        this.logger.log(await (0, sbv2_utils_1.prettyPrintCrank)(crankAccount, undefined, true));
    }
    async catch(error) {
        super.catch(error, "failed to print crank account");
    }
}
exports.default = CrankPrint;
CrankPrint.description = "print deserialized switchboard crank account";
CrankPrint.aliases = ["crank:print"];
CrankPrint.flags = {
    ...BaseCommand_1.default.flags,
};
CrankPrint.args = [
    {
        name: "crankKey",
        description: "public key of the crank account to deserialize",
    },
];
CrankPrint.examples = [
    "$ sbv2 crank:print 85L2cFUvXaeGQ4HrzP8RJEVCL7WvRrXM2msvEmQ82AVr",
];
