// @switchboard-xyz/switchboardv2-cli/dist/commands/print/job.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const web3_js_1 = require("@solana/web3.js");
const sbv2_utils_1 = require("@switchboard-xyz/sbv2-utils");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const BaseCommand_1 = tslib_1.__importDefault(require("../../BaseCommand"));
class JobPrint extends BaseCommand_1.default {
    async run() {
        const { args, flags } = await this.parse(JobPrint);
        const jobAccount = new switchboard_v2_1.JobAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.jobKey),
        });
        this.logger.log(await (0, sbv2_utils_1.prettyPrintJob)(jobAccount));
    }
    async catch(error) {
        super.catch(error, "failed to print job account");
    }
}
exports.default = JobPrint;
JobPrint.description = "Print the deserialized Switchboard job account";
JobPrint.aliases = ["job:print"];
JobPrint.flags = {
    ...BaseCommand_1.default.flags,
};
JobPrint.args = [
    {
        name: "jobKey",
        description: "public key of the job account to deserialize",
    },
];
JobPrint.examples = [
    "$ sbv2 job:print SzTvFZLz3hwjZFMwVWzuEnr1oUF6qyvXwXCvsqf7qeA",
];
