// @switchboard-xyz/switchboardv2-cli/dist/commands/print/oracle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const web3_js_1 = require("@solana/web3.js");
const sbv2_utils_1 = require("@switchboard-xyz/sbv2-utils");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const BaseCommand_1 = tslib_1.__importDefault(require("../../BaseCommand"));
class OraclePrint extends BaseCommand_1.default {
    async run() {
        const { args, flags } = await this.parse(OraclePrint);
        const oracleAccount = new switchboard_v2_1.OracleAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.oracleKey),
        });
        const data = await oracleAccount.loadData();
        this.logger.log(await (0, sbv2_utils_1.prettyPrintOracle)(oracleAccount, data, true));
    }
    async catch(error) {
        super.catch(error, "failed to print oracle account");
    }
}
exports.default = OraclePrint;
OraclePrint.description = "Print the deserialized Switchboard oracle account";
OraclePrint.aliases = ["oracle:print"];
OraclePrint.flags = {
    ...BaseCommand_1.default.flags,
};
OraclePrint.args = [
    {
        name: "oracleKey",
        description: "public key of the oracle account to deserialize",
    },
];
OraclePrint.examples = [
    "$ sbv2 oracle:print 9CmLriMhykZ8xAoNTSHjHbk6SkuMhie1NCZn9P6LCuZ4",
];
