// @switchboard-xyz/switchboardv2-cli/dist/commands/print/permission.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const web3_js_1 = require("@solana/web3.js");
const sbv2_utils_1 = require("@switchboard-xyz/sbv2-utils");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const BaseCommand_1 = tslib_1.__importDefault(require("../../BaseCommand"));
class PermissionPrint extends BaseCommand_1.default {
    async run() {
        const { args, flags } = await this.parse(PermissionPrint);
        const permissionAccount = new switchboard_v2_1.PermissionAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.permissionKey),
        });
        const data = await permissionAccount.loadData();
        this.logger.log(await (0, sbv2_utils_1.prettyPrintPermissions)(permissionAccount, data));
    }
    async catch(error) {
        super.catch(error, "failed to print permission account");
    }
}
exports.default = PermissionPrint;
PermissionPrint.description = "Print the deserialized Switchboard permission account";
PermissionPrint.aliases = ["permission:print"];
PermissionPrint.flags = {
    ...BaseCommand_1.default.flags,
};
PermissionPrint.args = [
    {
        name: "permissionKey",
        description: "public key of the permission account to deserialize",
    },
];
PermissionPrint.examples = [
    "$ sbv2 permission:print 94XXM72K2aKu2wcuJaawV8njuGaFZvhy8iKgPxoa1tJk",
];
