// @switchboard-xyz/switchboardv2-cli/dist/commands/print/vrf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const web3_js_1 = require("@solana/web3.js");
const sbv2_utils_1 = require("@switchboard-xyz/sbv2-utils");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const BaseCommand_1 = tslib_1.__importDefault(require("../../BaseCommand"));
class VrfPrint extends BaseCommand_1.default {
    async run() {
        const { args, flags } = await this.parse(VrfPrint);
        const vrfAccount = new switchboard_v2_1.VrfAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.vrfKey),
        });
        this.logger.log(await (0, sbv2_utils_1.prettyPrintVrf)(vrfAccount, await vrfAccount.loadData(), true));
    }
    async catch(error) {
        super.catch(error, "failed to print vrf account");
    }
}
exports.default = VrfPrint;
VrfPrint.enableJsonFlag = true;
VrfPrint.description = "Print the deserialized Switchboard VRF account";
VrfPrint.aliases = ["vrf:print"];
VrfPrint.flags = {
    ...BaseCommand_1.default.flags,
};
VrfPrint.args = [
    {
        name: "vrfKey",
        description: "public key of the vrf account to deserialize",
    },
];
VrfPrint.examples = ["$ sbv2 vrf:print"];
