// @switchboard-xyz/sbv2-utils/lib/cjs/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendAndConfirmRawTransaction = exports.packAndSend = void 0;
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
async function packAndSend(program, ixnsBatches, signers, feePayer) {
    const signatures = [];
    for await (const batch of ixnsBatches) {
        const { blockhash } = await program.provider.connection.getLatestBlockhash();
        const packedTransactions = (0, switchboard_v2_1.packInstructions)(batch, feePayer, blockhash);
        const signedTransactions = (0, switchboard_v2_1.signTransactions)(packedTransactions, signers);
        const signedTxs = await program.provider.wallet.signAllTransactions(signedTransactions);
        for (let k = 0; k < packedTransactions.length; k += 1) {
            const tx = signedTxs[k];
            const rawTx = tx.serialize();
            signatures.push(sendAndConfirmRawTransaction(program.provider.connection, rawTx, {
                maxRetries: 10,
                commitment: "processed",
            }).catch((error) => {
                console.error(error);
                throw error;
            }));
        }
        await Promise.all(signatures);
    }
    return Promise.all(signatures);
}
exports.packAndSend = packAndSend;
/**
 * Send and confirm a raw transaction
 *
 * If `commitment` option is not specified, defaults to 'max' commitment.
 */
async function sendAndConfirmRawTransaction(connection, rawTransaction, options) {
    const sendOptions = options && {
        skipPreflight: options.skipPreflight,
        preflightCommitment: options.preflightCommitment || options.commitment,
    };
    const signature = await connection.sendRawTransaction(rawTransaction, sendOptions);
    const status = (await connection.confirmTransaction(signature, options.commitment || "max")).value;
    if (status.err) {
        throw new Error(`Raw transaction ${signature} failed (${JSON.stringify(status)})`);
    }
    return signature;
}
exports.sendAndConfirmRawTransaction = sendAndConfirmRawTransaction;
