// @switchboard-xyz/switchboardv2-cli/dist/commands/print/oracle/permission.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const web3_js_1 = require("@solana/web3.js");
const sbv2_utils_1 = require("@switchboard-xyz/sbv2-utils");
const switchboard_v2_1 = require("@switchboard-xyz/switchboard-v2");
const BaseCommand_1 = tslib_1.__importDefault(require("../../../BaseCommand"));
class OraclePermissionPrint extends BaseCommand_1.default {
    async run() {
        const { args } = await this.parse(OraclePermissionPrint);
        const oracleAccount = new switchboard_v2_1.OracleAccount({
            program: this.program,
            publicKey: new web3_js_1.PublicKey(args.oracleKey),
        });
        const oracle = await oracleAccount.loadData();
        const queueAccount = new switchboard_v2_1.OracleQueueAccount({
            program: this.program,
            publicKey: oracle.queuePubkey,
        });
        const queue = await queueAccount.loadData();
        const [permissionAccount] = switchboard_v2_1.PermissionAccount.fromSeed(this.program, queue.authority, queueAccount.publicKey, oracleAccount.publicKey);
        try {
            const permission = await permissionAccount.loadData();
            this.logger.log(await (0, sbv2_utils_1.prettyPrintPermissions)(permissionAccount, permission));
        }
        catch {
            this.logger.error(`failed to find a permission account`);
        }
    }
    async catch(error) {
        super.catch(error, "failed to print oracle permission account");
    }
}
exports.default = OraclePermissionPrint;
OraclePermissionPrint.description = "Print the permission account associated with a Switchboard oracle account";
OraclePermissionPrint.aliases = ["oracle:permission:print", "oracle:print:permission"];
OraclePermissionPrint.flags = {
    ...BaseCommand_1.default.flags,
};
OraclePermissionPrint.args = [
    {
        name: "oracleKey",
        description: "public key of the oracle account to fetch permission account and deserialize",
    },
];
OraclePermissionPrint.examples = [
    "$ sbv2 oracle:permission:print 9CmLriMhykZ8xAoNTSHjHbk6SkuMhie1NCZn9P6LCuZ4",
];
