// @orca-so/sdk/dist/constants/devnet/farms.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.orcaDevnetFarmConfigs = exports.ethUsdcDoubleDip = exports.ethSolDoubleDip = exports.solUsdcDoubleDip = exports.orcaUsdcAqFarm = exports.orcaSolAqFarm = exports.ethUsdcAqFarm = exports.ethSolAqFarm = exports.usdcUsdtAqFarm = exports.solUsdtAqFarm = exports.solUsdcAqFarm = void 0;
const web3_js_1 = require("@solana/web3.js");
/**
 * The following content is auto-generated.
 */
exports.solUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("CzFnr34cRC1o8yrs5jTHBqDQ1FDckLapkPis6zTfqCF5"),
    farmTokenMint: new web3_js_1.PublicKey("AtJF9fbQ1pbz76NYo2jtKjHuzzyaFToHLCizYk6UoHHL"),
    rewardTokenMint: new web3_js_1.PublicKey("orcarKHSqC5CDDsGbho8GKvwExejWHxTqGzXgcewB9L"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("4GpUivZ2jvZqQ3vJRsoq5PwnYv6gdV9fJ9BzHT2JcRr7"),
    baseTokenDecimals: 6,
});
exports.solUsdtAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("ER1UZsRhzucvXY8xMmXGfbbxb442iTHk2u1rSKRMPakC"),
    farmTokenMint: new web3_js_1.PublicKey("E95ABqtqnJRMG6gqEKZ16dVMfSNDUdCbEYqGUVg3tFRX"),
    rewardTokenMint: new web3_js_1.PublicKey("orcarKHSqC5CDDsGbho8GKvwExejWHxTqGzXgcewB9L"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("2E4Mp6y2gFRteXiANnwzAJEhHwD3NX1wru3XvxJpGBq2"),
    baseTokenDecimals: 6,
});
exports.usdcUsdtAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("D7wWV4nxqEeCAk2UA2nEGWgKw2BgoZYFZFUfwWGBhDF8"),
    farmTokenMint: new web3_js_1.PublicKey("3dDqVehvEpc5HMcbJUAzwMwF54xqa4nNvYN9qis7HAPe"),
    rewardTokenMint: new web3_js_1.PublicKey("orcarKHSqC5CDDsGbho8GKvwExejWHxTqGzXgcewB9L"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("EBor1PR5XNGHjRVB6JDJuKVCQbvdr1CVZTaX1hTAdvQv"),
    baseTokenDecimals: 6,
});
exports.ethSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("3q4wtMRT2yKHzQGy6gxwD34i9rsafviKd78rpdDLrBfk"),
    farmTokenMint: new web3_js_1.PublicKey("5NH9rNaoLKbPPP5zwCJTaMPkFACtfJA8wEmqWrXzgCMa"),
    rewardTokenMint: new web3_js_1.PublicKey("orcarKHSqC5CDDsGbho8GKvwExejWHxTqGzXgcewB9L"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("8sFnpd7mM1AWxP1LXX2FWbbkaVtAopBPmPNZ9y6172WL"),
    baseTokenDecimals: 6,
});
exports.ethUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("3aRTv8TGtZkX4dsM4N9aUHKotcwRr6phWPHXPLN8DEYX"),
    farmTokenMint: new web3_js_1.PublicKey("EeqPcczEZH2cjYBnQyGXBQx1DGs7KG1pobwdPKcwALhD"),
    rewardTokenMint: new web3_js_1.PublicKey("orcarKHSqC5CDDsGbho8GKvwExejWHxTqGzXgcewB9L"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("9pRnvg7ihSJDLi6DGf3PLwr6xRRRrBPXsHYEgGL5hzgA"),
    baseTokenDecimals: 6,
});
exports.orcaSolAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("6YrLcQs5yFvXkRY5VkMGEfVgo5rwozJf7jXedpZxbKmi"),
    farmTokenMint: new web3_js_1.PublicKey("3z8o3b4gMBpnRsrDv7ruZPcVtgoULMFyEoEEGwTsw2TR"),
    rewardTokenMint: new web3_js_1.PublicKey("orcarKHSqC5CDDsGbho8GKvwExejWHxTqGzXgcewB9L"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("CmDdQhusZWyi9fue27VSktYgkHefm3JXNdzc9kCpyvYi"),
    baseTokenDecimals: 6,
});
exports.orcaUsdcAqFarm = Object.freeze({
    address: new web3_js_1.PublicKey("44GPPv5XedLXCNCc3Mbay1T18vc6x2bWihmmg7UJanvH"),
    farmTokenMint: new web3_js_1.PublicKey("52Tv8d6Z7Lb27dqCufZPUqadZBTv3EAKGzoV6hy185As"),
    rewardTokenMint: new web3_js_1.PublicKey("orcarKHSqC5CDDsGbho8GKvwExejWHxTqGzXgcewB9L"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("2ZEEntzoUN7XuMs88ukLGv5HRR1byL7wFWChryF5ZHri"),
    baseTokenDecimals: 6,
});
exports.solUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("L3XdTwLLtZRrXqSk5mqcBKaorqgmM2AMDGmzfBuhLMd"),
    farmTokenMint: new web3_js_1.PublicKey("7AYG2y3je8TqcLsugNkV6aPw14QyqdVFmK5EegeJmRFi"),
    rewardTokenMint: new web3_js_1.PublicKey("Ff5JqsAYUD4vAfQUtfRprT4nXu9e28tTBZTDFMnJNdvd"),
    rewardTokenDecimals: 9,
    baseTokenMint: new web3_js_1.PublicKey("AtJF9fbQ1pbz76NYo2jtKjHuzzyaFToHLCizYk6UoHHL"),
    baseTokenDecimals: 6,
});
exports.ethSolDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("5jGfcgoRNpZPwfiRTp5zxNpEKP3PTeM3F9EKS9EsvX67"),
    farmTokenMint: new web3_js_1.PublicKey("5Uv5T1BksJGKGFGVpfxGeGniFFgGLjavNpPWGkfUFNkr"),
    rewardTokenMint: new web3_js_1.PublicKey("6PE3Mwjzx9h8kCoBp5YPed9TFoG7du8L98yucBP5ps3x"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("5NH9rNaoLKbPPP5zwCJTaMPkFACtfJA8wEmqWrXzgCMa"),
    baseTokenDecimals: 6,
});
exports.ethUsdcDoubleDip = Object.freeze({
    address: new web3_js_1.PublicKey("B4CTmaFpNRtXL3ZLiLrgbMZknQHFdoGu4V9tbzsjFbRT"),
    farmTokenMint: new web3_js_1.PublicKey("5jdeMPvNrscuw8abZFf99X9kcxDhANcpuBHDNpnx7YPT"),
    rewardTokenMint: new web3_js_1.PublicKey("EmXq3Ni9gfudTiyNKzzYvpnQqnJEMRw2ttnVXoJXjLo1"),
    rewardTokenDecimals: 6,
    baseTokenMint: new web3_js_1.PublicKey("EeqPcczEZH2cjYBnQyGXBQx1DGs7KG1pobwdPKcwALhD"),
    baseTokenDecimals: 6,
});
/**
 * Mapping for OrcaFarm parameters
 * Key: baseTokenMint : OrcaFarmParams
 */
exports.orcaDevnetFarmConfigs = {
    "4GpUivZ2jvZqQ3vJRsoq5PwnYv6gdV9fJ9BzHT2JcRr7": exports.solUsdcAqFarm,
    "2E4Mp6y2gFRteXiANnwzAJEhHwD3NX1wru3XvxJpGBq2": exports.solUsdtAqFarm,
    EBor1PR5XNGHjRVB6JDJuKVCQbvdr1CVZTaX1hTAdvQv: exports.usdcUsdtAqFarm,
    "8sFnpd7mM1AWxP1LXX2FWbbkaVtAopBPmPNZ9y6172WL": exports.ethSolAqFarm,
    "9pRnvg7ihSJDLi6DGf3PLwr6xRRRrBPXsHYEgGL5hzgA": exports.ethUsdcAqFarm,
    CmDdQhusZWyi9fue27VSktYgkHefm3JXNdzc9kCpyvYi: exports.orcaSolAqFarm,
    "2ZEEntzoUN7XuMs88ukLGv5HRR1byL7wFWChryF5ZHri": exports.orcaUsdcAqFarm,
    AtJF9fbQ1pbz76NYo2jtKjHuzzyaFToHLCizYk6UoHHL: exports.solUsdcDoubleDip,
    "5NH9rNaoLKbPPP5zwCJTaMPkFACtfJA8wEmqWrXzgCMa": exports.ethSolDoubleDip,
    EeqPcczEZH2cjYBnQyGXBQx1DGs7KG1pobwdPKcwALhD: exports.ethUsdcDoubleDip,
};
